// @stack-spot/portal-network@0.66.0/src/error/dictionary/rte.ts
import { Dictionary } from '@stack-spot/portal-translate'

export const rteDictionary = {
  en: {
    RTE_400_BAD_REQUEST: 'Some fields are incorrect. Please check and try again.',
    RTE_400_WORKSPACE_ID_INVALID: 'Invalid Workspace. Please select another.',
    RTE_400_ENVIRONMENT_ID_INVALID: 'Invalid Environment. Please select another.',
    RTE_400_WORKSPACE_AND_ENVIRONMENT_NOT_INFORMED: 'To continue, enter the Environment and Workspace.',
    RTE_422_ROLE_INVALID: 'Invalid ARN. Please check the ARN.',
    RTE_422_CLOUD_ACCOUNT_ALREADY_EXIST: 'This cloud account is already connected. Please use another account.',
    RTE_422_ASSUME_ROLE_INVALID: 'Unable to assume the role. Please check permissions and try again.',
    RTE_404_APPLICATION_NOT_FOUND: 'To save, deploy the Application or Infrastructure first.',
  },
  pt: {
    RTE_400_BAD_REQUEST: 'Alguns campos estão incorretos. Por favor, verifique e tente novamente.',
    RTE_400_WORKSPACE_ID_INVALID: 'Workspace inválido. Por favor, selecione outro.',
    RTE_400_ENVIRONMENT_ID_INVALID: 'Ambiente inválido. Por favor, selecione outro.',
    RTE_400_WORKSPACE_AND_ENVIRONMENT_NOT_INFORMED: 'Para continuar, informe o Ambiente e o Workspace.',
    RTE_422_ROLE_INVALID: 'ARN inválido. Por favor, verifique o ARN informado.',
    RTE_422_CLOUD_ACCOUNT_ALREADY_EXIST: 'Essa conta de cloud já está conectada. Por favor, use outra conta.',
    RTE_422_ASSUME_ROLE_INVALID: 'Não foi possível assumir a role. Por favor, verifique as permissões e tente novamente.',
    RTE_404_APPLICATION_NOT_FOUND: 'Para salvar, primeiro faça deploy da Aplicação ou Infraestrutura.',
  },
} satisfies Dictionary
