// @stack-spot/portal-extension-backend@0.4.0/dist/operation-handler.js
import { jsx as _jsx } from "react/jsx-runtime";
import { SessionManager } from '@stack-spot/auth-react';
import { CitronNavigator } from '@stack-spot/citron-navigator';
import { overlay } from '@stack-spot/portal-layout';
import { Extension } from './Extension';
import { handleNetworkOperation } from './network-handler';
const operationHandler = {
    showModal: ({ variables: { path, ...variables }, slug, version }) => overlay.showModal({
        children: _jsx(Extension, { slug: slug, version: version, path: path }),
        ...variables,
    }),
    showRightPanel: ({ variables: { path, ...variables }, slug, version }) => overlay.showRightPanel({
        children: _jsx(Extension, { slug: slug, version: version, path: path }),
        ...variables,
    }),
    showCustomModal: ({ variables: { path, ...variables }, slug, version }) => overlay.showCustomModal(_jsx(Extension, { slug: slug, version: version, path: path }), variables),
    showCustomRightPanel: ({ variables: { path, ...variables }, slug, version }) => overlay.showCustomRightPanel(_jsx(Extension, { slug: slug, version: version, path: path }), variables),
    closeModal: () => overlay.closeModal(),
    closeRightPanel: () => overlay.closeRightPanel(),
    alert: ({ variables }) => overlay.alert(variables),
    confirm: ({ variables }) => overlay.confirm(variables),
    getExtensionSlug: ({ slug }) => slug,
    getExtensionVersion: ({ version }) => version,
    getUserData: () => {
        const { account_name = '', email = '' } = SessionManager.instance?.getSession()?.getTokenData() ?? {};
        return { name: account_name, email };
    },
    // @ts-ignore
    showToaster: ({ variables }) => overlay.showToaster(variables),
    getCurrentPage: () => {
        const key = CitronNavigator.instance?.currentRoute?.$key ?? '';
        const params = CitronNavigator.instance?.currentParams ?? {};
        return { key, params };
    },
    network: ({ variables }) => handleNetworkOperation(variables),
};
export function handleOperation(options) {
    return operationHandler[options.operation](options);
}
