// @stack-spot/portal-theme@1.1.0/dist/CSSToCitricAdapter.js
import { jsx as _jsx } from "react/jsx-runtime";
import { defaultTheme } from '@citric/core';
import { ThemeProvider } from 'styled-components';
import { theme } from './definition.js';
const citricTheme = {
    ...defaultTheme,
    colorScheme: {
        primary: theme.color.primary,
        secondary: theme.color.secondary,
        tertiary: theme.color.tertiary,
        success: theme.color.success,
        warning: theme.color.warning,
        danger: theme.color.danger,
        light: theme.color.light,
        inverse: theme.color.inverse,
        highlight: theme.color.highlight,
    },
    palette: {
        blue: theme.color.blue,
        cyan: theme.color.cyan,
        indigo: theme.color.indigo,
        moss: theme.color.moss,
        orange: theme.color.orange,
        pink: theme.color.pink,
        purple: theme.color.purple,
        red: theme.color.red,
        teal: theme.color.teal,
        yellow: theme.color.yellow,
        gray: theme.color.gray,
    },
};
/**
 * Citric is themed by Styled Components, while the StackSpot Portals are themed by CSS variables. This component creates an adapter from
 * CSS to a Styled Components theme.
 *
 * It is imperative to have this component on the tree before using any component from Citric.
 */
export const CSSToCitricAdapter = ({ children }) => (_jsx(ThemeProvider, { theme: citricTheme, children: children }));
