// @stack-spot/portal-extension-commons@0.3.0/dist/factory.js
let id = 0;
function next() {
    id = id === Number.MAX_SAFE_INTEGER ? 1 : id + 1;
    return id;
}
export function createReadyMessage() {
    return {
        signal: 'ready',
        id: next(),
        type: 'stackspot-extension-message',
    };
}
export function createResizeMessage(width, height) {
    return {
        signal: 'resize',
        id: next(),
        type: 'stackspot-extension-message',
        width,
        height,
    };
}
export function createConfigMessage(config) {
    return {
        signal: 'config',
        id: next(),
        type: 'stackspot-extension-message',
        ...config,
    };
}
export function createRequestMessage(...[operation, variables]) {
    return {
        type: 'stackspot-extension-message',
        id: next(),
        signal: 'request',
        operation,
        variables,
    };
}
export function createSuccessfulResponseMessage(operation, id, result) {
    return {
        type: 'stackspot-extension-message',
        id,
        signal: 'response-success',
        operation,
        result,
    };
}
export function createUnsuccessfulResponseMessage(operation, id, error, status) {
    return {
        type: 'stackspot-extension-message',
        id,
        signal: 'response-error',
        operation,
        error,
        status,
    };
}
