// @stack-spot/portal-layout@2.20.1/dist/components/menu/MenuSectionGroup.js
import { theme } from '@stack-spot/portal-theme';
import styled from 'styled-components';
import { MENU_CONTENT_PADDING as PADDING } from './constants.js';
export const MenuSectionGroup = styled.ul `
  padding: 0 0 0 16px;
  display: flex;
  flex-direction: column;
  visibility: hidden;
  transition: visibility 0s 0.3s;

  &.no-indentation {
    padding: 0;
  }

  .item-row {
    display: flex;
    flex-direction: row;
    gap: 8px;
    align-items: center;

    &.root {
      padding: 0 16px;
      margin-top: 16px;
      border-radius: 0;
    }

    .label {
      flex: 1;
      &.hidden, &.ellipsis {
        white-space: nowrap;
        overflow: hidden;
      }
      &.ellipsis {
        text-overflow: ellipsis;
      }
    }
  }

  li a {
    position: relative;
    height: 0;
    overflow: hidden;
    transition: height 0.3s, background-color 0.2s;

    &:hover {
      background-color: ${theme.color.light['500']};
    }
    
    &.action {
      &:before {
        content: '';
        position: absolute;
        left: 2px;
        width: 2px;
        height: 0;
        background: inherit;
        transition: height 0.2s;
      }

      &.active {
  
        &:hover {
          background-color: transparent;
        }
  
        &:before {
          background: ${theme.color.primary['500']};
          height: 24px;
        }
      }

      &:not(.active):hover:before {
        background: ${theme.color.light.contrastText};
        height: 24px;
      }
    }

    .chevron {
      transition: transform 0.5s;
      &:not(.open) {
        transform: rotate(180deg);
      }
    }

    .item-row-title {
      opacity: 0.7;
    }
  }

  &.open {
    visibility: visible;
    transition: unset;
    & > li > a  {
      height: 56px;
    }
  }

  &:not(.open) &.open > li > a {
    height: 0;
  }

  &.root {
    /* margin-bottom: ${PADDING}px; */

    & > li {
      .group-title {
        margin-left: ${PADDING}px;
        margin-bottom: 5px;
        margin-top: 40px;
        display: block;
      }

      &:first-child {
        .group-title {
          margin-top: 0;
        }
      }
    }
  }
`;
