// @citric/core@6.4.0/dist/default-theme.js
const defaultTheme = {
    globalStyle: {
        body: {
            margin: '0',
            font: 'body1',
            color: 'light.contrastText',
        },
    },
    radius: {
        full: '600rem',
        xxl: '8rem',
        xl: '4rem',
        lg: '2rem',
        md: '1rem',
        sm: '0.5rem',
        xs: '0.25rem',
        xxs: '0.125rem',
    },
    breakpoint: {
        xl: 1400,
        lg: 1200,
        md: 992,
        sm: 768,
        xs: 576,
    },
    spacing: (value) => {
        if (value === 0) {
            return '0rem';
        }
        return value === 1 ? '0.125rem' : `${0.25 * (value - 1)}rem`;
    },
    container: {
        width: {
            default: '1168px',
            xl: '1168px',
            lg: '976px',
            md: '776px',
            sm: '540px',
            xs: '540px',
        },
    },
    icon: {
        xs: {
            size: '0.75rem',
            spacing: '1rem',
        },
        sm: {
            size: '1rem',
            spacing: '1.5rem',
        },
        md: {
            size: '1.25rem',
            spacing: '2rem',
        },
        lg: {
            size: '1.5rem',
            spacing: '3rem',
        },
    },
    iconButton: {
        xs: {
            size: '0.75rem',
            spacing: '0.25rem',
        },
        sm: {
            size: '0.75rem',
            spacing: '0.375rem',
        },
        md: {
            size: '1rem',
            spacing: '0.5rem',
        },
        lg: {
            size: '1rem',
            spacing: '0.75rem',
        },
    },
    font: {
        family: {
            system1: 'Montserrat,sans-serif',
            system2: 'Roboto,sans-serif',
            code: 'Roboto Mono,monospace',
        },
        weight: {
            light: 300,
            regular: 400,
            medium: 500,
            semibold: 600,
            bold: 700,
        },
        appearance: {
            breadcrumb: {
                tag: 'p',
                weight: 'regular',
                size: '0.875rem',
                family: 'system2',
                lineHeight: '1.5',
            },
            button: {
                tag: 'button',
                weight: 'medium',
                size: '1rem',
                family: 'system2',
                lineHeight: '1.5',
            },
            body1: {
                tag: 'p',
                weight: 'regular',
                size: '1rem',
                family: 'system2',
                lineHeight: '1.5',
            },
            body2: {
                tag: 'p',
                weight: 'regular',
                size: '0.875rem',
                family: 'system2',
                lineHeight: '1.714',
            },
            code1: {
                tag: 'code',
                weight: 'regular',
                size: '1rem',
                family: 'code',
                lineHeight: '1.5',
            },
            code2: {
                tag: 'code',
                weight: 'regular',
                size: '0.875rem',
                family: 'code',
                lineHeight: '1.714',
            },
            display1: {
                tag: 'h1',
                weight: 'semibold',
                size: 'clamp(2.25rem,4vw + 1rem,4rem)',
                family: 'system1',
                lineHeight: '1.125',
                letterSpacing: '-0.1rem',
            },
            h1: {
                tag: 'h1',
                weight: 'semibold',
                size: 'clamp(2.25rem,4vw + 1rem,2.5rem)',
                family: 'system1',
                lineHeight: '1.2',
            },
            h2: {
                tag: 'h2',
                weight: 'semibold',
                size: 'clamp(1.5rem,4vw + 1rem,2rem)',
                family: 'system1',
                lineHeight: '1.25',
            },
            h3: {
                tag: 'h3',
                weight: 'semibold',
                size: 'clamp(1.75rem,4vw + 1rem,1.375rem)',
                family: 'system1',
                lineHeight: '1.333',
            },
            h4: {
                tag: 'h4',
                weight: 'semibold',
                size: 'clamp(1.125rem,4vw + 1rem,1.25rem)',
                family: 'system1',
                lineHeight: '1.2',
            },
            h5: {
                tag: 'h5',
                weight: 'semibold',
                size: '1rem',
                family: 'system1',
                lineHeight: '1.25',
            },
            h6: {
                tag: 'h6',
                weight: 'semibold',
                size: '0.875rem',
                family: 'system1',
                lineHeight: '1.429',
            },
            microtext1: {
                tag: 'small',
                weight: 'regular',
                size: '0.75rem',
                family: 'system2',
                lineHeight: '1.333',
            },
            subtitle1: {
                tag: 'p',
                weight: 'regular',
                size: '1.5rem',
                family: 'system1',
                lineHeight: '1.333',
            },
            subtitle2: {
                tag: 'p',
                weight: 'regular',
                size: '1.25rem',
                family: 'system1',
                lineHeight: '1.6',
            },
            subtitle3: {
                tag: 'p',
                weight: 'regular',
                size: '1rem',
                family: 'system1',
                lineHeight: '1.5',
            },
            subtitle4: {
                tag: 'p',
                weight: 'regular',
                size: '0.875rem',
                family: 'system1',
                lineHeight: '1.714',
            },
            overheader1: {
                tag: 'small',
                weight: 'semibold',
                size: '0.875rem',
                family: 'system1',
                lineHeight: '1.5',
                textTransform: 'uppercase',
            },
            overheader2: {
                tag: 'small',
                weight: 'semibold',
                size: '0.75rem',
                family: 'system1',
                lineHeight: '1.5',
                textTransform: 'uppercase',
            },
        },
    },
    colorScheme: {
        primary: {
            700: '#5C2512',
            600: '#AD4523',
            500: '#FF6633',
            400: '#FF9775',
            300: '#FFC8B5',
            contrastText: '#000000',
        },
        secondary: {
            700: '#14135C',
            600: '#2623AD',
            500: '#3834FF',
            400: '#7875FF',
            300: '#B7B6FF',
            contrastText: '#FFFFFF',
        },
        tertiary: {
            700: '#25124A',
            600: '#45238A',
            500: '#6633CC',
            400: '#9775DC',
            300: '#C8B5ED',
            contrastText: '#FFFFFF',
        },
        highlight: {
            700: '#5C3700',
            600: '#AD6800',
            500: '#FF9900',
            400: '#FFBA52',
            300: '#FFDAA3',
            contrastText: '#000000',
        },
        danger: {
            700: '#5C3232',
            600: '#AD5E5E',
            500: '#FF8A8A',
            400: '#FFB0B0',
            300: '#FFD5D5',
            contrastText: '#000000',
        },
        warning: {
            700: '#59512E',
            600: '#A89857',
            500: '#F8E080',
            400: '#FAEAA9',
            300: '#FCF4D1',
            contrastText: '#000000',
        },
        success: {
            700: '#344B34',
            600: '#628E62',
            500: '#91D191',
            400: '#B4E0B4',
            300: '#D7EED7',
            contrastText: '#000000',
        },
        light: {
            700: '#BCBCCF',
            600: '#2F2F3C',
            500: '#262631',
            400: '#1D1D25',
            300: '#14141A',
            contrastText: '#FAFAFB',
        },
        inverse: {
            700: '#707075',
            600: '#BCBCCF',
            500: '#D1D1DE',
            400: '#E5E5ED',
            300: '#F7F7F8',
            contrastText: '#0B0B0E',
        },
    },
    palette: {
        blue: {
            900: '#124A5C',
            800: '#1B6A85',
            700: '#238AAD',
            600: '#2BABD6',
            500: '#33CCFF',
            400: '#54D4FF',
            300: '#75DCFF',
            200: '#95E4FF',
            100: '#B5EDFF',
            50: '#D6F5FF',
        },
        cyan: {
            900: '#001A4D',
            800: '#004D80',
            700: '#0080B2',
            600: '#00B2E5',
            500: '#00E5FF',
            400: '#1AFFFF',
            300: '#4DFFFF',
            200: '#80FFFF',
            100: '#B3FFFF',
            50: '#E0FFFF',
        },
        gray: {
            900: '#000002',
            800: '#09090E',
            700: '#1C1C24',
            600: '#2F2F3C',
            500: '#444450',
            400: '#444450',
            300: '#6D6D77',
            200: '#97979E',
            100: '#C1C1C5',
            50: '#EAEAEC',
        },
        indigo: {
            900: '#14135C',
            800: '#1D1B85',
            700: '#2623AD',
            600: '#2F2CD6',
            500: '#3834FF',
            400: '#5855FF',
            300: '#7875FF',
            200: '#9795FF',
            100: '#B7B6FF',
            50: '#D7D6FF',
        },
        moss: {
            900: '#404600',
            800: '#5D6500',
            700: '#798400',
            600: '#95A300',
            500: '#B2C200',
            400: '#BECC29',
            300: '#CBD652',
            200: '#D7DF7A',
            100: '#E3E9A3',
            50: '#F0F3CC',
        },
        orange: {
            900: '#5C3112',
            800: '#85471B',
            700: '#AD5C23',
            600: '#D6722B',
            500: '#FF8833',
            400: '#FF9B54',
            300: '#FFAE75',
            200: '#FFC195',
            100: '#FFD4B5',
            50: '#FFE7D6',
        },
        pink: {
            900: '#5C254A',
            800: '#85356A',
            700: '#AD458A',
            600: '#D656AB',
            500: '#FF66CC',
            400: '#FF7FD4',
            300: '#FF97DC',
            200: '#FFAFE4',
            100: '#FFC8ED',
            50: '#FFE0F5',
        },
        purple: {
            900: '#27154C',
            800: '#381E6E',
            700: '#49278E',
            600: '#5B31B0',
            500: '#6C3AD2',
            400: '#845AD9',
            300: '#9B79E0',
            200: '#B298E8',
            100: '#CAB8EF',
            50: '#E2D8F6',
        },
        red: {
            50: '#FFCCCC',
            100: '#FFA3A3',
            200: '#FF7A7A',
            300: '#FF5252',
            400: '#FF2929',
            500: '#FF0000',
            600: '#D60000',
            700: '#AD0000',
            800: '#850000',
            900: '#5C0000',
        },
        teal: {
            900: '#004A37',
            800: '#006A50',
            700: '#008A68',
            600: '#00AB80',
            500: '#00CC99',
            400: '#29D4A9',
            300: '#52DCBA',
            200: '#7AE4CA',
            100: '#A3EDDA',
            50: '#CCF5EB',
        },
        yellow: {
            900: '#5C4A00',
            800: '#856A00',
            700: '#AD8A00',
            600: '#D6AB00',
            500: '#FFCC00',
            400: '#FFD429',
            300: '#FFDC52',
            200: '#FFE47A',
            100: '#FFEDA3',
            50: '#FFF5CC',
        },
    },
};
export default defaultTheme;
