// @citric/core@6.4.0/dist/Select.js
import { extendStyle, overwriteStyle } from './component-style.js';
import { getFieldBaseStyle } from './field-base.js';
import { css, styled } from './styled.js';
import { defaultConfig } from './utils/config.js';
import { getColor } from './utils/theme.js';
const Select = styled.select.withConfig(defaultConfig) `
  appearance: none;

  ${extendStyle('Select')};

  ${({ theme, colorScheme }) => getFieldBaseStyle(theme, colorScheme)};

  ${({ theme, multiple, colorScheme }) => multiple
    ? css `
          height: auto;
          overflow-y: scroll;
          &::-webkit-scrollbar {
            width: 0.125rem;
          }
          &::-webkit-scrollbar-track {
            background: ${getColor(theme, 'light.300')};
          }
          &::-webkit-scrollbar-thumb {
            background: ${getColor(theme, colorScheme || 'primary')};
          }
        `
    : css `
          background: url('data:image/svg+xml;utf8,${getIcon({ theme })}');
          background-repeat: no-repeat;
          background-position: right 0.5rem top 50%;
          background-size: 0.747rem;
          background-color: ${getColor(theme, 'light.300')};
        `}

  ${overwriteStyle('Select')};
`;
function getIcon({ theme }) {
    const color = getColor(theme, 'inverse.300');
    const fill = color?.replace('#', '%23');
    return `<svg xmlns="http://www.w3.org/2000/svg" fill="${fill}" width="16" height="16" viewBox="0 0 512 512"><path d="M505.172 182.112C512.966 174.285 512.939 161.622 505.112 153.828C497.285 146.034 484.622 146.061 476.828 153.888L255.5 376.157L35.1723 154.888C27.3785 147.061 14.7152 147.034 6.88807 154.828C-0.939081 162.622 -0.966059 175.285 6.82776 183.112L240.828 418.112C244.862 422.164 250.202 424.125 255.5 423.994C260.798 424.125 266.138 422.164 270.172 418.112L505.172 182.112Z"/></svg>`;
}
const Option = styled.option.withConfig(defaultConfig) `
  ${extendStyle('Option')};
  ${overwriteStyle('Option')};
`;
export { Option, Select };
