// @stack-spot/portal-layout@2.20.1/dist/components/Contact/show-contact-modal.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, Button, Flex, IconBox, Text } from '@citric/core';
import { ExternalLink } from '@citric/icons';
import { Card } from '@citric/ui';
import { useAnchorTag } from '@stack-spot/portal-components/anchor';
import { useTranslate } from '@stack-spot/portal-translate';
import { overlay } from '../../LayoutOverlayManager.js';
const CardItem = ({ title, description, hasMargin = false, buttonText, url, icon, onClick }) => {
    const Link = useAnchorTag();
    return _jsx(Card, { sx: { p: '7', bg: 'light.500', w: '100%' }, children: _jsxs(Flex, { flexDirection: "column", children: [_jsx(Text, { appearance: "h5", mb: "3", children: title }), _jsx(Text, { appearance: "microtext1", colorScheme: "light.700", mb: "8", children: description }), _jsx(Box, { mt: hasMargin ? 5 : 0, children: _jsx(Link, { onClick: onClick, href: url, target: "_blank", children: _jsxs(Button, { size: "md", colorScheme: "inverse", children: [buttonText, _jsx(IconBox, { sx: { ml: 2 }, colorIcon: "inverse.contrastText", children: icon })] }) }) })] }) });
};
const ContactUsComponent = ({ serviceNowUrl }) => {
    const t = useTranslate(dictionary);
    return _jsx(Flex, { sx: { gap: '8px' }, w: 12, mb: 10, flexWrap: "nowrap", children: _jsx(CardItem, { title: t.support, description: t.supportDescription, buttonText: t.supportButton, url: serviceNowUrl, icon: _jsx(ExternalLink, {}) }) });
};
export const showContactUsModal = ({ title, serviceNowUrl }) => {
    overlay.showModal({
        title,
        children: _jsx(ContactUsComponent, { serviceNowUrl: serviceNowUrl }),
        size: 'small',
    });
};
const dictionary = {
    en: {
        contactUs: 'Contact us',
        support: 'Support and feedback',
        supportDescription: 'Need help or have a feedback? Connect with our Support team',
        supportButton: 'Get in touch',
    },
    pt: {
        contactUs: 'Fale conosco',
        support: 'Suporte e feedback',
        supportDescription: 'Precisa de ajuda ou tem algum feedback? Fale com o nosso time de Suporte',
        supportButton: 'Entre em contato',
    },
};
