// @stack-spot/portal-layout@2.20.1/dist/components/PrivacyPolicyMessage/hooks.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useTranslate } from '@stack-spot/portal-translate';
import { useEffect, useState } from 'react';
import { overlay } from '../../LayoutOverlayManager.js';
import { PrivacyPolicyMessage } from './index.js';
export function usePrivacyPolicyEffect({ policy = 'okCookies', policyValue = 'true', domain = '.stackspot.com', privacyPoliceUrl, message, policyText, confirmText, }) {
    const t = useTranslate(dictionary);
    const [visible, setVisible] = useState(!document.cookie.match(new RegExp(`${policy}=([^;]+)`))?.[1]);
    async function checkPolicy() {
        if (visible) {
            await overlay.showBottomDialog({
                confirm: confirmText || t.ok,
                children: _jsx(PrivacyPolicyMessage, { privacyPoliceUrl: privacyPoliceUrl, message: message, policyText: policyText }),
            });
            document.cookie = `${policy}=${policyValue};path=/${location.host.endsWith(domain) ? `;domain=${domain}` : ''}`;
            setVisible(false);
        }
        else
            overlay.closeBottomDialog();
    }
    useEffect(() => {
        checkPolicy();
    }, [visible]);
}
const dictionary = {
    en: {
        ok: 'Ok',
    },
    pt: {
        ok: 'Ok',
    },
};
