// @stack-spot/portal-components@2.11.0/dist/components/Stepper/headers.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Box, Flex, Text } from '@citric/core';
import { theme } from '@stack-spot/portal-theme';
import { useContext } from 'react';
import { styled } from 'styled-components';
import { StepperContext } from './Stepper.js';
const StyledStepHeader = styled(Box) `
  flex: 1;
  border-top: 0.25rem solid ${theme.color.light[600]};
  padding-top: 1rem;

  &.active {
    border-color: ${theme.color.primary[500]};
  }

  &.finished {
    flex: unset;
    width: 4.5rem;
    border-color: ${theme.color.success[500]};
  }
`;
const StyledStepHeaderText = styled(Text) `
  color: ${theme.color.light.contrastText};
  width: 100%;
  height: 1.5rem;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;
/**
 * Renders a component that represents the visual part of a step within the stepper component.
 * The initial color of the step is primary.500 and when completed it is changed to the success.500 color.
 *
 * The component stepper wraps this component.
 */
export const StepperHeaders = ({ headers }) => {
    const { activeIndex } = useContext(StepperContext);
    return (_jsx(Flex, { sx: { gap: '1rem' }, children: headers.map((header, index) => (_jsx(StyledStepHeader, { className: index < activeIndex ? 'finished' : (index === activeIndex ? 'active' : ''), children: _jsx(StyledStepHeaderText, { children: header.title }) }, header.title))) }));
};
