// @stack-spot/portal-components@2.11.0/dist/components/error/UnderMaintenance.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useTranslate } from '@stack-spot/portal-translate';
import { ErrorFeedback } from './ErrorFeedback.js';
export const UnderMaintenance = ({ refresh }) => {
    const t = useTranslate(dictionary);
    return _jsx(ErrorFeedback, { title: t.underMaintenance, description: t.underMaintenanceDescription, help: t.underMaintenanceHelp, action: refresh ? {
            label: t.refresh,
            onClick: refresh,
        } : undefined });
};
const dictionary = {
    en: {
        underMaintenance: 'Under maintenance',
        underMaintenanceDescription: 'The site is currently undergoing maintenance.',
        underMaintenanceHelp: 'Please, try again in a few minutes.',
        refresh: 'Refresh page',
    },
    pt: {
        underMaintenance: 'Em manutenção',
        underMaintenanceDescription: 'O site está em manutenção.',
        underMaintenanceHelp: 'Por favor, tente novamente em alguns minutos.',
        refresh: 'Atualizar página',
    },
};
