// @stack-spot/portal-components@2.11.0/src/components/TimelineSection.tsx
import { Text } from '@citric/core'
import { Month, useMonthDictionary, useTranslate } from '@stack-spot/portal-translate'
import { styled } from 'styled-components'

interface Props {
  /**
   * The month. 0 for January, 11 for December.
   */
  month: Month,
  /**
   * The day of the month (1 - 31).
   */
  day: number,
  /**
   * The section's content.
   */
  children: React.ReactNode,
  style?: React.CSSProperties,
  className?: string,
}

const Styled = styled.div`
  display: flex;
  flex-direction: row;
  align-items: start;

  .timeline-section-header {
    width: 80px;
  }

  .timeline-section-content {
    min-width: 0;
    flex: 1;
  }
`

/**
 * Renders a section in a timeline where the left side is a header with the date and the right side is the content (children).
 * @param props React props: {@link Props}. 
 */
export const TimelineSection = ({ children, day, month, className, style }: Props) => {
  const months = useMonthDictionary()
  const t = useTranslate(locale)

  const ariaLabel = t.dayMonth.replace('%d', `${day}`).replace('%m', months[month])
  return (
    <Styled style={style} className={className}>
      <div className="timeline-section-header">
        <Text appearance="h2" aria-label={ariaLabel}>{day >= 10 ? day : `0${day}`}</Text>
        <Text appearance="subtitle4" aria-hidden="true">{months[month].toUpperCase().substring(0, 3)}</Text>
      </div>
      <div className="timeline-section-content">
        {children}
      </div>
    </Styled>
  )
}

const locale = {
  pt: {
    dayMonth: '%d de %m',
  },
  en: {
    dayMonth: '%m %d',
  },
}
