// @stack-spot/portal-network@0.66.0/src/api/account.ts
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://account-account-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://account-account-api.dev.stackspot.com"
};
export type UserScmInfoResponse = {
    /** Secrets name */
    secretName: string;
    /** Provider type */
    provider: "GITHUB" | "GITLAB" | "BITBUCKET" | "AZURE" | "GITLAB_ONPRM";
};
export type ValidationDetails = {
    code: string;
    field?: string;
    details?: string;
    values?: string[];
};
export type ErrorResponse = {
    code: string;
    status: number;
    details: string;
    validationDetails?: ValidationDetails[];
};
export type UserScmInfoRequest = {
    /** Repository Username */
    user: string;
    /** Repository personal token */
    token: string;
};
export type ScopeResponse = {
    /** Scope id */
    id: string;
    /** Scope name */
    name: string;
    /** Scope description */
    description?: string;
    /** Scope claim mapper */
    claimMapper?: {
        [key: string]: string;
    };
    /** When scope was created */
    createdAt: string;
    /** Last scope update */
    updatedAt?: string;
};
export type IamUpdateScopeRequest = {
    name: string;
    description?: string;
    claim_mapper: {
        [key: string]: string;
    };
};
export type AccountScmInfoResponse = {
    /** Secrets name */
    secretName?: string;
    /** SCM Workflow repository URL */
    workflowUrl?: string;
    /** SCM Type */
    "type"?: "PAT" | "GITHUBAPP";
    /** SCM repository provider */
    provider: "GITHUB" | "GITLAB" | "BITBUCKET" | "AZURE" | "GITLAB_ONPRM";
    /** Flag used to indicate if the SCM repository account credential configuration is mandated */
    mandate: boolean;
    /** Flag used to indicate if the user credentials to access the account SCM repository already has configured by User */
    hasUserConfiguration?: boolean;
    /** Connection URL for on-premises provider */
    urlProvider?: string;
};
export type Value = object;
export type AccountScmInfoUpdateRequest = {
    /** SCM Type */
    "type"?: "PAT" | "GITHUBAPP";
    /** SCM repository provider */
    provider?: "GITHUB" | "GITLAB" | "BITBUCKET" | "AZURE" | "GITLAB_ONPRM";
    value?: Value;
    /** SCM Workflow repository URL */
    workflowUrl?: string;
    /** Flag used when SCM repository account credential configuration is mandated */
    mandate?: boolean;
    urlProvider?: string;
};
export type AccountScmInfoSaveRequest = {
    /** SCM Type */
    "type": "PAT" | "GITHUBAPP";
    /** SCM repository provider */
    provider: "GITHUB" | "GITLAB" | "BITBUCKET" | "AZURE" | "GITLAB_ONPRM";
    value: Value;
    /** Connection URL for on-premises provider */
    urlProvider?: string;
    /** SCM Workflow repository URL */
    workflowUrl: string;
    /** Flag used when SCM repository account credential configuration is mandated */
    mandate: boolean;
};
export type UpdateAccountRoleRequest = {
    /** Account Role new name. */
    name?: string;
    /** Account Role new description. */
    description?: string;
    /** List of new permissions to add to the role. */
    permissionIds?: string[];
};
export type IamActionResponse = {
    id: string;
    client_id?: string;
    name: string;
    description: string;
    created_at: string;
    active?: boolean;
};
export type IamResourceType = {
    id: string;
    name: string;
    description: string;
    created_at: string;
    global_resource_id: string;
    resource_type_id: string;
    actions?: IamActionResponse[];
};
export type IamCreateStatementRequest = {
    actions: string[];
    resource: string;
};
export type UpdateResourceActionsRequest = {
    actions: string[];
};
export type AccountMemberStackspotIamDto = {
    id: string;
    username: string;
    email?: string;
    firstName?: string;
    lastName?: string;
    ddd?: string;
    ddi?: string;
    phone?: string;
    company?: string;
    isActive: boolean;
    isEmailVerified: boolean;
    properties?: {
        [key: string]: object;
    };
};
export type UpdateUserRequest = {
    /** User username */
    username?: string;
    /** User email */
    email?: string;
    /** User first name */
    firstName?: string;
    /** User last name */
    lastName?: string;
    /** User ddi */
    ddi?: string;
    /** User ddd */
    ddd?: string;
    /** User phone */
    phone?: string;
    /** Is user active? */
    isActive?: boolean;
    /** User Company */
    company?: string;
    /** User new password */
    password?: string;
};
export type UpdateUserResponse = {
    username?: string;
    email?: string;
    firstName?: string;
    lastName?: string;
    ddi?: string;
    ddd?: string;
    phone?: string;
    isActive?: boolean;
    company?: string;
    password?: string;
};
export type UpdatePasswordRequest = {
    /** New password */
    password: string;
};
export type InvitationUserResponse = {
    email: string;
};
export type UserInvitationResponse = {
    id: string;
    accountId: string;
    sender: InvitationUserResponse;
    invitedUser: InvitationUserResponse;
    createdAt: string;
    resentAt: string;
    expirationDate: string;
    roleName: string;
    assignAccount: boolean;
    magicLink: string;
    status: "ACCEPTED" | "PENDING" | "CANCELLED";
};
export type UpdateFeatureFlagRequest = {
    /** The feature flag  name */
    name: string;
    /** The feature flag slug */
    slug: string;
    /** The feature flag description */
    description: string;
};
export type UpdateSalesRepresentativesRequest = {
    /** List of emails from sales representatives responsible for the account. */
    salesRepresentatives: string[];
};
export type AccountStatusRequest = {
    /** Account enabled */
    enabled: boolean;
};
export type AttributeImporter = {
    /** External user attribute */
    externalUserAttribute?: string;
    /** Force attribute import */
    force?: string;
    /** Unique identifier for the attribute importer */
    id?: string;
    /** Mode of the attribute importer */
    importerMode?: string;
    /** Type of the attribute importer */
    importerType?: string;
    /** Mapping for the attribute importer */
    mapTo?: string;
    /** Name of the attribute importer */
    name?: string;
    /** Creation timestamp of the attribute importer */
    createdAt?: string;
};
export type IdMapper = {
    /** External IDP ID */
    externalIdpId?: string;
    /** Internal user ID */
    internalUserId?: string;
};
export type Config = {
    /** OpenId Authorization Endpoint */
    authorizationEndpoint?: string;
    /** OpenId End Session Endpoint */
    endSessionEndpoint?: string;
    /** OpenId Issuer */
    issuer?: string;
    /** OpenId JWKS Uri */
    jwksUri?: string;
    /** OpenId Revocation Endpoint */
    revocationEndpoint?: string;
    /** OpenId Token Endpoint */
    tokenEndpoint?: string;
    /** SSO Userinfo Endpoint */
    userinfoEndpoint?: string;
    /** OpenId Client ID */
    clientId?: string;
    /** OpenId Client Secret */
    clientSecret?: string;
};
export type Endpoint = {
    binding?: string;
    location?: string;
};
export type AccountSsoDetailedResponse = {
    /** Unique identifier for the SSO configuration */
    id?: string;
    /** Alias for the SSO configuration */
    alias?: string;
    /** SSO configuration name */
    name?: string;
    /** SSO callback url */
    callbackUrl?: string;
    /** List of attribute importers */
    attributeImporters: AttributeImporter[];
    /** Creation timestamp of the SSO configuration */
    createdAt?: string;
    /** Update timestamp of the SSO configuration */
    updatedAt?: string;
    /** Discovery endpoint for the SSO configuration */
    discoveryEndpoint?: string;
    /** Indicates if the SSO configuration is enabled */
    isEnabled?: boolean;
    idMapper?: IdMapper;
    /** Type of the SSO configuration */
    "type": "SAML" | "OPENID";
    config?: Config;
    /** SSO Endpoints */
    ssoEndpoints?: Endpoint[];
    /** SLO Endpoints */
    sloEndpoints?: Endpoint[];
    /** Indicates if the SSO is fully configured */
    isConfigured: boolean;
};
export type AccountSsoUpdateRequest = {
    /** SSO Alias */
    alias: string;
    /** SSO Name */
    name: string;
    /** SSO Callback URL */
    callbackUrl?: string;
    config?: Config;
};
export type AccountSsoPartialUpdateRequest = {
    /** SSO Status */
    isEnabled: boolean;
};
export type AccountSsoGroupMappingResponse = {
    /** SSO Created At */
    createdAt: string;
    /** SSO Group Mapping externalUserAttribute */
    attributeName?: string;
    /** SSO Group Mapping externalUserValue */
    attributeValue?: string;
    /** SSO Group Mapping force */
    force?: boolean;
    /** SSO Group Mapping id */
    id: string;
    /** SSO Group Mapping importerMode */
    importerMode?: string;
    /** SSO Group Mapping importerType */
    importerType?: "DEFAULT_ROLE" | "DEFAULT_GROUP" | "ATTR_TO_GROUP" | "ATTR_TO_ROLE" | "DYNAMIC_GROUP";
    /** SSO Group Mapping mapTo */
    mapTo?: string;
    /** SSO Group Mapping name */
    name: string;
};
export type AccountGroupMappingUpdateRequest = {
    /** Group Mapping externalUserAttribute */
    attributeName?: string;
    /** Group Mapping externalUserValue */
    attributeValue?: string;
    /** Group Mapping force */
    force?: boolean;
    /** Group Mapping importerMode */
    importerMode?: string;
    /** Group Mapping importerType */
    importerType: "DEFAULT_ROLE" | "DEFAULT_GROUP" | "ATTR_TO_GROUP" | "ATTR_TO_ROLE" | "DYNAMIC_GROUP";
    /** Group Mapping mapTo */
    mapTo: string;
    /** Group Mapping name */
    name: string;
};
export type ServiceCredentialCreateRequestV2 = {
    /** Service credential name */
    name: string;
    /** Service credential expiration days */
    description?: string;
    /** Service credential expiration days */
    expirationDays?: number;
    /** Service credential statements */
    scopes?: string[];
};
export type ServiceCredentialCreateResponseV2 = {
    /** Client ID service credential */
    clientId: string;
    /** Client Secret service credential */
    clientSecret?: string;
    /** Service credential scopes */
    scopes: string[];
    /** Service credential expiration date */
    expirationDate?: string;
    /** Account tenant */
    tenant: string;
};
export type ServiceCredentialResponse = {
    /** Service credential id */
    id: string;
    /** Service credential name */
    name: string;
    /** Service credential creator id */
    createdBy: string;
    /** Service credential creator username */
    creatorUsername: string;
    /** Service credential creator email */
    creatorEmail: string;
    /** Service credential created date */
    createdAt: string;
    /** Service credential expire date */
    expirationDate?: string;
    /** Service credential user email revoked */
    revokedBy?: string;
    /** Service credential revoked date */
    revokedAt?: string;
    /** Service credential status */
    status: "ACTIVE" | "DISABLED" | "REVOKED" | "EXPIRED";
};
export type StatementRequest = {
    /** Resource id */
    resource: string;
    /** Action id list */
    actions: string[];
};
export type ServiceCredentialCreateRequest = {
    /** Service credential name */
    name: string;
    /** Service credential expiration days */
    expirationDays?: number;
    /** Service credential permission ids */
    permissions?: string[];
    /** Service credential groups ids */
    teams?: string[];
    /** Service credential statements */
    statements?: StatementRequest[];
};
export type ServiceCredentialCreateResponse = {
    /** Client ID service credential */
    clientId: string;
    /** Client Secret service credential */
    clientSecret: string;
    /** Account tenant */
    tenant: string;
};
export type ServiceCredentialInfoRequest = {
    /** ID of the Service Credential */
    id: string;
    /** Name of the Service Credential */
    name: string;
};
export type ServiceCredentialExpirationNotifyRequest = {
    /** ID of the Service Credential creator */
    ownerId: string;
    /** Days until Service Credential expiration */
    daysToExpire: number;
    /** Tenant that represents the account related to the Service Credential */
    accountTenant: string;
    /** Information related to the Service Credential */
    credentials: ServiceCredentialInfoRequest[];
};
export type PagingRequest = {
    size?: number;
    page: number;
    sort?: string;
    direction: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: string;
};
export type IamCreateScopeRequest = {
    name: string;
    description?: string;
    claim_mapper?: {
        [key: string]: string;
    };
};
export type PermissionActionResponse = {
    /** The action id from StackSpot IAM */
    id: string;
    /** The client ID from StackSpot IAM */
    clientId?: string;
    /** Action's name */
    name: string;
    /** Action's description */
    description?: string;
    /** Action's creation timestamp */
    createdAt: string;
};
export type PermissionTypeResponse = {
    /** Resource type ID from StackSpot IAM */
    id: string;
    /** Resource type client ID from StackSpot IAM */
    clientId?: string;
    /** Resource type name */
    name: string;
    /** Resource type description */
    description?: string;
    /** Resource type creation timestamp */
    createdAt: string;
};
export type PermissionResourceResponse = {
    /** Resource ID from StackSpot IAM */
    id: string;
    /** Resource's client ID from StackSpot IAM */
    clientId?: string;
    /** Resource's name */
    name: string;
    /** Resource's description */
    description?: string;
    /** Resource's creation timestamp */
    createdAt: string;
    "type"?: PermissionTypeResponse;
};
export type PermissionStatementResponse = {
    actions: PermissionActionResponse;
    resource: PermissionResourceResponse;
};
export type PermissionResponse = {
    /** Permission ID from StackSpot IAM */
    id: string;
    /** Permission's name */
    name: string;
    /** Permission's description */
    description?: string;
    /** Permission's creation timestamp */
    createdAt: string;
    /** The statements defined in the permission */
    statements: PermissionStatementResponse[];
};
export type RoleResponse = {
    /** Role id from Keycloak */
    id: string;
    /** Role name */
    name: string;
    /** Role description */
    description: string;
    /** Role description */
    permissions: PermissionResponse[];
    /** Total members in role */
    totalMembers?: number;
};
export type CreateAccountRoleRequest = {
    /** Account Role name */
    name: string;
    /** Account Role description */
    description?: string;
};
export type IdResponse = {
    /** Id response. */
    id: string;
};
export type RoleMemberResponse = {
    /** Member ID */
    id: string;
    /** Member username */
    username: string;
    /** Member company email */
    email: string;
    /** Member first name */
    firstName?: string;
    /** Member last name */
    lastName?: string;
    /** Member creation timestamp */
    createdAt: string;
};
export type AddRoleToMemberRequest = {
    /** Account member IDs list to add role */
    ids: string[];
};
export type AddUsersToRoleResponse = {
    /** Status Message */
    msg: string;
    /** Permission resource */
    errors: string[];
};
export type RoleGroupResponse = {
    id: string;
    name: string;
    description?: string;
    imageURL?: string;
    createdAt: string;
};
export type RoleGroupIdsRequest = {
    ids: string[];
};
export type ResourceTypeDto = {
    /** Id of resource type. */
    id: string;
    /** Name of resource type. */
    name: string;
    /** Slug of resource type. */
    slug: string;
    /** Description of resource type. */
    description: string;
    /** Creation date of resource type. */
    createdAt: string;
};
export type ResourceResponse = {
    /** Id of resource. */
    id: string;
    /** Name of resource. */
    name: string;
    /** Description of resource. */
    description?: string;
    "type"?: ResourceTypeDto;
    /** Name of resource. */
    slug: string;
};
export type CreateResourceRequest = {
    /** Account resource name */
    name: string;
    /** Account resource slug */
    slug: string;
    /** Account resource type id */
    typeId: string;
    /** Account resource description */
    description?: string;
};
export type AccountMemberResponse = {
    /** Account member id */
    id: string;
    /** Account member username */
    username: string;
    /** Account member email */
    email?: string;
    /** Account member first name */
    firstName: string;
    /** Account member last name */
    lastName: string;
    /** Account member status */
    isActive: boolean;
    /** Account member avatar URL */
    avatarUrl?: string;
    /** Account member  full name */
    name: string;
};
export type CreateUserRequest = {
    /** User username */
    username: string;
    /** User email */
    email: string;
    /** User first name */
    firstName: string;
    /** User last name */
    lastName: string;
    /** User ddi */
    ddi?: string;
    /** User ddd */
    ddd?: string;
    /** User phone */
    phone?: string;
    /** Is user active? */
    isActive: boolean;
    /** User first password */
    password?: string;
    /** User Company */
    company?: string;
};
export type CreateUserResponse = {
    username: string;
    email: string;
    company?: string;
    firstName: string;
    lastName: string;
    ddi: string;
    ddd: string;
    phone: string;
    isActive: boolean;
    isEmailVerified?: boolean;
};
export type ReadMemberRoleResponse = {
    /** Role id from IAM */
    id: string;
    /** Role name */
    name: string;
    /** Role description */
    description: string;
    /** Total members in role */
    totalMembers: number;
    /** Role description */
    permissions: PermissionResponse[];
};
export type MemberIdsCommonRequest = {
    ids: string[];
};
export type GroupReadResponse = {
    /** Group id */
    id: string;
    /** Group name */
    name: string;
    /** Group description */
    description?: string;
    /** Group users */
    totalUsers?: number;
    /** Group image URL */
    imageUrl?: string;
    /** Creation time of group */
    createdAt?: string;
};
export type FavoriteResourcesResponse = {
    /** The type of the resource flag name */
    resourceType: string;
    /** List of resource ids */
    resources: string[];
};
export type CreateFavoriteResourceRequest = {
    /** The type of the resource flag name */
    resourceType: string;
    /** The resource identifier */
    resourceId: string;
};
export type MemberEmailRequest = {
    /** Member email */
    email: string;
};
export type InviteNewMemberRequest = {
    /** Member email */
    email: string;
    /** Member name */
    name: string;
    /** Member surname */
    surname: string;
    /** Member group ID */
    groupId?: string;
};
export type InviteMembersResponse = {
    /** Status Message */
    msg: string;
    /** Members with error */
    errors: string[];
};
export type CreateUserInvitationRequest = {
    /** User email */
    email: string;
    /** User role in account */
    role: string;
    /** Indicates whether account will be shared with user or not */
    assignAccount: boolean;
    /** Language to use in invitation e-mail */
    lang: string;
};
export type NewGroupRequest = {
    /** Group name */
    name: string;
    /** Group description */
    description: string;
    /** External Group name */
    externalName?: string;
    /** Group image base64 */
    image?: string;
};
export type IamIdResponse = {
    /** Id response. */
    id: string;
};
export type GroupRoleIdsRequest = {
    ids: string[];
};
export type AddResourcesToGroupRequest = {
    /** List of resource ids */
    ids: string[];
};
export type ReadGroupMembersResponse = {
    /** Account member id */
    id: string;
    /** Account member username */
    username: string;
    /** Account member first name */
    firstName?: string;
    /** Account member last name */
    lastName?: string;
    /** Account member email */
    companyEmail: string;
    /** Account member creation time */
    createdAt: string;
    /** Account member appended time */
    appendedAt: string;
};
export type GroupMemberIdsRequest = {
    ids: string[];
};
export type FeatureFlagsResponse = {
    /** The feature flag id */
    id: number;
    /** The feature flag name */
    name: string;
    /** The feature flag slug */
    slug: string;
    /** If the feature flag is enabled or not in the current account */
    enabled: boolean;
};
export type CreateFeatureFlagRequest = {
    /** The feature flag  name */
    name: string;
    /** The feature flag slug */
    slug: string;
    /** The feature flag description */
    description: string;
};
export type WelcomeEmailRequest = {
    /** Firstname of the contact requester */
    firstName: string;
    /** Firstname of the contact requester */
    lastName: string;
    /** Email of the contact requester */
    email: string;
};
export type ContactEmailRequest = {
    /** Name of the contact requester */
    name: string;
    /** Email of the contact requester */
    email: string;
    /** Subject of the contact */
    subject: string;
    /** Message of the contact */
    message: string;
    /** Telephone of the contact requester */
    telephone?: string;
};
export type SwitchAccountRequest = {
    /** Account slug to switch to */
    accountId: string;
};
export type ExchangeAccountResponse = {
    /** Personal Access Token generated by User Personal Client on user behalf */
    access_token: string;
    /** Refresh Token */
    refresh_token: string;
    /** The token type */
    token_type: string;
    /** The number of seconds before the access_token expires */
    expires_in: number;
};
export type PersonalClientCredentialsResponse = {
    /** The Client ID used for authentication */
    clientId: string;
    /** The Client Secret used for authentication */
    clientSecret: string;
    /** The account tenant */
    tenant: string;
};
export type GenerateTokenFromPersonalAccountRequest = {
    /** Audience */
    audience?: string;
};
export type PersonalAccessTokenResponse = {
    /** Personal Access Token generated by User Personal Client on user behalf */
    accessToken: string;
};
export type DeviceCodeResponse = {
    /** A long string used to verify the session between the client and the authorization server. The client uses this parameter to request the access token from the authorization server */
    deviceCode: string;
    /** A short string shown to the user that's used to identify the session on a secondary device */
    userCode: string;
    /** The URI the user should go to with the user_code in order to sign in */
    loginUrl: string;
    /** The account tenant */
    realm: string;
    /** The number of seconds before the device_code and user_code expire */
    expiration: number;
    /** The number of seconds the client should wait between polling requests */
    interval: number;
    /** OpenID Well Known url for current account */
    openIdWellKnownUrl: string;
};
export type AccountResponse = {
    /** Account id */
    accountApiId: string;
    /** Account slug */
    slug: string;
    /** Account name */
    name: string;
    /** Account logo image */
    image?: string;
    /** Account creation time */
    createdAt: string;
    /** Account creator */
    createdBy?: string;
    /** Account admin email */
    adminEmail: string;
    /** Account admin name */
    adminName: string;
    /** Account Email domains list */
    domains: string[];
    /** Account current status */
    isActive: boolean;
    /** Account sales representatives email list */
    salesRepresentatives: string[];
};
export type PageResponseAccountResponse = {
    /** Current page content */
    items?: AccountResponse[];
    /** Total elements found */
    totalElements: number;
    /** Current page number */
    page: number;
    /** Length of current page items */
    size: number;
};
export type CreateAccountOrderRequest = {
    /** The account's name. Example: Zup IT Innovation */
    name: string;
    /** The slug to use in URL as unique identifying part representing the account (https://developer.mozilla.org/en-US/docs/Glossary/Slug). Example: zup */
    slug: string;
    /** The account tenant should be formatted similarly to the slug and should identify the account uniquely. Only to use internal use. Example: zup-it */
    tenant: string;
    /** The existing accountID if created beforehand */
    accountId?: string;
    /** Account holder's first name */
    adminFirstName: string;
    /** Account holder's last name */
    adminLastName: string;
    /** Account holder's email */
    adminEmail: string;
    /** Email domains list that account users will use to login into StackSpot. Example: zup.com.br */
    domains: string[];
    /** Email list of the sales representatives responsible for this request. */
    salesRepresentatives: string[];
};
export type PageResponseAccountSsoGroupMappingResponse = {
    /** Current page content */
    items?: AccountSsoGroupMappingResponse[];
    /** Total elements found */
    totalElements: number;
    /** Current page number */
    page: number;
    /** Length of current page items */
    size: number;
};
export type AccountGroupMappingCreateRequest = {
    /** Group Mapping externalUserAttribute */
    attributeName?: string;
    /** Group Mapping externalUserValue */
    attributeValue?: string;
    /** Group Mapping force */
    force?: boolean;
    /** Group Mapping importerMode */
    importerMode?: string;
    /** Group Mapping importerType */
    importerType: "DEFAULT_ROLE" | "DEFAULT_GROUP" | "ATTR_TO_GROUP" | "ATTR_TO_ROLE" | "DYNAMIC_GROUP";
    /** Group Mapping mapTo */
    mapTo?: string;
    /** Group Mapping name */
    name: string;
};
export type Certificate = {
    data?: string;
};
export type SignKey = {
    name?: string;
    certificates?: Certificate[];
};
export type SsoConfigRequest = {
    /** SAML Entity Id */
    entityId?: string;
    signKey?: SignKey;
    /** OpenId Authorization Endpoint */
    authorizationEndpoint?: string;
    /** OpenId End Session Endpoint */
    endSessionEndpoint?: string;
    /** OpenId Issuer */
    issuer?: string;
    /** OpenId JWKS Uri */
    jwksUri?: string;
    /** OpenId Registration Endpoint */
    registrationEndpoint?: string;
    /** OpenId Revocation Endpoint */
    revocationEndpoint?: string;
    /** OpenId Token Endpoint */
    tokenEndpoint?: string;
    /** OpenId Userinfo Endpoint */
    userinfoEndpoint?: string;
    /** OpenId Client ID */
    clientId?: string;
    /** OpenId Client Secret */
    clientSecret?: string;
};
export type AccountSsoCreateRequest = {
    /** Alias from SSO configuration */
    alias: string;
    /** Name from SSO configuration */
    name: string;
    /** Callback URL from SSO configuration */
    callbackUrl: string;
    config: SsoConfigRequest;
    /** Activation configuration */
    isEnabled: boolean;
    /** SSO SAML Single Sign On endpoints */
    ssoEndpoints?: Endpoint[];
    /** SSO SAML Single Logout On endpoints */
    sloEndpoints?: Endpoint[];
    /** SSO SAML Name id Formats */
    nameIdFormats?: string[];
};
export type SsoIdResponse = {
    /** Id response. */
    id: string;
};
export type AccountSsoParserResponse = {
    config?: Config;
    /** SAML Want Signed Requests */
    wantSignedRequests?: boolean;
    /** SAML SSO Endpoints */
    ssoEndpoints: Endpoint[];
    /** SAML SLO Endpoints */
    sloEndpoints: Endpoint[];
    /** SAML Name Id Formats */
    nameIdFormats: string[];
    /** SSO callback url */
    callbackUrl?: string;
};
export type AccountFreemiumCreateRequest = {
    /** The account's name. Example: Zup IT Innovation */
    name: string;
    /** The slug to use in URL as unique identifying part representing the account (https://developer.mozilla.org/en-US/docs/Glossary/Slug). Example: zup */
    slug: string;
};
export type AccountCollaboratorRequest = {
    /** The API id of account requester */
    accountApiId: string;
};
export type CreateAccountPartnerOrderRequest = {
    /** The account's name. Example: Zup IT Innovation */
    name: string;
    /** Account holder's first name */
    adminFirstName: string;
    /** Account holder's last name */
    adminLastName: string;
    /** Account holder's email */
    adminEmail: string;
    /** Email domains list that account users will use to login into StackSpot. Example: zup.com.br */
    domains: string[];
    /** Account logo image in base64 */
    image?: string;
};
export type IdDefaultResponse = {
    /** Id response. */
    id: string;
};
export type ValidateAccountPartnerDataRequest = {
    /** The account's name. Example: Zup IT Innovation */
    name?: string;
    /** Email domains list that account users will use to login into StackSpot. Example: zup.com.br */
    domains?: string[];
};
export type RotateServiceCredentialSecretRequest = {
    /** Service credential secret */
    secret: string;
};
export type RotateServiceCredentialSecretResponse = {
    secret?: string;
};
export type ServiceCredentialAssociateGroupRequest = {
    /** Service credential groups ids */
    teams: string[];
};
export type ReadPreferencesResponse = {
    /** Dashboard widgets preferences */
    dashboardWidgets?: string[];
    /** Preferred language preferences */
    preferredLanguage?: "PT" | "EN";
    /** Theme preferences */
    theme?: "DARK" | "LIGHT";
};
export type UpdateMemberPreferencesRequest = {
    /** Dashboard widgets preferences Key */
    dashboardWidgets?: string[];
    /** Preferred language preferences Key */
    preferredLanguage?: "PT" | "EN";
    /** Theme preferences Key */
    theme?: "DARK" | "LIGHT";
};
export type DeleteMemberPermissionsRequest = {
    /** Permission Action */
    actionName: string;
    /** Slug from resource type */
    resourceTypeSlug: string;
    /** Slug from resource */
    resourceSlug: string;
};
export type UpdateMemberPermissionsRequest = {
    /** Permission Action */
    actionName: string;
    /** Slug from resource type */
    resourceTypeSlug: string;
    /** Slug from resource */
    resourceSlug: string;
};
export type MemberCredentialResponse = {
    /** Field to represents if credential is active */
    active: boolean;
    /** Member credential type */
    credentialType: "password" | "otp" | "fido";
    /** Member credential encoded data */
    aaguid?: string;
    /** Member credential identifier */
    id: number;
    /** Member credential was created */
    createdAt: string;
    /** Member credential last updated */
    updatedAt: string;
    /** Member id */
    userId: string;
};
export type MemberCredentialsListResponse = {
    /** Describes whether the user has fido configured */
    pendingConfiguration: boolean;
    /** List of User Credentials */
    items: MemberCredentialResponse[];
    /** Size of list */
    size: number;
    /** Total elements */
    totalElements: number;
};
export type GroupReadDetailResponse = {
    /** Group id */
    id: string;
    /** Group name */
    name: string;
    /** Group description */
    description?: string;
    /** Total users */
    totalUsers?: number;
    /** Group image URL */
    imageUrl?: string;
};
export type UpdateGroupRequest = {
    /** Group name */
    name?: string;
    /** Group description */
    description: string;
    /** Group image base64 */
    image?: string;
};
export type DeleteGroupPermissionsRequest = {
    /** Permission Action */
    actionName: string;
    /** Slug from resource type */
    resourceTypeSlug: string;
    /** Slug from resource */
    resourceSlug: string;
};
export type UpdateGroupPermissionsRequest = {
    /** Permission Action */
    actionName: string;
    /** Slug from resource type */
    resourceTypeSlug: string;
    /** Slug from resource */
    resourceSlug: string;
};
export type AddNewEmailDomainRequest = {
    /** Email domains list that account users will use to login into StackSpot. Example: zup.com.br */
    domains: string[];
};
export type AccountSsoAttributesConfigResponse = {
    /** Id from SSO */
    id: string;
    /** SSO Name attribute configuration */
    email?: string;
    /** SSO First Name attribute configuration */
    firstName?: string;
    /** SSO Last Name attribute configuration */
    lastName?: string;
};
export type AccountSsoAddAttributesConfigRequest = {
    /** SSO Name attribute configuration */
    email: string;
    /** SSO First Name attribute configuration */
    firstName: string;
    /** SSO Last Name attribute configuration */
    lastName?: string;
};
export type AccountPartnerAccountDataUpdateRequest = {
    /** The account's name. Example: Zup IT Innovation */
    name?: string;
    /** Account logo image in base64 */
    image?: string;
};
export type AccountPartnerAdminDataUpdateRequest = {
    /** Admin first name */
    firstName: string;
    /** Admin last name */
    lastName: string;
    /** Admin email */
    email: string;
};
export type AccountCollaborationInfoResponse = {
    /** Account ID */
    id: string;
    /** Account Name */
    name: string;
    /** Account Tenant */
    tenant: string;
    /** Account Slug */
    slug: string;
    /** Role type of user in Account */
    roleType: "OWNER" | "MEMBER";
};
export type AccountDataReadResponse = {
    /** Account ID */
    id: string;
    /** Account Name */
    name: string;
    /** Account Tenant */
    tenant: string;
    /** Account Slug */
    slug: string;
    /** Account Admin Name */
    adminName: string;
    /** Account Admin Email */
    adminEmail: string;
    /** Account Type */
    accountType: "ENTERPRISE" | "PARTNER" | "FREEMIUM";
    /** Account logo URL */
    logoBase64?: string;
};
export type ServiceCredentialDetailsResponse = {
    /** Service credential id */
    id: string;
    /** Service credential name */
    name: string;
    /** Service credential user email creator */
    createdBy: string;
    /** Service credential created date */
    createdAt: string;
    /** Service credential expire date */
    expirationDate?: string;
    /** Service credential user email revoked */
    revokedBy?: string;
    /** Service credential revoked date */
    revokedAt?: string;
    /** Service credential status */
    status: "ACTIVE" | "DISABLED" | "REVOKED" | "EXPIRED";
    /** Service credential scopes */
    scopes?: string[];
};
export type EditableActionResponse = {
    /** Id of action. */
    id: string;
    /** Client of action. */
    clientId?: string;
    /** Name of action. */
    name: string;
    /** Description of action. */
    description: string;
    /** Creation time of action. */
    createdAt: string;
    /** Action is active for parent resource. */
    active?: boolean;
};
export type FullResourceResponse = {
    /** Id of resource. */
    id: string;
    /** Name of resource. */
    name: string;
    /** Description of resource. */
    description?: string;
    actions?: EditableActionResponse[];
    globalResourceId: string;
    createdAt: string;
};
export type AccountScmProviderResponse = {
    /** SCM repository provider */
    provider: "GITHUB" | "GITLAB" | "BITBUCKET" | "AZURE" | "GITLAB_ONPRM";
};
export type AccountScmStatusResponse = {
    /** Flag used to indicate if the SCM repository account credential configuration is mandated */
    mandate: boolean;
    /** Flag used to indicate if the user credentials to access the account SCM repository already has configured by User */
    hasUserConfiguration: boolean;
};
export type RolePermissionResponse = {
    status?: boolean;
    /** Permission ID from StackSpot IAM */
    id: string;
    /** Permission's name */
    name: string;
    /** Permission's description */
    description?: string;
    /** Permission's creation timestamp */
    createdAt: string;
    /** The statements defined in the permission */
    statements: PermissionStatementResponse[];
};
export type MembersFromResourceResponse = {
    /** Member ID */
    id: string;
    /** Member username */
    username: string;
    /** Member company email */
    email: string;
    /** Member name */
    name?: string;
    /** Member permission actions */
    actions: string[];
};
export type GroupsFromResourceResponse = {
    /** Group ID */
    id: string;
    /** Group name */
    name: string;
    /** Group description */
    description?: string;
    /** Group slug */
    slug: string;
    /** Group member count */
    memberCount: number;
    /** Group permission actions */
    actions: string[];
};
export type ResourceTypeResponse = {
    id: string;
    name: string;
    slug: string;
    description: string;
    createdAt: string;
};
export type ResourceDto = {
    id: string;
    name: string;
    slug: string;
    description?: string;
    "type"?: ResourceTypeDto;
};
export type AccountAuthenticationConfigResponse = {
    /** OpenID Well Known url for current account */
    openIdWellKnownUrl: string;
};
export type AccountDataResponse = {
    /** Account tenant */
    tenant: string;
    /** Account IDP */
    idp?: string;
    /** Account Freemium available to user */
    freemiumAvailable: boolean;
    /** Account Type */
    accountType: "ENTERPRISE" | "PARTNER" | "FREEMIUM";
    /** Stackspot IAM Migrated */
    isIamMigrated: boolean;
    /** Stackspot IAM Tenant */
    openIdWellKnownUrl?: string;
    /** Account OpenID Well Known url */
    iamTenant?: string;
};
export type AccountSlugResponse = {
    /** Account slug */
    slug: string;
};
export type AccountSsoResponse = {
    /** SSO configuration name */
    name: string;
    /** Unique identifier for the SSO configuration */
    id: string;
    /** Alias for the SSO configuration */
    alias: string;
    /** List of attribute importers */
    attributeImporters: AttributeImporter[];
    /** Creation timestamp of the SSO configuration */
    createdAt?: string;
    /** Discovery endpoint for the SSO configuration */
    discoveryEndpoint?: string;
    /** Indicates if the SSO configuration is enabled */
    isEnabled?: boolean;
    idMapper?: IdMapper;
    /** Type of the SSO configuration */
    "type": "SAML" | "OPENID";
    /** Indicates if the SSO is fully configured */
    isConfigured: boolean;
};
export type PageResponseAccountSsoResponse = {
    /** Current page content */
    items?: AccountSsoResponse[];
    /** Total elements found */
    totalElements: number;
    /** Current page number */
    page: number;
    /** Length of current page items */
    size: number;
};
export type AccountDataAvailableResponse = {
    /** Account data (name and slug) is available to be registered */
    available: boolean;
};
export type AccountPartnerResponse = {
    /** Account id */
    accountApiId: string;
    /** Account slug */
    slug: string;
    /** Account name */
    name: string;
    /** Account logo image */
    image?: string;
    /** Account creation time */
    createdAt: string;
    /** Account current status */
    isActive: boolean;
};
/**
 * List User repository credential's.
 */
export function listScmCredentials(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UserScmInfoResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/users/scm-credentials", {
        ...opts
    }));
}
/**
 * Update User repository credential's.
 */
export function scmCredentialUpdate({ userScmInfoRequest }: {
    userScmInfoRequest: UserScmInfoRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/users/scm-credentials", oazapfts.json({
        ...opts,
        method: "PUT",
        body: userScmInfoRequest
    })));
}
/**
 * Save User repository credential's.
 */
export function scmCredentialSave({ userScmInfoRequest }: {
    userScmInfoRequest: UserScmInfoRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/users/scm-credentials", oazapfts.json({
        ...opts,
        method: "POST",
        body: userScmInfoRequest
    })));
}
/**
 * Delete an user SCM credentials.
 */
export function scmDelete(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/users/scm-credentials", {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get scope by ID
 */
export function getById({ scopeId }: {
    scopeId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ScopeResponse;
    } | {
        status: 400;
        data: ScopeResponse;
    } | {
        status: 404;
        data: ScopeResponse;
    } | {
        status: 500;
        data: ScopeResponse;
    }>(`/v1/scopes/${encodeURIComponent(scopeId)}`, {
        ...opts
    }));
}
/**
 * Update scope information
 */
export function updateScope({ scopeId, iamUpdateScopeRequest }: {
    scopeId: string;
    iamUpdateScopeRequest: IamUpdateScopeRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/scopes/${encodeURIComponent(scopeId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: iamUpdateScopeRequest
    })));
}
/**
 * Delete scope by ID
 */
export function deleteScope({ scopeId }: {
    scopeId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/scopes/${encodeURIComponent(scopeId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * List repository credential's.
 */
export function listScmCredentials1(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountScmInfoResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/scm-credentials", {
        ...opts
    }));
}
/**
 * Update repository credential's.
 */
export function scmCredentialUpdate1({ accountScmInfoUpdateRequest }: {
    accountScmInfoUpdateRequest: AccountScmInfoUpdateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/scm-credentials", oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountScmInfoUpdateRequest
    })));
}
/**
 * Save repository credential's.
 */
export function scmCredentialSave1({ accountScmInfoSaveRequest }: {
    accountScmInfoSaveRequest: AccountScmInfoSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/scm-credentials", oazapfts.json({
        ...opts,
        method: "POST",
        body: accountScmInfoSaveRequest
    })));
}
/**
 * Update a new role for Account
 */
export function updateAccountRole({ roleId, updateAccountRoleRequest }: {
    roleId: string;
    updateAccountRoleRequest: UpdateAccountRoleRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateAccountRoleRequest
    })));
}
/**
 * Delete a role for Account
 */
export function deleteAccountRole({ roleId }: {
    roleId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get all Global Reources and Actions compared to role
 */
export function getResourcesAndActionsWithStatus({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    roleId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IamResourceType[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Update role with new resource and actions
 */
export function updateRoleWithNewActions({ roleId, body }: {
    roleId: string;
    body: IamCreateStatementRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/resources`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Updates actions from a resource
 */
export function updateResourceActions({ roleId, resourceId, updateResourceActionsRequest }: {
    roleId: string;
    resourceId: string;
    updateResourceActionsRequest: UpdateResourceActionsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/resources/${encodeURIComponent(resourceId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateResourceActionsRequest
    })));
}
/**
 * Get user by id
 */
export function getMemberById({ memberId }: {
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountMemberStackspotIamDto;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}`, {
        ...opts
    }));
}
/**
 * Update user on current tenant
 */
export function updateUser({ memberId, updateUserRequest }: {
    memberId: string;
    updateUserRequest: UpdateUserRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UpdateUserResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateUserRequest
    })));
}
/**
 * Update user password on current tenant
 */
export function updateUserPassword({ memberId, updatePasswordRequest }: {
    memberId: string;
    updatePasswordRequest: UpdatePasswordRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/password`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updatePasswordRequest
    })));
}
/**
 * Accept user invitation
 */
export function accept({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UserInvitationResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 409;
        data: ErrorResponse;
    } | {
        status: 424;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/invitations/${encodeURIComponent(id)}/accept`, {
        ...opts,
        method: "PUT"
    }));
}
/**
 * Update a Feature Flag
 */
export function update({ id, updateFeatureFlagRequest }: {
    id: number;
    updateFeatureFlagRequest: UpdateFeatureFlagRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
    }>(`/v1/feature-flags/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateFeatureFlagRequest
    })));
}
/**
 * Sets a new sales representative list to an existing account
 */
export function updateSalesRepresentatives({ accountId, updateSalesRepresentativesRequest }: {
    accountId: string;
    updateSalesRepresentativesRequest: UpdateSalesRepresentativesRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/admin/accounts/${encodeURIComponent(accountId)}/sales-representatives`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateSalesRepresentativesRequest
    })));
}
/**
 * Update status account
 */
export function updateAccountStatus({ accountId, accountStatusRequest }: {
    accountId: string;
    accountStatusRequest: AccountStatusRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/status`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountStatusRequest
    })));
}
/**
 * Get a configured SSO from account
 */
export function getAccountSso({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountSsoDetailedResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Update a configured SSO from account
 */
export function updateSso({ id, accountSsoUpdateRequest }: {
    id: string;
    accountSsoUpdateRequest: AccountSsoUpdateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountSsoUpdateRequest
    })));
}
/**
 * Delete a configured Account SSO from account
 */
export function deleteSso({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update partially a configured SSO from account
 */
export function partialUpdateSso({ id, accountSsoPartialUpdateRequest }: {
    id: string;
    accountSsoPartialUpdateRequest: AccountSsoPartialUpdateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountSsoPartialUpdateRequest
    })));
}
/**
 * Get SSO Group Mapping By Id
 */
export function getGroupMapById({ id, attributeImporterId }: {
    id: string;
    attributeImporterId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountSsoGroupMappingResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping/${encodeURIComponent(attributeImporterId)}`, {
        ...opts
    }));
}
/**
 * Update SSO Group Mapping
 */
export function updateGroupMapping({ id, attributeImporterId, accountGroupMappingUpdateRequest }: {
    id: string;
    attributeImporterId: string;
    accountGroupMappingUpdateRequest: AccountGroupMappingUpdateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 204;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping/${encodeURIComponent(attributeImporterId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: accountGroupMappingUpdateRequest
    })));
}
/**
 * Delete SSO Group Mapping
 */
export function deleteGroupMapping({ id, attributeImporterId }: {
    id: string;
    attributeImporterId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping/${encodeURIComponent(attributeImporterId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Create Service Credential v2
 */
export function createServiceCredential({ serviceCredentialCreateRequestV2 }: {
    serviceCredentialCreateRequestV2: ServiceCredentialCreateRequestV2;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ServiceCredentialCreateResponseV2;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v2/service-credentials", oazapfts.json({
        ...opts,
        method: "POST",
        body: serviceCredentialCreateRequestV2
    })));
}
/**
 * Get Service Credentials
 */
export function getServiceCredentials({ name, status, page, size }: {
    name?: string;
    status?: string;
    page?: string;
    size?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ServiceCredentialResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/service-credentials${QS.query(QS.explode({
        name,
        status,
        page,
        size
    }))}`, {
        ...opts
    }));
}
/**
 * Create Service Credential
 */
export function createServiceCredential1({ serviceCredentialCreateRequest }: {
    serviceCredentialCreateRequest: ServiceCredentialCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ServiceCredentialCreateResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/service-credentials", oazapfts.json({
        ...opts,
        method: "POST",
        body: serviceCredentialCreateRequest
    })));
}
/**
 * Notify about Service Credential expiration
 */
export function notifyServiceCredentialExpiration({ serviceCredentialExpirationRequest }: {
    serviceCredentialExpirationRequest: ServiceCredentialExpirationNotifyRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
    } | {
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/service-credentials/expiration", oazapfts.json({
        ...opts,
        method: "POST",
        body: serviceCredentialExpirationRequest
    })));
}
/**
 * Find all the existing scopes
 */
export function findScopes({ page }: {
    page?: PagingRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ScopeResponse[];
    } | {
        status: 400;
        data: ScopeResponse[];
    } | {
        status: 500;
        data: ScopeResponse[];
    }>(`/v1/scopes${QS.query(QS.explode({
        page
    }))}`, {
        ...opts
    }));
}
/**
 * Create a new scope
 */
export function createScope({ iamCreateScopeRequest }: {
    iamCreateScopeRequest: IamCreateScopeRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText("/v1/scopes", oazapfts.json({
        ...opts,
        method: "POST",
        body: iamCreateScopeRequest
    })));
}
/**
 * Get Roles
 */
export function getRoles({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RoleResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Create a new role for Account
 */
export function createAccountRole({ createAccountRoleRequest }: {
    createAccountRoleRequest: CreateAccountRoleRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/roles", oazapfts.json({
        ...opts,
        method: "POST",
        body: createAccountRoleRequest
    })));
}
/**
 * Get role members
 */
export function getRoleMembers({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    roleId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RoleMemberResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Add Role to Member
 */
export function addRoleToMember({ roleId, addRoleToMemberRequest }: {
    roleId: string;
    addRoleToMemberRequest: AddRoleToMemberRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AddUsersToRoleResponse;
    } | {
        status: 201;
        data: AddUsersToRoleResponse;
    } | {
        status: 207;
        data: AddUsersToRoleResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/members`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addRoleToMemberRequest
    })));
}
/**
 * Get Groups from role
 */
export function getRoleGroups({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    roleId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RoleGroupResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Bind roles groups
 */
export function bindRoleGroups({ roleId, roleGroupIdsRequest }: {
    roleId: string;
    roleGroupIdsRequest: RoleGroupIdsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
    } | {
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/groups`, oazapfts.json({
        ...opts,
        method: "POST",
        body: roleGroupIdsRequest
    })));
}
/**
 * Get Resources
 */
export function getResources({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ResourceResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Create a new resource for Account
 */
export function createResource({ createResourceRequest }: {
    createResourceRequest: CreateResourceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/resources", oazapfts.json({
        ...opts,
        method: "POST",
        body: createResourceRequest
    })));
}
/**
 * Get Account Members
 */
export function getAccountMembers1({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountMemberResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Create user on current tenant
 */
export function createUser({ createUserRequest }: {
    createUserRequest: CreateUserRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: CreateUserResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/members", oazapfts.json({
        ...opts,
        method: "POST",
        body: createUserRequest
    })));
}
/**
 * Get member roles
 */
export function getRoles1({ memberId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    memberId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ReadMemberRoleResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Bind users to roles
 */
export function bindToRoles({ memberId, memberIdsCommonRequest }: {
    memberId: string;
    memberIdsCommonRequest: MemberIdsCommonRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountMemberStackspotIamDto;
    } | {
        status: 201;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/roles`, oazapfts.json({
        ...opts,
        method: "POST",
        body: memberIdsCommonRequest
    })));
}
/**
 * Reset member OTP
 */
export function resetOtp({ memberId }: {
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/reset-otp`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get member Groups
 */
export function getMemberGroups({ memberId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    memberId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GroupReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Bind users to groups
 */
export function bindToGroups({ memberId, memberIdsCommonRequest }: {
    memberId: string;
    memberIdsCommonRequest: MemberIdsCommonRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountMemberStackspotIamDto;
    } | {
        status: 201;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/groups`, oazapfts.json({
        ...opts,
        method: "POST",
        body: memberIdsCommonRequest
    })));
}
/**
 * Get account member favorite resources
 */
export function listAccountMemberFavorites({ memberId }: {
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FavoriteResourcesResponse[];
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/favorites`, {
        ...opts
    }));
}
/**
 * Add account member favorite resource
 */
export function addAccountMemberFavorite({ memberId, createFavoriteResourceRequest }: {
    memberId: string;
    createFavoriteResourceRequest: CreateFavoriteResourceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 409;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/favorites`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createFavoriteResourceRequest
    })));
}
/**
 * Send Email to reset password
 */
export function resetPassword({ memberEmailRequest }: {
    memberEmailRequest: MemberEmailRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/members/reset-password", oazapfts.json({
        ...opts,
        method: "POST",
        body: memberEmailRequest
    })));
}
/**
 * Invite new members to account
 */
export function inviteMembers({ body }: {
    body: InviteNewMemberRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: InviteMembersResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/members/invitation", oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * List user invitations for account
 */
export function listUserInvitations({ filterBy, filterValue, page, size }: {
    filterBy?: string;
    filterValue?: string;
    page?: string;
    size?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: UserInvitationResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/invitations${QS.query(QS.explode({
        filterBy,
        filterValue,
        page,
        size
    }))}`, {
        ...opts
    }));
}
/**
 * Create user invitations for account
 */
export function createUserInvitation({ createUserInvitationRequest }: {
    createUserInvitationRequest: CreateUserInvitationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 201;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/invitations", oazapfts.json({
        ...opts,
        method: "POST",
        body: createUserInvitationRequest
    })));
}
/**
 * Get Groups
 */
export function getGroups({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn, includeDefaultGroup }: {
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
    includeDefaultGroup?: boolean;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GroupReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn,
        includeDefaultGroup
    }))}`, {
        ...opts
    }));
}
/**
 * Create Group
 */
export function save({ newGroupRequest }: {
    newGroupRequest: NewGroupRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IamIdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/groups", oazapfts.json({
        ...opts,
        method: "POST",
        body: newGroupRequest
    })));
}
/**
 * Get Group members
 */
export function getRoles2({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    groupId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RoleResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/roles${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Bind Group roles
 */
export function bindRoles({ groupId, groupRoleIdsRequest }: {
    groupId: string;
    groupRoleIdsRequest: GroupRoleIdsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
    } | {
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/roles`, oazapfts.json({
        ...opts,
        method: "POST",
        body: groupRoleIdsRequest
    })));
}
/**
 * Get Group Resources
 */
export function getGroupResources({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    groupId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ResourceResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/resources${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Add Multiple Resources to Group
 */
export function addResourcesToGroup({ groupId, addResourcesToGroupRequest }: {
    groupId: string;
    addResourcesToGroupRequest: AddResourcesToGroupRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/resources`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addResourcesToGroupRequest
    })));
}
/**
 * Get Group members
 */
export function getMembers({ groupId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    groupId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ReadGroupMembersResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Bind Group member
 */
export function bindGroupMembers({ groupId, groupMemberIdsRequest }: {
    groupId: string;
    groupMemberIdsRequest: GroupMemberIdsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
    } | {
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/members`, oazapfts.json({
        ...opts,
        method: "POST",
        body: groupMemberIdsRequest
    })));
}
/**
 * List Feature Flags in an Account
 */
export function getFeatures(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FeatureFlagsResponse[];
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: FeatureFlagsResponse[];
    }>("/v1/feature-flags", {
        ...opts
    }));
}
/**
 * Create a Feature Flag
 */
export function create({ createFeatureFlagRequest }: {
    createFeatureFlagRequest: CreateFeatureFlagRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
    }>("/v1/feature-flags", oazapfts.json({
        ...opts,
        method: "POST",
        body: createFeatureFlagRequest
    })));
}
/**
 * Send Welcome Email
 */
export function sendWelcomeEmail({ welcomeEmailRequest }: {
    welcomeEmailRequest: WelcomeEmailRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/emails/welcome", oazapfts.json({
        ...opts,
        method: "POST",
        body: welcomeEmailRequest
    })));
}
/**
 * Send Download Email
 */
export function sendDownloadEmail(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/emails/download", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Send contact email
 */
export function sendContactEmail({ contactEmailRequest }: {
    contactEmailRequest: ContactEmailRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/emails/contact", oazapfts.json({
        ...opts,
        method: "POST",
        body: contactEmailRequest
    })));
}
/**
 * Switch user to account passed in the body
 */
export function switchAccount({ switchAccountRequest }: {
    switchAccountRequest: SwitchAccountRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ExchangeAccountResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/authentication/switch-account", oazapfts.json({
        ...opts,
        method: "POST",
        body: switchAccountRequest
    })));
}
/**
 * Get credentials from personal service client (create if not exists one to the logged user).
 */
export function getPersonalClientCredentials(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PersonalClientCredentialsResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/authentication/personal-client-credentials", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get an Access Token generated With Audience on logged user behalf, since previously authorized.
 */
export function getPersonalAccessToken1({ apiAuthorization, generateTokenFromPersonalAccountRequest }: {
    apiAuthorization: string;
    generateTokenFromPersonalAccountRequest?: GenerateTokenFromPersonalAccountRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PersonalAccessTokenResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/authentication/personal-access-token", oazapfts.json({
        ...opts,
        method: "POST",
        body: generateTokenFromPersonalAccountRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "Api-Authorization": apiAuthorization
        })
    })));
}
/**
 * Disable client that generates Personal Access Tokens.
 */
export function disablePersonalAccessTokenGeneration(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/authentication/personal-access-token/unauthorize", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Verify if exists a client created (authorized) to generate Access Token on logged user behalf.
 */
export function personalAccessTokenAuthorization(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/authentication/personal-access-token/authorize", {
        ...opts
    }));
}
/**
 * Authorize creating a client to generate Personal Access Tokens on your behalf.
 */
export function createPersonalAccessToken(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/authentication/personal-access-token/authorize", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Get an Access Token generated on logged user's service account behalf, since previously authorized.
 */
export function getPersonalAccessTokenFromServiceAccountOnly(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PersonalAccessTokenResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/authentication/personal-access-token-sa", {
        ...opts,
        method: "POST"
    }));
}
/**
 * Generate device code.
 */
export function deviceCode({ email, freemiumAccess, codeChallenge, body }: {
    email: string;
    freemiumAccess?: boolean;
    codeChallenge: string;
    body?: Blob;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DeviceCodeResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/authentication/device-code${QS.query(QS.explode({
        email,
        freemium_access: freemiumAccess,
        code_challenge: codeChallenge
    }))}`, oazapfts.form({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Find accounts
 */
export function getAccounts1({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PageResponseAccountResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/admin/accounts${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Create an Enterprise Account Order
 */
export function create1({ createAccountOrderRequest }: {
    createAccountOrderRequest: CreateAccountOrderRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
    }>("/v1/admin/accounts", oazapfts.json({
        ...opts,
        method: "POST",
        body: createAccountOrderRequest
    })));
}
/**
 * Get All SSO Group Mapping
 */
export function getAllGroupMapping({ id, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    id: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PageResponseAccountSsoGroupMappingResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Create SSO Group Mapping
 */
export function createGroupMapping({ id, accountGroupMappingCreateRequest }: {
    id: string;
    accountGroupMappingCreateRequest: AccountGroupMappingCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}/group-mapping`, oazapfts.json({
        ...opts,
        method: "POST",
        body: accountGroupMappingCreateRequest
    })));
}
/**
 * Configure new SSO in account
 */
export function ssoConfigure({ protocol, accountSsoCreateRequest }: {
    protocol: "SAML" | "OPENID";
    accountSsoCreateRequest: AccountSsoCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
        data: SsoIdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/protocol/${encodeURIComponent(protocol)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: accountSsoCreateRequest
    })));
}
/**
 * Parse SSO configuration file
 */
export function ssoParseConfigurationFile({ protocol, body }: {
    protocol: "SAML" | "OPENID";
    body?: {
        name?: string;
        alias?: string;
        file: Blob;
    };
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountSsoParserResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/protocol/${encodeURIComponent(protocol)}/parser`, oazapfts.multipart({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Create Account Freemium
 */
export function createAccountFreemium({ accountFreemiumCreateRequest }: {
    accountFreemiumCreateRequest: AccountFreemiumCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/accounts/freemium", oazapfts.json({
        ...opts,
        method: "POST",
        body: accountFreemiumCreateRequest
    })));
}
/**
 * Adding Collaborator Account
 */
export function addCollaborator({ accountCollaboratorRequest }: {
    accountCollaboratorRequest: AccountCollaboratorRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/accounts/freemium/collaborators", oazapfts.json({
        ...opts,
        method: "POST",
        body: accountCollaboratorRequest
    })));
}
/**
 * Create an Account Order for Partner
 */
export function createPartner({ createAccountPartnerOrderRequest }: {
    createAccountPartnerOrderRequest: CreateAccountPartnerOrderRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
        data: IdDefaultResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: IdDefaultResponse;
    }>("/v1/account-orders/partners", oazapfts.json({
        ...opts,
        method: "POST",
        body: createAccountPartnerOrderRequest
    })));
}
/**
 * Validate new Partner account data
 */
export function validateNewPartnerData({ validateAccountPartnerDataRequest }: {
    validateAccountPartnerDataRequest: ValidateAccountPartnerDataRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
    }>("/v1/account-orders/partners/validations/creation-data", oazapfts.json({
        ...opts,
        method: "POST",
        body: validateAccountPartnerDataRequest
    })));
}
/**
 * Rotate Service Credential Secret v2
 */
export function rotateServiceCredentialSecret({ id, rotateServiceCredentialSecretRequest }: {
    id: string;
    rotateServiceCredentialSecretRequest: RotateServiceCredentialSecretRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RotateServiceCredentialSecretResponse;
    } | {
        status: 204;
        data: RotateServiceCredentialSecretResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/service-credentials/${encodeURIComponent(id)}/secret`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: rotateServiceCredentialSecretRequest
    })));
}
/**
 * Revoke Service Credential v2
 */
export function revokeServiceCredential({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/service-credentials/${encodeURIComponent(id)}/revoke`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Associate Group to Service Credential
 */
export function associateGroupToServiceCredential({ id, serviceCredentialAssociateGroupRequest }: {
    id: string;
    serviceCredentialAssociateGroupRequest: ServiceCredentialAssociateGroupRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/service-credentials/${encodeURIComponent(id)}/teams`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: serviceCredentialAssociateGroupRequest
    })));
}
/**
 * Revoke Service Credential
 */
export function revokeServiceCredential1({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/service-credentials/${encodeURIComponent(id)}/revocation`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Update member preferences
 */
export function getMemberPreferences({ memberId }: {
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
        data: ReadPreferencesResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/preferences`, {
        ...opts
    }));
}
/**
 * Update member preferences
 */
export function updateMemberPreferences({ memberId, updateMemberPreferencesRequest }: {
    memberId: string;
    updateMemberPreferencesRequest: UpdateMemberPreferencesRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/preferences`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateMemberPreferencesRequest
    })));
}
/**
 * Delete a member permissions
 */
export function deleteMemberPermissions({ memberId, deleteMemberPermissionsRequest }: {
    memberId: string;
    deleteMemberPermissionsRequest: DeleteMemberPermissionsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/permissions`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteMemberPermissionsRequest
    })));
}
/**
 * Update member permissions
 */
export function updateMemberPermissions({ memberId, updateMemberPermissionsRequest }: {
    memberId: string;
    updateMemberPermissionsRequest: UpdateMemberPermissionsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/permissions`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateMemberPermissionsRequest
    })));
}
/**
 * Get All Fido credentials for given user
 */
export function getAllMemberFidoCredentials({ memberId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    memberId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: MemberCredentialsListResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/fido-credentials${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Disables Fido credentials for given user
 */
export function deactivateFidoCredentials({ memberId }: {
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/fido-credentials`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Enable Fido credentials for given user
 */
export function enableFidoCredentials({ memberId }: {
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/fido-credentials`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Revoke user invitation
 */
export function cancelUserInvitation({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 409;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/invitations/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Resend user invitation
 */
export function resendUserInvitation({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 409;
        data: ErrorResponse;
    } | {
        status: 429;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/invitations/${encodeURIComponent(id)}`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Get Group by ID
 */
export function getGroupById({ groupId }: {
    groupId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GroupReadDetailResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}`, {
        ...opts
    }));
}
/**
 * Delete Group
 */
export function deleteV1GroupsByGroupId({ groupId }: {
    groupId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update Group
 */
export function update1({ groupId, updateGroupRequest }: {
    groupId: string;
    updateGroupRequest: UpdateGroupRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IamIdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateGroupRequest
    })));
}
/**
 * Delete Group permissions
 */
export function deleteGroupPermissions({ groupId, deleteGroupPermissionsRequest }: {
    groupId: string;
    deleteGroupPermissionsRequest: DeleteGroupPermissionsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/permissions`, oazapfts.json({
        ...opts,
        method: "DELETE",
        body: deleteGroupPermissionsRequest
    })));
}
/**
 * Update Group permissions
 */
export function updateGroupPermissions({ groupId, updateGroupPermissionsRequest }: {
    groupId: string;
    updateGroupPermissionsRequest: UpdateGroupPermissionsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/permissions`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateGroupPermissionsRequest
    })));
}
/**
 * Enable a Feature Flag in an Account
 */
export function enable({ id }: {
    id: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
    }>(`/v1/feature-flags/${encodeURIComponent(id)}/enable`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Disable a Feature Flag in an Account
 */
export function disable({ id }: {
    id: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
    }>(`/v1/feature-flags/${encodeURIComponent(id)}/disable`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Adds a new email domain to an existing account
 */
export function addNewEmailDomain({ accountId, addNewEmailDomainRequest }: {
    accountId: string;
    addNewEmailDomainRequest: AddNewEmailDomainRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/admin/accounts/${encodeURIComponent(accountId)}/domains`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: addNewEmailDomainRequest
    })));
}
/**
 * Get configured account SSO attributes
 */
export function ssoGetAttributesConfig({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountSsoAttributesConfigResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}/attributes-configuration`, {
        ...opts
    }));
}
/**
 * Configure account SSO attributes
 */
export function ssoAddAttributes({ id, accountSsoAddAttributesConfigRequest }: {
    id: string;
    accountSsoAddAttributesConfigRequest: AccountSsoAddAttributesConfigRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso/${encodeURIComponent(id)}/attributes-configuration`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountSsoAddAttributesConfigRequest
    })));
}
/**
 * Get Partner by account Id
 */
export function getPartnerAccount({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account-partners/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Delete Partner
 */
export function deletePartner({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account-partners/${encodeURIComponent(id)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update Partner Account data.
 */
export function updatePartnerAccountData({ id, accountPartnerAccountDataUpdateRequest }: {
    id: string;
    accountPartnerAccountDataUpdateRequest: AccountPartnerAccountDataUpdateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account-partners/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountPartnerAccountDataUpdateRequest
    })));
}
/**
 * Update Partner Account Admin data.
 */
export function updatePartnerAccountAdminData({ id, accountPartnerAdminDataUpdateRequest }: {
    id: string;
    accountPartnerAdminDataUpdateRequest: AccountPartnerAdminDataUpdateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account-partners/${encodeURIComponent(id)}/admin`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: accountPartnerAdminDataUpdateRequest
    })));
}
/**
 * Verify User repository credential configuration status.
 */
export function isCreatedScmCredentials(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/users/scm-credentials/configuration-status", {
        ...opts
    }));
}
/**
 * List Accounts of User can use and collaborate.
 */
export function getUserAccountWithRole(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountCollaborationInfoResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/users/accounts", {
        ...opts
    }));
}
/**
 * Get Accounts
 */
export function getAccounts({ accountType, slug, id }: {
    accountType?: string;
    slug?: string;
    id?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountDataReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/stackspot/accounts${QS.query(QS.explode({
        accountType,
        slug,
        id
    }))}`, {
        ...opts
    }));
}
/**
 * Get Account by id
 */
export function getAccount({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountDataReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/stackspot/accounts/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Get Account Members
 */
export function getAccountMembers({ id, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    id: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountMemberResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/stackspot/accounts/${encodeURIComponent(id)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Service Credential by ID (Client ID)
 */
export function getServiceCredential({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ServiceCredentialDetailsResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/service-credentials/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * List service credential permissions
 */
export function getServiceCredentialPermissions({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: FullResourceResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/service-credentials/${encodeURIComponent(id)}/permissions`, {
        ...opts
    }));
}
/**
 * List service credential groups
 */
export function getServiceCredentialGroups({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GroupReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/service-credentials/${encodeURIComponent(id)}/groups`, {
        ...opts
    }));
}
/**
 * Get SCM provider.
 */
export function getScmProvider(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountScmProviderResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/scm-credentials/provider", {
        ...opts
    }));
}
/**
 * Verify repository credential configuration status.
 */
export function isCreatedScmCredentials1(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountScmStatusResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/scm-credentials/configuration-status", {
        ...opts
    }));
}
/**
 * Get Permissions from role
 */
export function getPermissionsWithStatus({ roleId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    roleId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: RolePermissionResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/permissions${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Members from Resource
 */
export function getResourceMembers({ resourceId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    resourceId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: MembersFromResourceResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/resources/${encodeURIComponent(resourceId)}/members${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Groups from Resource
 */
export function getResourceGroups({ resourceId, size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    resourceId: string;
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GroupsFromResourceResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/resources/${encodeURIComponent(resourceId)}/groups${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Resource Types
 */
export function getResourceTypes({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ResourceTypeResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/resource-types${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Validate permission access
 */
export function getAccess({ resourceType, resource, action, attribute }: {
    resourceType: string;
    resource: string;
    action: string;
    attribute?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 403;
        data: ErrorResponse;
    }>(`/v1/permissions/access${QS.query(QS.explode({
        resourceType,
        resource,
        action,
        attribute
    }))}`, {
        ...opts
    }));
}
/**
 * Get Resources Members
 */
export function getResources1({ memberId }: {
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ResourceDto[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/resources`, {
        ...opts
    }));
}
/**
 * Get member favorite resources by type
 */
export function listMemberFavoritesByResource({ memberId, resourceType }: {
    memberId: string;
    resourceType: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: any;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/favorites/resources/${encodeURIComponent(resourceType)}`, {
        ...opts
    }));
}
/**
 * Get Account Collaborators Members
 */
export function getAccountMembersToCollaborators({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountMemberResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/collaborators${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Get Authentication Configuration for account
 */
export function getAuthenticationConfig({ tenantSlug }: {
    tenantSlug: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountAuthenticationConfigResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/authentication/config${QS.query(QS.explode({
        tenant_slug: tenantSlug
    }))}`, {
        ...opts
    }));
}
/**
 * Get an account by ID
 */
export function getAccount1({ accountId }: {
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/admin/accounts/${encodeURIComponent(accountId)}`, {
        ...opts
    }));
}
/**
 * Get account data (tenant, IDP alias, account freemium available to user) by email
 */
export function getAccountDataByEmail({ email }: {
    email: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountDataResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts${QS.query(QS.explode({
        email
    }))}`, {
        ...opts
    }));
}
/**
 * Get account Slug - this API is temporary
 */
export function getAccountSlug({ accountId }: {
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountSlugResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/${encodeURIComponent(accountId)}`, {
        ...opts
    }));
}
/**
 * Get all SSO configured in account
 */
export function getAllAccountSso({ size, page, sort, direction, search, filterMode, filterBy, filterValue, multiFilterMode, filterIn }: {
    size?: any;
    page?: any;
    sort?: string;
    direction?: "ASC" | "DESC";
    search?: string;
    filterMode?: "MATCH" | "CONTAINS" | "IN";
    filterBy?: string;
    filterValue?: string;
    multiFilterMode?: string;
    filterIn?: any;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PageResponseAccountSsoResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 409;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/sso${QS.query(QS.explode({
        size,
        page,
        sort,
        direction,
        search,
        filterMode,
        filterBy,
        filterValue,
        multiFilterMode,
        filterIn
    }))}`, {
        ...opts
    }));
}
/**
 * Returns information if account data (name and slug) are available to be registered
 */
export function accountDataIsAvailable({ name, slug }: {
    name: string;
    slug: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountDataAvailableResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/accounts/freemium/available${QS.query(QS.explode({
        name,
        slug
    }))}`, {
        ...opts
    }));
}
/**
 * Get Partners who are sharing content.
 */
export function getSharingPartners({ name }: {
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountPartnerResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account-partners/sharing${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Get Partners with whom it is allowed to share content.
 */
export function getPartnersSharingAllowed({ name }: {
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountPartnerResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account-partners/sharing-allowed${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Validate Partner association limit
 */
export function validatePartnerAssociationLimit(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
    }>("/v1/account-orders/partners/validations/association-limit", {
        ...opts
    }));
}
/**
 * Disassociate Group to Service Credential
 */
export function disassociateGroupToServiceCredential({ id, groupId }: {
    id: string;
    groupId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/service-credentials/${encodeURIComponent(id)}/teams/${encodeURIComponent(groupId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Remove Role from Member
 */
export function removeRoleFromMember({ roleId, memberId }: {
    roleId: string;
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/roles/${encodeURIComponent(roleId)}/members/${encodeURIComponent(memberId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete a resource from Account
 */
export function deleteResource({ resourceId }: {
    resourceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/resources/${encodeURIComponent(resourceId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Remove resource from member favorites
 */
export function deleteMemberFavorite({ memberId, resourceType, resourceId }: {
    memberId: string;
    resourceType: string;
    resourceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/members/${encodeURIComponent(memberId)}/favorites/${encodeURIComponent(resourceType)}/${encodeURIComponent(resourceId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete Group role
 */
export function deleteRole({ groupId, roleId }: {
    groupId: string;
    roleId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/roles/${encodeURIComponent(roleId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete Resource from Group
 */
export function deleteResourceFromGroup({ groupId, resourceId }: {
    groupId: string;
    resourceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/resources/${encodeURIComponent(resourceId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete Group member
 */
export function deleteMember({ groupId, memberId }: {
    groupId: string;
    memberId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/groups/${encodeURIComponent(groupId)}/members/${encodeURIComponent(memberId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
