// @stack-spot/portal-network@0.66.0/src/api/cloudPlatformHorizon.ts
/**
 * PlatformHorizonApplication
 * 0.0.2
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://cloud-platform-horizon.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://cloud-platform-horizon.dev.stackspot.com"
};
export type ManifestoResource = {
    schemaVersion: string;
    kind: string;
    metadata: {
        [key: string]: string;
    };
    spec: {
        [key: string]: object;
    };
};
export type ValidationDetails = {
    code: string;
    field: string;
    values: string[];
};
export type ErrorBody = {
    code: string;
    details: string;
    status: number;
    validationDetails: ValidationDetails[];
};
export type StatusType = number;
/**
 * Applies a manifesto to the control plane.
 */
export function apply({ manifestoResource }: {
    manifestoResource: ManifestoResource;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 202;
    } | {
        status: 403;
        data: ErrorBody;
    } | {
        status: 422;
        data: {
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
            detail?: string;
            title?: string;
            status?: StatusType;
        };
    } | {
        status: 500;
        data: {
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
            detail?: string;
            title?: string;
            status?: StatusType;
        };
    }>("/v1/admin/control-plane/apply", oazapfts.json({
        ...opts,
        method: "POST",
        body: manifestoResource
    })));
}
