// @stack-spot/portal-network@0.66.0/src/api/eventBus.ts
/**
 * EventBusApi
 * 0.2.7
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://idp-event-bus-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://idp-event-bus-api.dev.stackspot.com"
};
export type Event = {
    /** Mandatory, defined by the client application */
    id: string;
    /** Mandatory, '/{workspaceSlug}/{app_name}' */
    source: string;
    /** Mandatory, fixed to '1.0' */
    specversion: string;
    /** Mandatory, UTC ISO 8601 ('YYYY-MM-DDTHH:MM:SSZ'), defined by client application */
    time: string;
    /** Mandatory, 'workspace_slug.event_name' */
    "type": string;
    /** Mandatory, application-specific event data */
    data: {
        [key: string]: object;
    };
};
export type SendEventsRequest = {
    /** List of events */
    events: Event[];
};
export type EventResult = object;
export type Error = EventResult & {
    eventId?: string;
    errorCode?: string;
    errorMessage?: string;
};
export type Success = EventResult & {
    eventId?: string;
    partition?: number;
    offset?: number;
    timestamp?: number;
};
export type EventProcessingResponse = {
    errorCount: number;
    results: (Error | Success)[];
};
export type StatusType = number;
export type SendEncryptedEventsRequest = {
    /** Base64 encoded + encrypted Events */
    data: string;
    /** Base64 encoded + encrypted random Symmetric Key */
    sessionKey: string;
};
export function sendEvents({ stackspotCustomerIp, sendEventsRequest }: {
    stackspotCustomerIp?: string;
    sendEventsRequest: SendEventsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: EventProcessingResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
            status?: StatusType;
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
            status?: StatusType;
        };
    }>("/v1/events", oazapfts.json({
        ...opts,
        method: "POST",
        body: sendEventsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    })));
}
export function sendEncryptedEvents({ stackspotCustomerIp, sendEncryptedEventsRequest }: {
    stackspotCustomerIp?: string;
    sendEncryptedEventsRequest: SendEncryptedEventsRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: EventProcessingResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
            status?: StatusType;
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
            status?: StatusType;
        };
    }>(`/v1/events/workspace/${encodeURIComponent(workspaceSlug)}/event/${encodeURIComponent(eventName)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: sendEncryptedEventsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    })));
}
export function getPublicKey(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
            status?: StatusType;
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
            status?: StatusType;
        };
    }>("/v1/events/key", {
        ...opts
    }));
}
