// @stack-spot/portal-network@0.66.0/src/api/insights.ts
/**
 * Insights API
 * v1.0.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://insights-insights-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://insights-insights-api.dev.stackspot.com"
};
export type NewPanelRequest = {
    panelType: string;
};
export type IdResponse = {
    /** Id response. */
    id?: number;
};
export type ValidationDetails = {
    code: string;
    field?: string;
    details?: string;
    values?: string[];
};
export type ErrorResponse = {
    code: string;
    status: number;
    details: string;
    validationDetails?: ValidationDetails[];
};
export type AddPanelToDashboardRequest = {
    panel: string;
};
export type NewDatasetRequest = {
    name: string;
    description?: string;
    scope: string;
    query: string;
    panelId: number;
};
export type Panel = {
    /** Panel id. */
    id?: number;
    /** The name of the dataset used to generate this panel. */
    datasetName?: string;
    /** The type of the panel. */
    panelType?: string;
    /** Order to display this panel. */
    order?: number;
    /** Data for this panel. */
    data?: {
        [key: string]: object;
    }[];
};
export type DashboardReadResponse = {
    /** Panels in a dashboard. */
    panels?: Panel[];
};
/**
 * Create Panel
 */
export function save({ newPanelRequest }: {
    newPanelRequest: NewPanelRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/panels", oazapfts.json({
        ...opts,
        method: "POST",
        body: newPanelRequest
    })));
}
/**
 * Add Panel to Dashboard Studio
 */
export function addToDashboardStudio({ accountSlug, studioSlug, addPanelToDashboardRequest }: {
    accountSlug: string;
    studioSlug: string;
    addPanelToDashboardRequest: AddPanelToDashboardRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/panels/account/${encodeURIComponent(accountSlug)}/studio/${encodeURIComponent(studioSlug)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addPanelToDashboardRequest
    })));
}
/**
 * Remove Panel from Dashboard Studio
 */
export function removeFromDashboardStudio({ accountSlug, studioSlug, panelId }: {
    accountSlug: string;
    studioSlug: string;
    panelId: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/panels/account/${encodeURIComponent(accountSlug)}/studio/${encodeURIComponent(studioSlug)}${QS.query(QS.explode({
        panelId
    }))}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Create Dataset
 */
export function save1({ newDatasetRequest }: {
    newDatasetRequest: NewDatasetRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/datasets", oazapfts.json({
        ...opts,
        method: "POST",
        body: newDatasetRequest
    })));
}
/**
 * Download Studio Data
 */
export function downloadStudioData({ accountSlug, studioSlug, startDate, endDate, stackspotCustomerIp }: {
    accountSlug: string;
    studioSlug: string;
    startDate: string;
    endDate: string;
    stackspotCustomerIp?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/assets/account/${encodeURIComponent(accountSlug)}/studio/${encodeURIComponent(studioSlug)}/csv${QS.query(QS.explode({
        startDate,
        endDate
    }))}`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    }));
}
/**
 * Download Event account data file
 */
export function streamingAccountData({ accountSlug, startDate, endDate, dataset }: {
    accountSlug: string;
    startDate: string;
    endDate: string;
    dataset: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchBlob<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/assets/account/${encodeURIComponent(accountSlug)}/export${QS.query(QS.explode({
        startDate,
        endDate,
        dataset
    }))}`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Download Account Data
 */
export function downloadAccountData({ accountSlug, startDate, endDate, stackspotCustomerIp }: {
    accountSlug: string;
    startDate: string;
    endDate: string;
    stackspotCustomerIp?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/assets/account/${encodeURIComponent(accountSlug)}/csv${QS.query(QS.explode({
        startDate,
        endDate
    }))}`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    }));
}
/**
 * Get Dashboard Studio
 */
export function getDashboardStudio({ accountSlug, studioSlug, stackspotCustomerIp }: {
    accountSlug: string;
    studioSlug: string;
    stackspotCustomerIp?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: DashboardReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/dashboards/account/${encodeURIComponent(accountSlug)}/studio/${encodeURIComponent(studioSlug)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    }));
}
