// @stack-spot/portal-network@0.66.0/src/api/notification.ts
/**
 * Notification Engine
 * 1.0.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://account-notification-engine.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://account-notification-engine.dev.stackspot.com"
};
export type NotificationContentResponse = {
    title: string;
    description: string;
};
export type GetTenantNotificationsResponse = {
    id: string;
    content: {
        [key: string]: NotificationContentResponse;
    };
    broadcast_level: "ACCOUNT" | "INDIVIDUAL" | "PLATFORM" | "RESOURCE";
    context: "ACCOUNT" | "STUDIO" | "WORKSPACE" | "AI";
    target?: string;
    criticality: "LOW" | "MEDIUM" | "HIGH";
    call_to_action: string;
    persistent: boolean;
    committed: boolean;
    last_until?: string;
    trigger_at: string;
    created_at?: string;
};
export type PagingResponseModelGetTenantNotificationsResponse = {
    items: GetTenantNotificationsResponse[];
    size: number;
    total_elements: number;
};
export type ErrorDetails = {
    code: "INTEGRATION_EXCEPTION" | "INVALID_TENANT" | "VALIDATION_ERROR" | "MANDATORY_FIELD_NOT_INFORMED" | "INVALID_REQUEST" | "NOTIFICATION_NOT_FOUND" | "INVALID_FIELD" | "MISSING_TOKEN";
    field?: string;
    details?: string;
    values?: string[];
};
export type ApiErrorResponse = {
    code?: "INTEGRATION_EXCEPTION" | "INVALID_TENANT" | "VALIDATION_ERROR" | "MANDATORY_FIELD_NOT_INFORMED" | "INVALID_REQUEST" | "NOTIFICATION_NOT_FOUND" | "INVALID_FIELD" | "MISSING_TOKEN";
    status: number;
    message: string;
    details?: ErrorDetails[];
};
export type NotificationContentRequest = {
    /** User friendly title for the notification */
    title: string;
    /** User friendly description for the notification (allows the use of markdown) */
    description: string;
};
export type CreateTenantNotificationRequest = {
    /** Map of language codes to notification content. Each entry represents the content of the notification in a specific language. */
    content: {
        [key: string]: NotificationContentRequest;
    };
    /** Level of broadcasting to the notification. */
    broadcast_level: "ACCOUNT" | "INDIVIDUAL" | "PLATFORM" | "RESOURCE" | "ACCOUNT" | "INDIVIDUAL" | "RESOURCE";
    /** The target of the notification, user id for 'individual', resource id for 'resource', leave empty for any other */
    target: string;
    /** Criticality of the notification. */
    criticality: "LOW" | "MEDIUM" | "HIGH" | "LOW" | "MEDIUM" | "HIGH";
    /** Which context to publish the notification */
    context: "ACCOUNT" | "STUDIO" | "WORKSPACE" | "AI" | "Account" | "Studio" | "Workspace";
    /** Possible redirect link for the notification button. */
    call_to_action: string;
    /** If the notification should be re-consumed until 'last_until' is reached. */
    persistent: boolean;
    /** UTC formatted timestamp to when the notification should finally end, if the persistent is set to true. */
    last_until?: string;
    /** UTC formatted timestamp for when to triggers the notification. */
    trigger_at: string;
};
export type CreatePlatformNotificationRequest = {
    /** Map of language codes to notification content. Each entry represents the content of the notification in a specific language. */
    content: {
        [key: string]: NotificationContentRequest;
    };
    /** Criticality of the notification. */
    criticality: "LOW" | "MEDIUM" | "HIGH" | "LOW" | "MEDIUM" | "HIGH";
    /** Which context to publish the notification */
    context: "ACCOUNT" | "STUDIO" | "WORKSPACE" | "AI" | "Account" | "Studio" | "Workspace";
    /** Possible redirect link for the notification button. */
    call_to_action: string;
    /** If the notification should be re-consumed until last_until is reached. */
    persistent: boolean;
    /** UTC formatted timestamp to when the notification should finally ended if the persistent is set to true. */
    last_until?: string;
    /** UTC formatted timestamp for when to triggers the notification. */
    trigger_at: string;
};
export type CreatePlatformNotificationResponse = {
    id: string;
};
export type GetCommittedNotificationsResponse = {
    /** If there are uncommitted notifications for the user */
    uncommitted_messages: boolean;
};
/**
 * Get notifications for tenant
 */
export function findAll1({ size, page, createdSince, context, criticality, content, committed, tenantId }: {
    size?: number;
    page?: number;
    createdSince?: number;
    context?: "ACCOUNT" | "STUDIO" | "WORKSPACE" | "AI";
    criticality?: "LOW" | "MEDIUM" | "HIGH";
    content?: string;
    committed?: boolean;
    tenantId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PagingResponseModelGetTenantNotificationsResponse;
    } | {
        status: 403;
        data: ApiErrorResponse;
    } | {
        status: 500;
        data: ApiErrorResponse;
    }>(`/v1/tenants/${encodeURIComponent(tenantId)}/notifications${QS.query(QS.explode({
        size,
        page,
        createdSince,
        context,
        criticality,
        content,
        committed
    }))}`, {
        ...opts
    }));
}
/**
 * Generate a new notification intent to the specified tenant
 */
export function create({ tenantId, createTenantNotificationRequest }: {
    tenantId: string;
    createTenantNotificationRequest: CreateTenantNotificationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: object;
    } | {
        status: 403;
        data: ApiErrorResponse;
    } | {
        status: 500;
        data: ApiErrorResponse;
    }>(`/v1/tenants/${encodeURIComponent(tenantId)}/notifications`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createTenantNotificationRequest
    })));
}
/**
 * Allows to mark a notification as committed
 */
export function commit({ notificationId, tenantId }: {
    notificationId: string;
    tenantId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ApiErrorResponse;
    } | {
        status: 500;
        data: ApiErrorResponse;
    }>(`/v1/tenants/${encodeURIComponent(tenantId)}/notifications/${encodeURIComponent(notificationId)}/commit`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Generate a new notification intent to the specified tenant
 */
export function create1({ createPlatformNotificationRequest }: {
    createPlatformNotificationRequest: CreatePlatformNotificationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CreatePlatformNotificationResponse;
    }>("/v1/notifications", oazapfts.json({
        ...opts,
        method: "POST",
        body: createPlatformNotificationRequest
    })));
}
/**
 * Get notifications for tenant
 */
export function findAll({ size, page, createdSince, context, criticality, content, committed, tenantId }: {
    size?: number;
    page?: number;
    createdSince?: number;
    context?: "ACCOUNT" | "STUDIO" | "WORKSPACE" | "AI";
    criticality?: "LOW" | "MEDIUM" | "HIGH";
    content?: string;
    committed?: boolean;
    tenantId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PagingResponseModelGetTenantNotificationsResponse;
    } | {
        status: 403;
        data: ApiErrorResponse;
    } | {
        status: 500;
        data: ApiErrorResponse;
    }>(`/v2/tenants/${encodeURIComponent(tenantId)}/notifications${QS.query(QS.explode({
        size,
        page,
        createdSince,
        context,
        criticality,
        content,
        committed
    }))}`, {
        ...opts
    }));
}
/**
 * Allows to check if there are any uncommitted notifications for the user
 */
export function committedNotifications({ tenantId }: {
    tenantId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetCommittedNotificationsResponse;
    } | {
        status: 403;
        data: ApiErrorResponse;
    } | {
        status: 500;
        data: ApiErrorResponse;
    }>(`/v1/tenants/${encodeURIComponent(tenantId)}/notifications/committed`, {
        ...opts
    }));
}
/**
 * Actuator web endpoint 'health'
 */
export function health(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>("/healthz", {
        ...opts
    }));
}
/**
 * Actuator web endpoint 'health-path'
 */
export function getHealthz(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: object;
    }>("/healthz/**", {
        ...opts
    }));
}
/**
 * Delete notification by id
 */
export function deleteV1TenantsByTenantIdNotificationsAndNotificationId({ notificationId, tenantId }: {
    notificationId: string;
    tenantId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ApiErrorResponse;
    } | {
        status: 500;
        data: ApiErrorResponse;
    }>(`/v1/tenants/${encodeURIComponent(tenantId)}/notifications/${encodeURIComponent(notificationId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Allows to unmark a notification as committed
 */
export function uncommit({ notificationId, tenantId }: {
    notificationId: string;
    tenantId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ApiErrorResponse;
    } | {
        status: 500;
        data: ApiErrorResponse;
    }>(`/v1/tenants/${encodeURIComponent(tenantId)}/notifications/${encodeURIComponent(notificationId)}/uncommit`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Batches deletes the notification in all tenants
 */
export function delete1({ notificationId }: {
    notificationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/notifications/${encodeURIComponent(notificationId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
