// @stack-spot/portal-network@0.66.0/src/api/serviceCatalog.ts
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://catalog-tech-products-catalog-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://catalog-tech-products-catalog-api.dev.stackspot.com"
};
export type ProductResponse = {
    id: string;
    accountId?: string;
    name: string;
    description: string;
    updatedAt: string;
    createdAt: string;
    workspaceId: string;
    status: string;
    imageUrl?: string;
};
export type StatusType = number;
export type ProductRequest = {
    name?: string;
    description?: string;
    imageBase64?: string;
};
export type ClientApplicationResponse = {
    id?: string;
    name?: string;
    displayName?: string;
    description?: string;
};
export type ClientApplicationRequest = {
    name: string;
    displayName: string;
    description: string;
};
export type ProductWithApiAndEnvironmentRequest = {
    name: string;
    description: string;
    environments?: string[];
    apis?: string[];
    imageBase64?: string;
};
export type CatalogApiByEnvResponse = {
    updatedAt: string;
    createdAt: string;
    environmentId: string;
    status: string;
    version: string;
    gatewayUrl?: string;
    hasOpenApiContent: boolean;
};
export type CatalogApiResponse = {
    id: string;
    name: string;
    description: string;
    workspaceId: string;
    connectionInterfaceSlug?: string;
    byEnvironment: CatalogApiByEnvResponse[];
};
export type CatalogApiDefinitionRequest = {
    environmentId: string;
    gatewayUrl: string;
    openApiContent: string;
};
export type CatalogApiRequest = {
    name: string;
    description: string;
    connectionInterfaceSlug?: string;
    definitions?: CatalogApiDefinitionRequest[];
};
export type CatalogApiDefinitionResponse = {
    apiId: string;
    updatedAt: string;
    createdAt: string;
    environmentId: string;
    status: string;
    version: string;
    basicInfo?: CatalogApiResponse;
};
export type AccessRequestRequest = {
    status: "GRANTED" | "DENIED";
    oauthClientId?: string;
};
export type AccessRequestRequestWrite = {
    status: "GRANTED" | "DENIED";
    oAuthClientId?: string;
    oauthClientId?: string;
};
export type AccessRequestBatchRequest = {
    status: "GRANTED" | "DENIED";
    accessRequests: string[];
    oauthClientId?: string;
};
export type AccessRequestBatchRequestWrite = {
    status: "GRANTED" | "DENIED";
    oAuthClientId?: string;
    accessRequests: string[];
    oauthClientId?: string;
};
export type CatalogApiDetailsDto = {
    id: string;
    updatedAt: string;
    createdAt: string;
    environmentId: string;
    status: string;
    version: string;
    name: string;
    description: string;
    workspaceId: string;
    connectionInterfaceSlug?: string;
};
export type EndpointParameterResponse = {
    name: string;
    "type"?: string;
    description?: string;
    required?: boolean;
    "in"?: string;
    format?: string;
};
export type BodySchema = {
    name?: string;
    "type"?: string;
    description?: string;
    format?: string;
};
export type EndpointResponseResponse = {
    code?: string;
    description?: string;
    responseExample?: {
        [key: string]: object;
    };
    schema?: BodySchema[];
};
export type CatalogApiMethodResponse = {
    "type": string;
    endpointName: string;
    endpointUrl: string;
    name: string;
    description: string;
    endpointParameters: EndpointParameterResponse[];
    curlExample: string;
    bodyExample?: object;
    responses: EndpointResponseResponse[];
    produces?: string[];
    consumes?: string[];
};
export type CatalogApiMethodsResponse = CatalogApiMethodResponse[];
export type SimpleProductResponse = {
    id: string;
    accountId?: string;
    name: string;
    description: string;
    workspaceId: string;
    status: string;
    imageUrl?: string;
    environments: string[];
};
export type CatalogApiDetailsResponse = {
    id: string;
    updatedAt: string;
    createdAt: string;
    environmentId: string;
    status: string;
    version: string;
    name: string;
    description: string;
    workspaceId: string;
};
export type SimpleApiResponse = {
    id: string;
    name: string;
    description: string;
    workspaceId: string;
    accountId?: string;
    environments: string[];
};
export type CatalogApiAccessRequestResponse = {
    id: string;
    environmentId: string;
    authType: string;
    status: string;
    updatedAt: string;
    createdAt: string;
    grantedAt?: string;
    grantedBy?: string;
    updatedBy?: string;
    clientApplication: {
        [key: string]: object;
    };
    api: {
        [key: string]: object;
    };
    oauthClientId: string;
};
export type CatalogApiAccessRequestResponseWrite = {
    id: string;
    environmentId: string;
    oAuthClientId?: string;
    authType: string;
    status: string;
    updatedAt: string;
    createdAt: string;
    grantedAt?: string;
    grantedBy?: string;
    updatedBy?: string;
    clientApplication: {
        [key: string]: object;
    };
    api: {
        [key: string]: object;
    };
    oauthClientId: string;
};
export type PartnerProductResponse = {
    id: string;
    accountId?: string;
    name: string;
    description: string;
    workspaceId: string;
    status?: string;
    imageUrl?: string;
    environments: string[];
};
export type PartnerApiResponse = {
    id: string;
    accountId?: string;
    name: string;
    description: string;
    workspaceId: string;
    environments: string[];
};
export type ClientApplicationWithStatusResponse = {
    id?: string;
    name?: string;
    displayName?: string;
    description?: string;
    lastAccessRequestStatus?: "DENIED" | "GRANTED" | "REQUESTED";
};
export type ClientApplicationWithGrantedAccessResponse = {
    id?: string;
    name?: string;
    displayName?: string;
    description?: string;
    clientId: string;
};
export type EnvironmentResponse = {
    id: string;
    name: string;
};
export type HealthzResponse = {
    status: "UP" | "DOWN";
    groups: ("liveness" | "readiness")[];
};
/**
 * Get one product by product_id
 */
export function getOne({ workspaceId, productId }: {
    workspaceId: string;
    productId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}`, {
        ...opts
    }));
}
/**
 * Update product's name and/or description
 */
export function updateProduct({ workspaceId, productId, productRequest }: {
    workspaceId: string;
    productId: string;
    productRequest: ProductRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: productRequest
    })));
}
/**
 * Get one product by product_id
 */
export function deleteProduct({ workspaceId, productId }: {
    workspaceId: string;
    productId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Change product status
 */
export function changeProductStatus({ workspaceId, productId, action }: {
    workspaceId: string;
    productId: string;
    action: "publish" | "unpublish";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}${QS.query(QS.explode({
        action
    }))}`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Get Client Application by ID
 */
export function getOneClientApplication({ clientApplicationId }: {
    clientApplicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ClientApplicationResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/client-applications/${encodeURIComponent(clientApplicationId)}`, {
        ...opts
    }));
}
/**
 * Update a Client Application
 */
export function updateClientApplication({ clientApplicationId, clientApplicationRequest }: {
    clientApplicationId: string;
    clientApplicationRequest: ClientApplicationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/client-applications/${encodeURIComponent(clientApplicationId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: clientApplicationRequest
    })));
}
/**
 * Delete a Client Application by ID
 */
export function deleteClientApplication({ clientApplicationId }: {
    clientApplicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/client-applications/${encodeURIComponent(clientApplicationId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get all products by workspace Id
 */
export function getAll({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products`, {
        ...opts
    }));
}
/**
 * save product in a workspace Id
 */
export function saveProduct({ workspaceId, productWithApiAndEnvironmentRequest }: {
    workspaceId: string;
    productWithApiAndEnvironmentRequest: ProductWithApiAndEnvironmentRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products`, oazapfts.json({
        ...opts,
        method: "POST",
        body: productWithApiAndEnvironmentRequest
    })));
}
export function linkProductApi({ workspaceId, productId, apiId }: {
    workspaceId: string;
    productId: string;
    apiId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts,
        method: "POST"
    }));
}
export function unlinkProductApi({ workspaceId, productId, apiId }: {
    workspaceId: string;
    productId: string;
    apiId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get one API by application id
 */
export function getOne1({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api`, {
        ...opts
    }));
}
/**
 * save API in a workspace Id
 */
export function saveApi({ workspaceId, applicationId, catalogApiRequest }: {
    workspaceId: string;
    applicationId: string;
    catalogApiRequest: CatalogApiRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api`, oazapfts.json({
        ...opts,
        method: "POST",
        body: catalogApiRequest
    })));
}
/**
 * delete one API by application id
 */
export function deleteApi({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * get API definition
 */
export function getApiDefinition({ workspaceId, applicationId, environmentId }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiDefinitionResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api/environments/${encodeURIComponent(environmentId)}/definition`, {
        ...opts
    }));
}
/**
 * save API definition in a workspace Id
 */
export function saveApiDefinition({ workspaceId, applicationId, environmentId, gatewayUrl, body }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    gatewayUrl?: string;
    body: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api/environments/${encodeURIComponent(environmentId)}/definition${QS.query(QS.explode({
        gatewayUrl
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * update API definition in a workspace Id
 */
export function updateApiDefinition({ workspaceId, applicationId, environmentId, gatewayUrl, connectionInterfaceSlug, body }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    gatewayUrl?: string;
    connectionInterfaceSlug?: string;
    body?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api/environments/${encodeURIComponent(environmentId)}/definition${QS.query(QS.explode({
        gatewayUrl,
        connectionInterfaceSlug
    }))}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Get all apis by workspace Id
 */
export function getAll1({ workspaceId, offset, limit, status }: {
    workspaceId: string;
    offset?: number;
    limit?: number;
    status?: "DRAFT" | "PUBLISHED" | "UNPUBLISHED" | "DELETED";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis${QS.query(QS.explode({
        offset,
        limit,
        status
    }))}`, {
        ...opts
    }));
}
/**
 * save API in a workspace Id
 */
export function saveApi1({ workspaceId, catalogApiRequest }: {
    workspaceId: string;
    catalogApiRequest: CatalogApiRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis`, oazapfts.json({
        ...opts,
        method: "POST",
        body: catalogApiRequest
    })));
}
/**
 * get API definition
 */
export function getApiDefinition1({ workspaceId, apiId, environmentId }: {
    workspaceId: string;
    apiId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiDefinitionResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}/definition`, {
        ...opts
    }));
}
/**
 * save API definition in a workspace Id
 */
export function saveNewApiDefinition({ workspaceId, apiId, environmentId, gatewayUrl, body }: {
    workspaceId: string;
    apiId: string;
    environmentId: string;
    gatewayUrl?: string;
    body: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}/definition${QS.query(QS.explode({
        gatewayUrl
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * update API definition in a workspace Id
 */
export function updateApiDefinition1({ workspaceId, apiId, environmentId, gatewayUrl, connectionInterfaceSlug, body }: {
    workspaceId: string;
    apiId: string;
    environmentId: string;
    gatewayUrl?: string;
    connectionInterfaceSlug?: string;
    body?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: boolean;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}/definition${QS.query(QS.explode({
        gatewayUrl,
        connectionInterfaceSlug
    }))}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Grant partner access permission to access the product
 */
export function givePermission({ partnerAccountId, environmentId, productId }: {
    partnerAccountId: string;
    environmentId: string;
    productId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Remove partner access permission from the product
 */
export function removePermission({ partnerAccountId, environmentId, productId }: {
    partnerAccountId: string;
    environmentId: string;
    productId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Grant partner access permission to access the API
 */
export function givePermission1({ partnerAccountId, environmentId, apiId }: {
    partnerAccountId: string;
    environmentId: string;
    apiId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Remove partner access permission from the API
 */
export function removePermission1({ partnerAccountId, environmentId, apiId }: {
    partnerAccountId: string;
    environmentId: string;
    apiId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Request access to many client application by Product
 */
export function createAccessRequestByProduct({ environmentId, productId, body }: {
    environmentId: string;
    productId: string;
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}/access-requests`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Request access to many client application by Api
 */
export function createAccessRequestByApi({ environmentId, apiId, body }: {
    environmentId: string;
    apiId: string;
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/access-requests`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Update access request by status and clientID
 */
export function updateAccessRequest({ environmentId, accessRequestId, accessRequestRequest }: {
    environmentId: string;
    accessRequestId: string;
    accessRequestRequest: AccessRequestRequestWrite;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/environments/${encodeURIComponent(environmentId)}/access-requests/${encodeURIComponent(accessRequestId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: accessRequestRequest
    })));
}
/**
 * Get all Client Applications from Account
 */
export function getClientApplication(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ClientApplicationResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>("/v1/partner/client-applications", {
        ...opts
    }));
}
/**
 * Save a new Client Application for Account
 */
export function createClientApplication({ clientApplicationRequest }: {
    clientApplicationRequest: ClientApplicationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: string;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>("/v1/partner/client-applications", oazapfts.json({
        ...opts,
        method: "POST",
        body: clientApplicationRequest
    })));
}
/**
 * Update batch access request by status and clientID
 */
export function updateAccessRequestBatch({ accessRequestBatchRequest }: {
    accessRequestBatchRequest: AccessRequestBatchRequestWrite;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>("/v1/partner/access-requests", oazapfts.json({
        ...opts,
        method: "POST",
        body: accessRequestBatchRequest
    })));
}
/**
 * Update API status by application id
 */
export function updateApiStatusByWorkspaceAndApplicationId({ workspaceId, applicationId, environmentId, action }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    action: "PUBLISHED" | "UNPUBLISHED" | "DRAFT";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api/environments/${encodeURIComponent(environmentId)}${QS.query(QS.explode({
        action
    }))}`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Update API status
 */
export function updateApiStatusByWorkspaceAndApiId({ workspaceId, apiId, envId, action }: {
    workspaceId: string;
    apiId: string;
    envId: string;
    action: "PUBLISHED" | "UNPUBLISHED" | "DRAFT";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(envId)}${QS.query(QS.explode({
        action
    }))}`, {
        ...opts,
        method: "PATCH"
    }));
}
export function getApisByProductId({ workspaceId, productId }: {
    workspaceId: string;
    productId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}/apis`, {
        ...opts
    }));
}
/**
 * Get one API by api_id
 */
export function getOne2({ workspaceId, apiId }: {
    workspaceId: string;
    apiId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts
    }));
}
/**
 * Get all products of an API
 */
export function getProductsByApiId({ workspaceId, apiId }: {
    workspaceId: string;
    apiId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/products`, {
        ...opts
    }));
}
/**
 * get API Details
 */
export function apiDetails({ workspaceId, apiId, environmentId }: {
    workspaceId: string;
    apiId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiDetailsDto;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}`, {
        ...opts
    }));
}
/**
 * get API Definition
 */
export function getApiWorkspaceMethods({ workspaceId, apiId, environmentId }: {
    workspaceId: string;
    apiId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiMethodsResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}/methods`, {
        ...opts
    }));
}
/**
 * List all the Account's Products
 */
export function getProducts(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SimpleProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>("/v1/service-catalog/products", {
        ...opts
    }));
}
export function getApisByProductId1({ productId, environmentId }: {
    productId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}/apis`, {
        ...opts
    }));
}
export function getApisByProductId2({ productId, environmentId, accountId }: {
    productId: string;
    environmentId: string;
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}/apis`, {
        ...opts
    }));
}
/**
 * Get one product by product_id
 */
export function getOneProduct({ accountId, productId, environmentId }: {
    accountId: string;
    productId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}`, {
        ...opts
    }));
}
/**
 * Get one product by product_id
 */
export function getOneProduct1({ productId, environmentId }: {
    productId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}`, {
        ...opts
    }));
}
/**
 * Get all products
 */
export function getAllProducts({ accountId, environmentId, limit }: {
    accountId: string;
    environmentId: string;
    limit?: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/products${QS.query(QS.explode({
        limit
    }))}`, {
        ...opts
    }));
}
/**
 * Get all products
 */
export function getAllProducts1({ environmentId, limit }: {
    environmentId: string;
    limit?: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/products${QS.query(QS.explode({
        limit
    }))}`, {
        ...opts
    }));
}
/**
 * Get all products of an API
 */
export function getProductsByApiId1({ apiId, environmentId, accountId }: {
    apiId: string;
    environmentId: string;
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/products`, {
        ...opts
    }));
}
/**
 * Get all products of an API
 */
export function getProductsByApiId2({ apiId, environmentId }: {
    apiId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/products`, {
        ...opts
    }));
}
/**
 * get API definition
 */
export function getApiMethods({ accountId, apiId, environmentId }: {
    accountId: string;
    apiId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiMethodsResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/methods`, {
        ...opts
    }));
}
/**
 * get API definition
 */
export function getApiMethods1({ apiId, environmentId }: {
    apiId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiMethodsResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/methods`, {
        ...opts
    }));
}
/**
 * get API definition
 */
export function getApiDefinition2({ apiId, environmentId, accountId }: {
    apiId: string;
    environmentId: string;
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiDetailsResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/details`, {
        ...opts
    }));
}
/**
 * get API definition
 */
export function getApiDefinition3({ apiId, environmentId }: {
    apiId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiDetailsResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/details`, {
        ...opts
    }));
}
/**
 * Get one API by api_id
 */
export function getOneApi({ apiId, environmentId }: {
    apiId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts
    }));
}
/**
 * Get one API by api_id
 */
export function getOneApi1({ apiId, environmentId, accountId }: {
    apiId: string;
    environmentId: string;
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts
    }));
}
/**
 * Get all apis
 */
export function getAllApis({ environmentId, limit }: {
    environmentId: string;
    limit?: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis${QS.query(QS.explode({
        limit
    }))}`, {
        ...opts
    }));
}
/**
 * Get all apis
 */
export function getAllApis1({ accountId, environmentId, limit }: {
    accountId: string;
    environmentId: string;
    limit?: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis${QS.query(QS.explode({
        limit
    }))}`, {
        ...opts
    }));
}
/**
 * List all the Account's Apis
 */
export function getApis(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SimpleApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>("/v1/service-catalog/apis", {
        ...opts
    }));
}
/**
 * Get all partner APIs whose status is requested or granted
 */
export function getRequestedOrGrantedAccessApis({ partnerAccountId }: {
    partnerAccountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiAccessRequestResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/requested-or-granted-access-apis`, {
        ...opts
    }));
}
/**
 * Get all Partner's products
 */
export function getProducts1({ partnerAccountId }: {
    partnerAccountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PartnerProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/products`, {
        ...opts
    }));
}
/**
 * Get all Partner's products by environment
 */
export function getProductsByEnv({ partnerAccountId, environmentId }: {
    partnerAccountId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PartnerProductResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/products`, {
        ...opts
    }));
}
/**
 * Get all Partner's apis by environment
 */
export function getApisByEnv({ partnerAccountId, environmentId }: {
    partnerAccountId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PartnerApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/apis`, {
        ...opts
    }));
}
/**
 * Get all Partner's apis
 */
export function getApis1({ partnerAccountId }: {
    partnerAccountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PartnerApiResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/${encodeURIComponent(partnerAccountId)}/apis`, {
        ...opts
    }));
}
/**
 * Get all Client Applications that are eligible to request access to API
 */
export function getEligibleClientApplications({ environmentId, apiId }: {
    environmentId: string;
    apiId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ClientApplicationWithStatusResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/client-applications`, {
        ...opts
    }));
}
export function getAllApplications({ environmentId, apiId }: {
    environmentId: string;
    apiId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ClientApplicationWithGrantedAccessResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/approved-client-applications`, {
        ...opts
    }));
}
/**
 * access request list requested
 */
export function findAccessRequestRequestedStatus({ environmentId }: {
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiAccessRequestResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/environments/${encodeURIComponent(environmentId)}/access-requests`, {
        ...opts
    }));
}
/**
 * Access request list history
 */
export function findAccessRequestRequestedHistory({ environmentId }: {
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CatalogApiAccessRequestResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/partner/environments/${encodeURIComponent(environmentId)}/access-requests/history`, {
        ...opts
    }));
}
/**
 * Get allowed environments
 */
export function getEnvironmentsByAccountId({ accountId }: {
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: EnvironmentResponse[];
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>(`/v1/accounts/${encodeURIComponent(accountId)}/environments`, {
        ...opts
    }));
}
/**
 * Get health status
 */
export function healthz(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: HealthzResponse;
    } | {
        status: 422;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    } | {
        status: 500;
        data: {
            title?: string;
            detail?: string;
            status?: StatusType;
            instance?: string;
            "type"?: string;
            parameters?: {
                [key: string]: object;
            };
        };
    }>("/healthz", {
        ...opts
    }));
}
