// @stack-spot/portal-network@0.66.0/src/api/workspace.ts
/**
 * Workspace API
 * v1.0.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults: Oazapfts.Defaults<Oazapfts.CustomHeaders> = {
    headers: {},
    baseUrl: "https://workspace-workspace-api.stg.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://workspace-workspace-api.stg.stackspot.com"
};
export type WorkspaceVariableResponse = {
    /** Workspace variable name. */
    name: string;
    /** Workspace variable value. */
    value?: string;
    /** Workspace variable description. */
    description: string;
    /** Workspace variable mandate flag. */
    mandate: boolean;
    /** Variable value source. */
    source: string;
    /** Workspace variable creation data */
    createdAt: string;
};
export type ValidationDetails = {
    code: string;
    field?: string;
    details?: string;
    values?: string[];
};
export type ErrorResponse = {
    code: string;
    status: number;
    details: string;
    validationDetails?: ValidationDetails[];
};
export type UpsertWorkspaceVariableRequest = {
    /** Workspace variable value. */
    value: string;
};
export type ContextAttribute = {
    key: string;
    value: object;
    final: boolean;
};
export type ContextValuesRequest = {
    /** Input attributes context values. */
    valuesContext: ContextAttribute[];
    /** Plugin o Action Version Id. */
    externalVersionId: string;
    /** Environment Name (For specific environments). */
    envName?: string;
};
export type ImportContextValueRequest = {
    /** Context type (PLUGIN, ACTION). */
    "type": "PLUGIN" | "ACTION";
    /** PluginId or ActionId. */
    externalId: string;
    /** PluginVersionId or ActionVersionId. */
    externalVersionId: string;
    /** List of keys to be imported. */
    attributes: string[];
};
export type ImportContextV2Request = {
    /** StackVersionId reference to be imported from. */
    stackVersionReference: string;
    /** Target contexts to be imported. */
    targetContexts: ImportContextValueRequest[];
};
export type ContextAttributeDto = {
    key: string;
    value: object;
    final: boolean;
    source?: "ACCOUNT" | "STACK" | "WORKSPACE";
};
export type ActionContextResponse = {
    actionId: string;
    versionRange: string;
    stepType?: "CREATE_REPO";
    valuesContext: ContextAttributeDto[];
};
export type WorkflowContextResponse = {
    accountId: string;
    actionsBefore: ActionContextResponse[];
    actionsAfter: ActionContextResponse[];
};
export type WorkflowActionContextRequest = {
    actionId: string;
    valuesContext: ContextAttribute[];
};
export type WorkspaceWorkflowCreateRequest = {
    actionsBefore: WorkflowActionContextRequest[];
    actionsAfter: WorkflowActionContextRequest[];
};
export type WorkflowActionResponse = {
    actionVersionId: string;
    orderStep: number;
    createRepo?: boolean;
};
export type WorkflowConditionResponse = {
    condition: "BEFORE" | "AFTER";
    actions: WorkflowActionResponse[];
};
export type WorkflowReadResponse = {
    identifier: string;
    workflowConditions: WorkflowConditionResponse[];
};
export type WorkflowActionsRequest = {
    /** Workflow action version id. */
    actionVersionId: string;
    /** Workspace order step. */
    orderStep: number;
    /** This action will create repository. */
    createRepo: boolean;
};
export type WorkflowConditionRequest = {
    /** Workflow condition. */
    condition: "BEFORE" | "AFTER";
    /** Workflow actions. */
    actions: WorkflowActionsRequest[];
};
export type AddWorkspaceWorkflowStepRequest = {
    /** Workflow identifier. */
    identifier: string;
    workflowConditions: WorkflowConditionRequest[];
};
export type UpdateSharedInfraStatusRequest = {
    /** Shared Infrastructure Status. */
    status: "CREATED" | "CREATE_FAILED" | "DRAFT" | "DELETING" | "DELETE_FAILED";
};
export type IdResponse = {
    /** Id response. */
    id: string;
};
export type SharedInfraVisibleLinksResponse = {
    id: string;
    visible: boolean;
};
export type UpdateSharedInfraLinkVisibilityRequest = {
    id: string;
    visible: boolean;
};
export type Task = {
    runTaskId: string;
    "type": "IAC" | "DEPLOY" | "DESTROY" | "IAC_SELF_HOSTED" | "DEPLOY_SELF_HOSTED" | "DESTROY_SELF_HOSTED";
    status: "EXTERNAL_ERROR" | "INTERNAL_ERROR" | "ABORTED" | "FAILED" | "SUCCEEDED" | "RUNNING" | "PENDING" | "READY_TO_RUN";
    message?: string;
};
export type Plugins = {
    pluginAlias: string;
    tasks: Task[];
};
export type ConnectionInterface = {
    selected: string;
    "type": string;
    alias: string;
};
export type Connections = {
    requires?: ConnectionInterface[];
    generates?: ConnectionInterface[];
};
export type GeneratesLinks = {
    name: string;
    url: string;
    "type": string;
    picture: string;
};
export type PluginLink = {
    generates: GeneratesLinks[];
};
export type PluginsStatusInCloud = {
    statusInCloud: "EXTERNAL_ERROR" | "INTERNAL_ERROR" | "ABORTED" | "FAILED" | "SUCCEEDED" | "RUNNING" | "PENDING" | "READY_TO_RUN";
    name: string;
    alias: string;
    "type": string;
    pluginVersionId: string;
    stackVersionId: string;
    inputs: {
        [key: string]: object;
    };
    connections: Connections;
    link?: PluginLink;
};
export type UpdateRunRequest = {
    /** Activity status to update. */
    status: "RUNNING" | "SUCCEEDED" | "USER_ERROR" | "RUNTIME_ERROR" | "EXTERNAL_ERROR" | "ABORTED" | "ABORTING";
    /** Message. */
    message?: string;
    /** List of plugins that contain in deploy, rollback or destroy. */
    plugins: Plugins[];
    /** List of plugins in cloud. */
    pluginsStatusInCloud: PluginsStatusInCloud[];
};
export type RunTaskRequest = {
    /** Plugin Id. */
    runTaskId: string;
    /** Plugin type. */
    "type": "IAC" | "DEPLOY" | "DESTROY" | "IAC_SELF_HOSTED" | "DEPLOY_SELF_HOSTED" | "DESTROY_SELF_HOSTED";
    /** Plugin status to update. */
    status: "EXTERNAL_ERROR" | "INTERNAL_ERROR" | "ABORTED" | "FAILED" | "SUCCEEDED" | "RUNNING" | "PENDING" | "READY_TO_RUN";
    /** Plugin Alias. */
    pluginAlias: string;
    /** Add message when plugin deployment gives error. */
    message?: string;
};
export type DeploymentTagRequest = {
    /** Application id. */
    applicationId?: string;
    /** Shared Infra id. */
    sharedInfrastructureId?: string;
    /** Related tag. */
    relatedTag: string;
};
export type ConnectionInterfaceAttribute = {
    /** Attribute key. */
    key: string;
    /** Attribute value. */
    value: object;
    /** Indicates if this attribute value is sensitive or not. */
    sensitive: boolean;
};
export type UpdateConnectionInterfaceAttributesRequest = {
    /** Environment id. */
    environmentId: string;
    /** List of attributes */
    attributes: ConnectionInterfaceAttribute[];
};
export type ConnectionInterfaceAttributesByEnv = {
    /** Environment id. */
    environmentId: string;
    /** List of attributes */
    attributes: ConnectionInterfaceAttribute[];
};
export type UpsertConnectionInterfaceRequest = {
    applicationId?: string;
    sharedInfraId?: string;
    slug: string;
    typeId: string;
    sourcePluginVersionId?: string;
    attributesByEnv: ConnectionInterfaceAttributesByEnv[];
};
export type UpdateApplicationStatusRequest = {
    /** Application Status. */
    status: "CREATED" | "CREATE_FAILED" | "DRAFT" | "DELETING" | "DELETE_FAILED" | "ARCHIVING" | "ARCHIVED" | "ARCHIVE_FAILED";
};
export type ApplicationVisibleLinksResponse = {
    id: string;
    visible: boolean;
};
export type UpdateApplicationLinkVisibilityRequest = {
    id: string;
    visible: boolean;
};
export type WorkflowSettingsResponse = {
    name: string;
    description: string;
    "type": string;
    isDefault: boolean;
    workflowDefaultId?: string;
    stacksId: string[];
};
export type SettingsWorkflowSaveRequest = {
    name: string;
    description: string;
    "type": "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
    isDefault: boolean;
    stacksId: string[];
};
export type ContextWorkflowRequest = {
    actionId: string;
    valuesContext: ContextAttribute[];
};
export type ContextsWorkflowSaveRequest = {
    actionsBefore: ContextWorkflowRequest[];
    actionsAfter: ContextWorkflowRequest[];
};
export type ActionWorkflowResponse = {
    actionId: string;
    order: number;
    versionRange: string;
    stepType?: "CREATE_REPO";
};
export type WorkflowActionsResponse = {
    accountId: string;
    actionsBefore: ActionWorkflowResponse[];
    actionsAfter: ActionWorkflowResponse[];
};
export type WorkflowActionRequest = {
    actionId: string;
    order: number;
    versionRange: string;
    stepType?: "CREATE_REPO";
};
export type ActionsWorkflowSaveRequest = {
    actionsBefore: WorkflowActionRequest[];
    actionsAfter: WorkflowActionRequest[];
};
export type AccountVariableResponse = {
    /** Account variable name */
    name: string;
    /** Account variable value stored */
    value?: string;
    /** Account variable description */
    description: string;
    /** Account variable mandatory flag */
    mandate: boolean;
    /** Account variable creation data */
    createdAt: string;
};
export type UpdateAccountVariableRequest = {
    /** New account variable value */
    value?: string;
    /** New account variable description */
    description: string;
    /** New account variable mandatory flag */
    mandate: boolean;
};
export type AddAccountWorkflowStepRequest = {
    /** Workflow identifier. */
    identifier: string;
    workflowConditions: WorkflowConditionRequest[];
};
export type StackWorkflowCreateRequest = {
    actionsBefore: WorkflowActionContextRequest[];
    actionsAfter: WorkflowActionContextRequest[];
};
export type WorkspaceResponse = {
    /** Workspace id. */
    id: string;
    /** Workspace name */
    name: string;
    /** Workspace description */
    description?: string;
    /** Workspace image */
    imageUrl?: string;
};
export type WorkspacePageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: WorkspaceResponse[];
};
export type GetWorkspacePageRequest = {
    name?: string;
    createdBy?: string;
    favorites?: string[];
};
export type GetSharedInfraRequestBody = {
    name?: string;
    stackVersionIds?: string[];
};
export type ExtraRepositoriesReadResponse = {
    /** Extra repository name. */
    name: string;
    /** Extra repository description. */
    description: string;
    /** Extra repository provider. */
    provider: "GITHUB" | "GITLAB" | "BITBUCKET" | "AZUREDEVOPS";
    /** Extra repository url. */
    url: string;
};
export type SharedInfraDetailsResponse = {
    /** Shared infrastructure id. */
    id: string;
    /** Shared infrastructure name. */
    name: string;
    /** Shared infrastructure description. */
    description?: string;
    /** Shared infrastructure repository url. */
    repoUrl?: string;
    /** Shared infrastructure repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this shared infrastructure. */
    stackVersionId: string;
    /** Starter used to generate this shared infrastructure. */
    starterId?: string;
    /** Shared Infrastructure status. */
    status: string;
    /** Application extra repositories. */
    extraRepositories: ExtraRepositoriesReadResponse[];
    /** Shared Infrastructure creator. */
    createdBy: string;
    /** Shared Infrastructure creation date. */
    createdAt: string;
    /** Shared Infrastructure updater. */
    updatedBy: string;
    /** Shared Infrastructure  update date. */
    updatedAt: string;
};
export type WorkspaceSharedInfraPageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: SharedInfraDetailsResponse[];
};
export type GetApplicationsRequestBody = {
    name?: string;
    stackVersionIds?: string[];
};
export type ApplicationReadResponse = {
    /** Application id. */
    id: string;
    /** Application name. */
    name: string;
    /** Application description. */
    description?: string;
    /** Application repository url. */
    repoUrl?: string;
    /** Application repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this application. */
    stackVersionId: string;
    /** Starter used to generate this application. */
    starterId: string;
    /** Application status. */
    status: string;
    /** Application extra repositories. */
    extraRepositories: ExtraRepositoriesReadResponse[];
    /** Application creator. */
    createdBy: string;
    /** Application creation date. */
    createdAt: string;
    /** Application updater. */
    updatedBy: string;
    /** Application update date. */
    updatedAt: string;
};
export type WorkspaceApplicationPageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: ApplicationReadResponse[];
};
export type EnvironmentSimpleResponse = {
    /** Environment id. */
    id: string;
    /** Environment name. */
    name: string;
};
export type TargetByPluginUsageResponse = {
    /** Plugin Version ID. */
    pluginVersionId: string;
    environment: EnvironmentSimpleResponse;
};
export type PluginUsageByTargetResponse = {
    /** Application id. */
    id: string;
    /** Application name. */
    name: string;
    /** Application description. */
    description?: string;
    /** Amount of linked versions. */
    totalLinkedVersions: number;
    workspace: WorkspaceResponse;
};
export type WorkspaceReadResponse = {
    /** Workspace id. */
    id: string;
    /** Workspace name. */
    name: string;
    /** Workspace description. */
    description?: string;
    /** Workspace image URL. */
    imageUrl?: string;
    /** Workspace manage runtime. */
    manageRuntime: boolean;
    /** Workspace accountId. */
    accountId: string;
    /** Workspace teams. - deprecated field */
    teams?: string[];
    /** Workspace creator. */
    createdBy: string;
    /** Workspace creation date. */
    createdAt: string;
    /** Workspace updater. */
    updatedBy: string;
    /** Workspace update date. */
    updatedAt: string;
};
export type NewWorkspaceRequest = {
    /** Workspace name. */
    name: string;
    /** Workspace description. */
    description?: string;
    /** Workspace image base64. */
    image?: string;
    /** Workspace manage runtime. */
    manageRuntime: boolean;
};
export type StackInWorkspaceReadResponse = {
    /** Stack version id. */
    stackVersionId: string;
    /** Workspace id. */
    workspaceId: string;
    /** Account id. */
    accountId: string;
};
export type StackVersionInfoRequest = {
    /** Stack id. */
    stackId: string;
    /** Stack slug. */
    stackSlug: string;
    /** Stack display name. */
    stackDisplayName: string;
    /** Stack semantic version. */
    stackVersionSemVer: string;
    /** Studio id. */
    studioId: string;
    /** Studio slug. */
    studioSlug: string;
};
export type AddStackInWorkspaceRequest = {
    /** Stack version id. */
    stackVersionId: string;
    stackVersionInfo: StackVersionInfoRequest;
};
export type StackVersionResponse = {
    stackVersionId: string;
    pluginIdsWithContext: string[];
};
export type StackWithWorkspaceContextResponse = {
    /** List of stack version with pluginVersionId list that has workspace context. */
    stackVersions: StackVersionResponse[];
    /** Workspace id. */
    workspaceId: string;
    /** Account id. */
    accountId: string;
};
export type ExtraRepositoriesRequest = {
    /** Extra repository name. */
    name: string;
    /** Extra repository description. */
    description: string;
    /** Extra repository provider. */
    provider: "GITHUB" | "GITLAB" | "BITBUCKET" | "AZUREDEVOPS";
    /** Extra repository url. */
    url: string;
};
export type CreateSharedInfraRequest = {
    /** Shared infrastructure name. */
    name: string;
    /** Shared infrastructure description. */
    description?: string;
    /** Shared infrastructure repository url. */
    repoUrl?: string;
    /** Shared infrastructure repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this shared infrastructure. */
    stackVersionId: string;
    /** Starter used to generate this shared infrastructure. */
    starterId?: string;
    /** Extra repository this shared infrastructure. */
    extraRepositories?: ExtraRepositoriesRequest[];
};
export type RecreateSharedInfraRequest = {
    /** Shared infrastructure name. */
    name: string;
    /** Shared infrastructure description. */
    description?: string;
    /** Shared infrastructure repository url. */
    repoUrl?: string;
    /** Shared infrastructure repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this shared infrastructure. */
    stackVersionId: string;
    /** Starter used to generate this shared infrastructure. */
    starterId?: string;
};
export type UpdateSharedInfraRequest = {
    /** Shared infrastructure name. */
    name?: string;
    /** Shared infrastructure description. */
    description?: string;
    /** Shared infrastructure repository URL. */
    repoUrl?: string;
    /** Shared infrastructure repository base branch. */
    repoBaseBranch?: string;
    /** Extra repository for this application. */
    extraRepositories?: ExtraRepositoriesRequest[];
};
export type SharedInfraLinkReadResponse = {
    /** Shared infra link id. */
    id: string;
    /** Shared infra link name. */
    name: string;
    /** Shared infra link url. */
    url: string;
    /** Shared infra link image URL. */
    imageUrl?: string;
};
export type CreateSharedInfraLinkRequest = {
    /** Shared infra link name. */
    name: string;
    /** Shared infra URL. */
    url: string;
    /** Shared infra image base64. */
    image?: string;
};
export type RequiredConnectionInterface = {
    connectionInterfaceId: string;
    "type": string;
    existing: boolean;
};
export type Attribute = {
    key: string;
    value: object;
};
export type PluginElementSnapshot = {
    pluginVersionId: string;
    "type": string;
    requiredConnectionInterfaces: RequiredConnectionInterface[];
    attributes: Attribute[];
};
export type SharedInfraDeploySnapshotRequest = {
    /** Deployment Id. */
    deploymentId: string;
    /** Deployment Id. */
    environmentId: string;
    /** Input attributes for applied app plugin. */
    plugins: PluginElementSnapshot[];
};
export type StartRunRequest = {
    /** Requestor's email. */
    requestedBy: string;
    /** Action to perform. */
    action: "CREATE" | "UPDATE" | "DELETE" | "RUN";
    /** Run ID. */
    runId: string;
    /** Workflow executionId. */
    executionId?: string;
    /** Environment id. */
    environmentId: string;
    /** Application id. */
    applicationId?: string;
    /** Shared Infra id. */
    sharedInfrastructureId?: string;
    /** Deploy type. */
    "type": "DEPLOY" | "ROLLBACK" | "DESTROY" | "DEPLOY_SELF_HOSTED" | "DESTROY_SELF_HOSTED" | "ROLLBACK_SELF_HOSTED";
    /** Pipeline URL for self hosted deployment. */
    pipelineLink?: string;
    /** Application Manifest. */
    appManifesto?: string;
    /** List of plugins that contain in deploy, rollback or destroy. */
    plugins: Plugins[];
};
export type WorkspaceLinkReadResponse = {
    /** Shared infra link id. */
    id: string;
    /** Workspace link name. */
    name: string;
    /** Workspace link url. */
    url: string;
    /** Workspace link image URL. */
    imageUrl?: string;
};
export type CreateWorkspaceLinkRequest = {
    /** Workspace link name. */
    name: string;
    /** Workspace link URL. */
    url: string;
    /** Workspace link image base64. */
    image?: string;
};
export type WorkspaceEmbeddedLinkReadResponse = {
    /** Workspace embedded link id. */
    id: string;
    /** Workspace embedded link name. */
    name: string;
    /** Workspace embedded link url. */
    url: string;
    /** Workspace embedded link image URL. */
    imageUrl?: string;
    /** Workspace embedded link layout. */
    layout?: object;
};
export type CreateWorkspaceEmbeddedLinkRequest = {
    /** Workspace embedded link name. */
    name: string;
    /** Workspace embedded link URL. */
    url: string;
    /** Workspace embedded link image base64. */
    image?: string;
};
export type ConnectionInterfaceSummaryResponse = {
    /** Connection interface id. */
    id: string;
    /** Connection interface id type generate by client. */
    typeId: string;
    /** Shared infra that was used to generated this Connection interface. */
    sharedInfraId?: string;
    /** Application that was used to generated this Connection interface. */
    applicationId?: string;
    /** Connection interface id generated by client. */
    connectionInterfaceId: string;
    /** Indicates if this connection interface was generated by a plugin or manually. */
    automaticallyGenerated: boolean;
    /** The version ID of the plugin that generated this. */
    sourcePluginVersionId?: string;
};
export type CreateConnectionInterfaceRequest = {
    /** Connection interface type id. */
    typeId: string;
    /** Connection interface id generated by client. */
    connectionInterfaceId: string;
    /** Connection interface owner applications. */
    applicationId?: string;
    /** Connection interface owner shared infra. */
    sharedInfraId?: string;
    /** Version id from plugin that generated this connection interface. */
    sourcePluginVersionId?: string;
    /** Indicates if this connection interface was generated by a plugin or manually. */
    automaticallyGenerated: boolean;
    /** Connection interface attributes. */
    attributesByEnv: ConnectionInterfaceAttributesByEnv[];
};
export type UpdateConnectionInterfaceVisibilityRequest = {
    ids: string[];
    "type": "ACCOUNT" | "APPLICATION" | "WORKSPACE";
};
export type CreateApplicationRequest = {
    /** Application name. */
    name: string;
    /** Application description. */
    description?: string;
    /** Application repository url. */
    repoUrl?: string;
    /** Application repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this application. */
    stackVersionId: string;
    /** Starter used to generate this application. */
    starterId: string;
    /** Extra repository this shared infrastructure. */
    extraRepositories?: ExtraRepositoriesRequest[];
};
export type RecreateApplicationRequest = {
    /** Application name. */
    name: string;
    /** Application description. */
    description?: string;
    /** Application repository url. */
    repoUrl?: string;
    /** Application repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this application. */
    stackVersionId: string;
    /** Starter used to generate this application. */
    starterId: string;
};
export type UpdateApplicationRequest = {
    /** Application name. */
    name?: string;
    /** Application description. */
    description?: string;
    /** Application repository url. */
    repoUrl?: string;
    /** Application repository base branch. */
    repoBaseBranch?: string;
    /** Extra repository for this application. */
    extraRepositories?: ExtraRepositoriesRequest[];
};
export type ApplicationLinkReadResponse = {
    /** Shared infra link id. */
    id: string;
    /** Application link name. */
    name: string;
    /** Application link url. */
    url: string;
    /** Application link image URL. */
    imageUrl?: string;
};
export type CreateApplicationLinkRequest = {
    /** Application link name. */
    name: string;
    /** Application link URL. */
    url: string;
    /** Application link image base64. */
    image?: string;
};
export type ApplicationDestroyRequest = {
    /** Destroy status. */
    status: "SUCCESS" | "FAILED";
};
export type MetaDataRequest = {
    name: string;
};
export type ConnectionInterfaceRequest = {
    /** Connection interface name */
    selected: string;
    /** Connection interface type */
    "type": string;
    /** Connection interface alias */
    alias: string;
};
export type ConnectionsRequest = {
    /** Connections interface requires per plugin */
    requires?: ConnectionInterfaceRequest[];
    /** Connections interface generates per plugin */
    generates?: ConnectionInterfaceRequest[];
};
export type GeneratesRequest = {
    /** Generated link name */
    name: string;
    /** Generated link url */
    url: string;
    /** Generated link type */
    "type": string;
    /** Generated link picture */
    picture: string;
};
export type PluginLinkRequest = {
    generates: GeneratesRequest[];
};
export type AppliedPluginsRequest = {
    /** Plugin name */
    name: string;
    /** Plugin alias */
    alias: string;
    /** Plugin type */
    "type": string;
    /** Plugin version id */
    pluginVersionId: string;
    /** Stack version id */
    stackVersionId: string;
    /** Plugin inputs */
    inputs: {
        [key: string]: object;
    };
    connections: ConnectionsRequest;
    links?: PluginLinkRequest;
};
export type SpecRequest = {
    /** Deploy specification id */
    id: string;
    /** Deploy specification repository */
    repository: string;
    /** Plugins applied in deploy  */
    appliedPlugins: AppliedPluginsRequest[];
};
export type AppManifestoRequest = {
    /** Deployment schema version */
    schemaVersion: string;
    /** Deployment kind */
    kind: string;
    metadata: MetaDataRequest;
    spec: SpecRequest;
};
export type ApplicationDeployRequest = {
    /** Deploy status. */
    status: "SUCCESS" | "FAILED";
    /** Deploy version. */
    version: string;
    appManifesto: AppManifestoRequest;
};
export type ApplicationEmbeddedLinkReadResponse = {
    /** Application embedded link id. */
    id: string;
    /** Application embedded link name. */
    name: string;
    /** Application embedded link url. */
    url: string;
    /** Application embedded link image URL. */
    imageUrl?: string;
    /** Application embedded link layout. */
    layout?: object;
};
export type CreateApplicationEmbeddedLinkRequest = {
    /** Application embedded link name. */
    name: string;
    /** Application embedded link URL. */
    url: string;
    /** Application embedded link image base64. */
    image?: string;
};
export type ApplicationDeploySnapshotRequest = {
    /** Deployment Id. */
    deploymentId: string;
    /** Deployment Id. */
    environmentId: string;
    /** Input attributes for applied app plugin. */
    plugins: PluginElementSnapshot[];
};
export type EnvironmentLinkedPluginResponse = {
    id: string;
    name: string;
    totalLinkedApps: number;
    totalLinkedInfra: number;
};
export type EnvironmentStackResponse = {
    stackVersionId: string;
    environments: EnvironmentLinkedPluginResponse[];
};
export type StackUsageByWorkspaceResponse = {
    /** Workspace id. */
    id: string;
    /** Workspace name */
    name: string;
    /** Workspace description */
    description?: string;
    /** Workspace image */
    imageUrl?: string;
    /** Amount of linked versions. */
    totalLinkedVersions: number;
    /** Stack Versions linked. */
    stackVersions: EnvironmentStackResponse[];
};
export type GetAllSharedInfraRequestBody = {
    favorites?: string[];
    name?: string;
    stackVersionIds?: string[];
};
export type SharedInfraWorkspaceReadResponse = {
    /** Application id. */
    id: string;
    /** Shared Infra name. */
    name: string;
    /** Shared Infra description. */
    description?: string;
    /** Shared Infra repository url. */
    repoUrl?: string;
    /** Shared Infra repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this Shared Infra. */
    stackVersionId: string;
    /** Starter used to generate this Shared Infra. */
    starterId?: string;
    /** Shared Infra status. */
    status: string;
    /** Shared Infra creator. */
    createdBy: string;
    /** Shared Infra creation date. */
    createdAt: string;
    /** Shared Infra updater. */
    updatedBy: string;
    /** Shared Infra update date. */
    updatedAt: string;
    workspace: WorkspaceReadResponse;
};
export type SharedInfraPageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: SharedInfraWorkspaceReadResponse[];
};
export type PluginUsageAmountResponse = {
    pluginVersionId: string;
    totalUsedInInfras: number;
    totalUsedInApps: number;
};
export type PluginUsageAmountConsolidatedResponse = {
    totalUsedInInfras: number;
    totalUsedInApps: number;
    details: PluginUsageAmountResponse[];
};
export type EnvironmentReadResponse = {
    /** Environment id. */
    id: string;
    /** Environment name. */
    name: string;
    /** Environment description. */
    description: string;
    /** Environment accountId. */
    accountId: string;
};
export type EnvironmentSaveRequest = {
    /** Environment name. */
    name: string;
    /** Environment description. */
    description: string;
};
export type EnvironmentReadBatchResponse = {
    /** Environment id. */
    id: string;
    /** Environment name. */
    name: string;
};
export type GetAllApplicationsRequestBody = {
    favorites?: string[];
    name?: string;
    stackVersionIds?: string[];
};
export type ApplicationWorkspaceReadResponse = {
    /** Application id. */
    id: string;
    /** Application name. */
    name: string;
    /** Application description. */
    description?: string;
    /** Application repository url. */
    repoUrl?: string;
    /** Application repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this application. */
    stackVersionId: string;
    /** Starter used to generate this application. */
    starterId: string;
    /** Application status. */
    status: string;
    /** Application creator. */
    createdBy: string;
    /** Application creation date. */
    createdAt: string;
    /** Application updater. */
    updatedBy: string;
    /** Application update date. */
    updatedAt: string;
    workspace: WorkspaceReadResponse;
};
export type ApplicationPageResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    totalPages: number;
    items: ApplicationWorkspaceReadResponse[];
};
export type WorkflowResponse = {
    id: string;
    name: string;
    description: string;
    "type": string;
    isDefault: boolean;
};
export type WorkflowActionWithContextRequest = {
    actionId: string;
    order: number;
    versionRange: string;
    stepType?: "CREATE_REPO";
    valuesContext: ContextAttribute[];
};
export type AccountWorkflowCreateRequest = {
    name: string;
    description: string;
    "type": "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
    isDefault: boolean;
    stacksId: string[];
    actionsBefore: WorkflowActionWithContextRequest[];
    actionsAfter: WorkflowActionWithContextRequest[];
};
export type PaginatedAccountVariableResponse = {
    /** Current page requested */
    currentPage: number;
    /** Last page available */
    lastPage: number;
    /** Page size requested */
    pageSize: number;
    /** Total items found */
    totalItems: number;
    /** Account variables for current page */
    items: AccountVariableResponse[];
};
export type CreateAccountVariableRequest = {
    /** Account variable name */
    name: string;
    /** Account variable value */
    value?: string;
    /** Account variable description */
    description: string;
    /** Account variable mandatory flag */
    mandate: boolean;
};
export type AccountStackContextResponse = {
    /** List of stack version ids with pluginVersionId list with account context. */
    stackVersions: StackVersionResponse[];
    /** Account id. */
    accountId: string;
};
export type UpdateWorkspaceRequest = {
    /** Workspace name. */
    name?: string;
    /** Workspace description. */
    description?: string;
    /** Workspace image base64. */
    image?: string;
    /** Workspace manage runtime. */
    manageRuntime?: boolean;
};
export type UpdateSharedInfraLinkRequest = {
    /** Shared infra link URL. */
    url?: string;
    /** Shared infra link image base64. */
    image?: string;
    name?: string;
};
export type UpdateWorkspaceLinkRequest = {
    /** Workspace link URL. */
    url?: string;
    /** Workspace link image base64. */
    image?: string;
    name?: string;
};
export type UpdateWorkspaceEmbeddedLinkRequest = {
    /** Workspace embedded link name. */
    name?: string;
    /** Workspace embedded link URL. */
    url?: string;
    /** Workspace embedded link image base64. */
    image?: string;
};
export type UpsertWorkspaceEmbeddedLinkRequest = {
    /** Workspace embedded link id. */
    id: string;
    /** Workspace embedded link layout. */
    layout: object;
};
export type UpdateApplicationLinkRequest = {
    /** Shared infra link URL. */
    url?: string;
    /** Shared infra link image base64. */
    image?: string;
    name?: string;
};
export type UpdateApplicationEmbeddedLinkRequest = {
    /** Application embedded link name. */
    name?: string;
    /** Application embedded link URL. */
    url?: string;
    /** Application embedded link image base64. */
    image?: string;
};
export type UpsertApplicationEmbeddedLinkRequest = {
    /** Application embedded link id. */
    id: string;
    /** Application embedded link layout. */
    layout: object;
};
export type EnvironmentUpdateRequest = {
    /** Environment name. */
    name?: string;
    /** Environment description. */
    description?: string;
};
export type PaginatedWorkspaceVariableResponse = {
    currentPage: number;
    pageSize: number;
    lastPage: number;
    totalItems: number;
    items: WorkspaceVariableResponse[];
};
export type AppliedConnectionInterface = {
    slug: string;
    "type": string;
    alias: string;
};
export type ConnectionsResponse = {
    requires?: AppliedConnectionInterface[];
    generates?: AppliedConnectionInterface[];
};
export type Link = {
    name: string;
    url: string;
    "type": string;
    picture: string;
};
export type AppliedPluginV2Response = {
    name: string;
    "type"?: string;
    alias: string;
    pluginVersionId: string;
    stackVersionId: string;
    status?: "PENDING" | "READY_TO_RUN" | "RUNNING" | "SUCCEEDED" | "FAILED" | "INTERNAL_ERROR" | "ABORTED" | "EXTERNAL_ERROR" | "DESTROYED";
    runType?: "DEPLOY" | "DEPLOY_SELF_HOSTED" | "DESTROY" | "DESTROY_SELF_HOSTED" | "IAC" | "IAC_SELF_HOSTED";
    inCloud?: boolean;
    inputs?: {
        [key: string]: object;
    };
    connections?: ConnectionsResponse;
    links?: Link[];
};
export type ConnectionInterfaceSummaryV2Response = {
    /** Connection interface api id. */
    apiId: string;
    /** Connection interface id type generate by client. */
    typeId: string;
    /** Connection interface id generated by client. */
    slug: string;
    /** Indicates if this connection interface was generated by a plugin or manually. */
    automaticallyGenerated: boolean;
    /** The version ID of the plugin that generated this. */
    sourcePluginVersionId?: string;
};
export type Visibility = {
    /** Connection visibility type. */
    "type": "ACCOUNT" | "APPLICATION" | "WORKSPACE";
    /** IDs (based on type) for which connection interface is visible. */
    ids: string[];
};
export type ConnectionInterfaceDetailsV2Response = {
    /** Connection interface type. */
    typeId: string;
    /** Connection interface id generated by client. */
    slug: string;
    visibility: Visibility;
    /** Version id from plugin that generated this connection interface. */
    sourcePluginVersionId?: string;
    /** Indicates if this connection interface was generated by a plugin or manually. */
    automaticallyGenerated: boolean;
    /** Connection interface attributes. */
    attributesByEnv: ConnectionInterfaceAttributesByEnv[];
};
export type PluginInUseResponse = {
    name: string;
    alias: string;
    pluginVersionId: string;
    stackVersionId: string;
    inputs?: {
        [key: string]: object;
    };
    connections?: ConnectionsResponse;
};
export type ConnectionInterfaceInUseByTargetIdDetailsResponse = {
    environmentId: string;
    targets: PluginInUseResponse[];
};
export type WorkspaceVariableReadV1Response = {
    /** Workspace variable name. */
    name: string;
    /** Workspace variable value. */
    value: string;
    /** Workspace variable description. */
    description: string;
};
export type ConsolidatedAttr = {
    key: string;
    value: object;
    mandate: boolean;
    source: string;
};
export type ConsolidatedAttrByEnv = {
    environmentId: string;
    envName: string;
    attributes: ConsolidatedAttr[];
};
export type ConsolidatedContextResponse = {
    externalId: string;
    attributesByEnv: ConsolidatedAttrByEnv[];
};
export type AccountWorkflowResponse = {
    workflowId: string;
    name: string;
    description: string;
    isDefault: boolean;
    "type"?: "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
    actionsBefore: WorkflowActionResponse[];
    actionsAfter: WorkflowActionResponse[];
};
export type AppliedAttribute = {
    key: string;
    value: string;
};
export type AppliedPluginResponse = {
    pluginVersionId: string;
    "type": string;
    attributes: AppliedAttribute[];
    connectionInterfaces: AppliedConnectionInterface[];
};
export type ActivityResponse = {
    title?: string;
    version?: string;
    owner?: string;
    status: "SUCCESS" | "FAILED" | "SKIPPED" | "CANCELLED" | "RUNTIME_ERROR" | "INTERNAL_ERROR" | "EXTERNAL_ERROR" | "ABORTED" | "ABORTING" | "SUCCEEDED" | "USER_ERROR" | "DRIFT" | "TIMEOUT" | "RUNNING" | "READY_TO_RUN" | "PENDING" | "NONE";
    "type": "DEPLOY" | "ACTION" | "WORKFLOW" | "DRIFT" | "DESTROY" | "ROLLBACK" | "IAC" | "DEPLOY_SELF_HOSTED" | "DESTROY_SELF_HOSTED" | "ROLLBACK_SELF_HOSTED" | "IAC_SELF_HOSTED" | "CUSTOMER_WORKFLOW";
    requestedBy?: string;
    action?: "CREATE" | "UPDATE" | "DELETE" | "RUN";
    time: number;
    detail: {
        [key: string]: object;
    };
};
export type PaginatedActivityResponse = {
    items: ActivityResponse[];
    currentPage: number;
    totalItems: number;
    totalPages: number;
    pageSize: number;
    hasNextPage: boolean;
};
export type AppDependencyConnInterfaceResponse = {
    name: string;
    "type": string;
};
export type Dependencies = {
    environmentName: string;
    hasSelfHostedInfra: boolean;
    hasApplication: boolean;
    connectionInterfaces: AppDependencyConnInterfaceResponse[];
};
export type AppAndInfraDependencyResponse = {
    canDelete: boolean;
    hasSelfHostedInfra: boolean;
    hasApplication: boolean;
    dependencies: Dependencies[];
};
export type StacksVersionIdsResponse = {
    stackVersionIds: string[];
};
export type ConnectionInterfaceDetailsResponse = {
    /** Connection interface id. */
    id: string;
    /** Connection interface type id. */
    typeId: string;
    /** Connection interface id generated by client. */
    connectionInterfaceId: string;
    visibility: Visibility;
    /** Version id from plugin that generated this connection interface. */
    sourcePluginVersionId?: string;
    /** Indicates if this connection interface was generated by a plugin or manually. */
    automaticallyGenerated: boolean;
    /** Connection interface attributes. */
    attributesByEnv: ConnectionInterfaceAttributesByEnv[];
};
export type DeploymentTargetsResponse = {
    targetId: string;
    targetType?: "APPLICATION" | "SHARED_INFRA";
    environmentId: string;
};
export type ConnectionInterfaceInUseDetailsResponse = {
    slug: string;
    targets: DeploymentTargetsResponse[];
};
export type ConnectionInterfaceInUseSummaryResponse = {
    targetId: string;
    targetType?: "APPLICATION" | "SHARED_INFRA";
    environmentId: string;
};
export type MetaDataResponse = {
    name: string;
};
export type SpecResponse = {
    id?: string;
    repository?: string;
    appliedPlugins: AppliedPluginV2Response[];
};
export type ManifestoResponse = {
    schemaVersion: string;
    kind: string;
    metadata: MetaDataResponse;
    spec: SpecResponse;
};
export type AppDeployResponse = {
    deploymentId: string;
    targetId: string;
    targetType: "APPLICATION" | "SHARED_INFRA";
    targetVersion?: string;
    workspaceId: string;
    accountId: string;
    environmentId: string;
    status: "ACTIVE" | "INACTIVE";
    deployStatus: "SUCCEEDED" | "FAILED";
    requestedBy: string;
    dateTime: string;
    manifesto: ManifestoResponse;
};
export type InfraDeployResponse = {
    deploymentId: string;
    targetId: string;
    targetType?: "APPLICATION" | "SHARED_INFRA";
    targetVersion?: string;
    workspaceId: string;
    accountId: string;
    environmentId: string;
    "type"?: "DEPLOY" | "ROLLBACK" | "DESTROY" | "IAC" | "DEPLOY_SELF_HOSTED" | "DESTROY_SELF_HOSTED" | "ROLLBACK_SELF_HOSTED" | "IAC_SELF_HOSTED";
    status: "ACTIVE" | "INACTIVE";
    action?: "CREATE" | "UPDATE" | "DELETE" | "RUN";
    requestedBy?: string;
    manifesto: ManifestoResponse;
};
export type ApplicationDeployInfoResponse = {
    appDeploy?: AppDeployResponse;
    infraDeploy?: InfraDeployResponse;
};
export type EnvironmentsRequest = {
    environmentId: string;
    environmentName?: string;
    hasSaasDeploy: boolean;
    hasSelHostedDeploy: boolean;
    hasActiveApplicationDeploy: boolean;
};
export type ApplicationCanBeDeletedResponse = {
    targetId: string;
    targetType: "APPLICATION" | "SHARED_INFRA";
    environments: EnvironmentsRequest[];
};
export type WorkspaceInfo = {
    /** Workspace id. */
    id: string;
    /** Workspace name */
    name: string;
};
export type GetSharedInfraResponse = {
    /** Shared infrastructure id. */
    id: string;
    /** Shared infrastructure name. */
    name: string;
    /** Shared infrastructure description. */
    description?: string;
    /** Shared infrastructure repository url. */
    repoUrl?: string;
    /** Shared infrastructure repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this shared infrastructure. */
    stackVersionId: string;
    workspace: WorkspaceInfo;
};
export type PluginInUseDetailsResponse = {
    workspaceId: string;
    targets: DeploymentTargetsResponse[];
};
export type CheckConnectionSlugAvailabilityResponse = {
    /** Whether connection slug is available or not. */
    available: boolean;
};
export type ApplicationRepoUrlReadResponse = {
    /** Application id. */
    id: string;
    /** Application name. */
    name: string;
    /** Application status. */
    status: "CREATED" | "CREATE_FAILED" | "DRAFT" | "DELETING" | "DELETE_FAILED" | "ARCHIVING" | "ARCHIVED" | "ARCHIVE_FAILED";
    /** Application description. */
    description?: string;
    workspace?: WorkspaceResponse;
};
export type GetApplicationResponse = {
    /** Application id. */
    id: string;
    /** Application name. */
    name: string;
    /** Application description. */
    description?: string;
    /** Application repository url. */
    repoUrl?: string;
    /** Application repository base branch. */
    repoBaseBranch: string;
    /** Stack used to generate this application. */
    stackVersionId: string;
    /** Starter used to generate this application. */
    starterId: string;
    workspace: WorkspaceInfo;
};
export type WorkflowActionUsageResponse = {
    versionRanges: string[];
};
export type AccountVariableUsageResponse = {
    /** How many plugins are using this account variable */
    pluginsInUse: number;
    /** How many actions are using this account variable */
    actionsInUse: number;
    /** How many contexts are using this account variable */
    contextsInUse: number;
};
export type AccountAttr = {
    key: string;
    value: object;
    mandate: boolean;
};
export type AccountAttrByEnv = {
    environmentId: string;
    envName: string;
    attributes: AccountAttr[];
};
export type AccountContextResponse = {
    externalId: string;
    attributesByEnv: AccountAttrByEnv[];
};
export type WorkflowSimpleResponse = {
    id: string;
    name: string;
    "type": string;
};
export type StackWorkflowResponse = {
    id: string;
    workflows: WorkflowSimpleResponse[];
};
export type StacksByAccountResponse = {
    accountId: string;
    stacks: StackWorkflowResponse[];
};
/**
 * Get a workspace variable by name
 */
export function workspaceVariableV2ControllerfindByName({ accountId, workspaceId, name }: {
    accountId?: string;
    workspaceId: string;
    name: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceVariableResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/variables/${encodeURIComponent(name)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Upsert workspace variable based on existing account variable
 */
export function workspaceVariableV2Controllerupsert({ workspaceId, name, upsertWorkspaceVariableRequest }: {
    workspaceId: string;
    name: string;
    upsertWorkspaceVariableRequest: UpsertWorkspaceVariableRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/variables/${encodeURIComponent(name)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: upsertWorkspaceVariableRequest
    })));
}
/**
 * Add context to a typed element in a workspace's stack.
 */
export function contextV2ControlleraddTypedContextInWorkspace({ workspaceId, stackVersionId, externalId, $type, contextValuesRequest }: {
    workspaceId: string;
    stackVersionId: string;
    externalId: string;
    $type: "plugin" | "action";
    contextValuesRequest: ContextValuesRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/${encodeURIComponent($type)}s/${encodeURIComponent(externalId)}/input-context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: contextValuesRequest
    })));
}
/**
 * Import context in a workspace's stack.
 */
export function contextV2ControllerimportContextInWorkspace({ workspaceId, stackVersionId, importContextV2Request }: {
    workspaceId: string;
    stackVersionId: string;
    importContextV2Request: ImportContextV2Request;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/input-context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: importContextV2Request
    })));
}
/**
 * Add context to a typed element in an account's stack.
 */
export function contextV2ControlleraddTypedContextInAccount({ stackVersionId, externalId, $type, contextValuesRequest }: {
    stackVersionId: string;
    externalId: string;
    $type: "plugin" | "action";
    contextValuesRequest: ContextValuesRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/account/stacks/${encodeURIComponent(stackVersionId)}/${encodeURIComponent($type)}s/${encodeURIComponent(externalId)}/input-context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: contextValuesRequest
    })));
}
/**
 * Import context in an account's stack.
 */
export function contextV2ControllerimportContextInAccount({ stackVersionId, importContextV2Request }: {
    stackVersionId: string;
    importContextV2Request: ImportContextV2Request;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/account/stacks/${encodeURIComponent(stackVersionId)}/input-context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: importContextV2Request
    })));
}
/**
 * List all configured context of an workflow by workspace and stack.
 */
export function workflowWorkspaceControllerlistContextByWorkspaceStack({ workspaceId, workflowId, stackId }: {
    workspaceId: string;
    workflowId: string;
    stackId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowContextResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/workflows/${encodeURIComponent(workflowId)}/stacks/${encodeURIComponent(stackId)}/context`, {
        ...opts
    }));
}
/**
 * Save workspace workflow context
 */
export function workflowWorkspaceControllersaveWorkspaceWorkflowContext({ workspaceId, workflowId, stackId, workspaceWorkflowCreateRequest }: {
    workspaceId: string;
    workflowId: string;
    stackId: string;
    workspaceWorkflowCreateRequest: WorkspaceWorkflowCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/workflows/${encodeURIComponent(workflowId)}/stacks/${encodeURIComponent(stackId)}/context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: workspaceWorkflowCreateRequest
    })));
}
/**
 * Get Workflow of a stack in a workspace.
 */
export function workflowControllergetWorkspaceWorkflow({ workspaceId, stackVersionId, identifier }: {
    workspaceId: string;
    stackVersionId: string;
    identifier?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowReadResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/workflows${QS.query(QS.explode({
        identifier
    }))}`, {
        ...opts
    }));
}
/**
 * Add step to a workflow belonging to a workspace.
 */
export function workflowControlleraddWorkflowStepInWorkspace({ workspaceId, stackVersionId, addWorkspaceWorkflowStepRequest }: {
    workspaceId: string;
    stackVersionId: string;
    addWorkspaceWorkflowStepRequest: AddWorkspaceWorkflowStepRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/workflows`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: addWorkspaceWorkflowStepRequest
    })));
}
/**
 * Update shared infrastructure status.
 */
export function workspaceSharedInfrastructureControllerupdateStatus({ workspaceId, sharedInfraId, updateSharedInfraStatusRequest }: {
    workspaceId: string;
    sharedInfraId: string;
    updateSharedInfraStatusRequest: UpdateSharedInfraStatusRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/status`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateSharedInfraStatusRequest
    })));
}
/**
 * Get shared infra links visibility
 */
export function sharedInfraLinkControllergetVisibleLinks({ workspaceId, sharedInfraId }: {
    workspaceId: string;
    sharedInfraId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraVisibleLinksResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/links/visibility-management`, {
        ...opts
    }));
}
/**
 * Update application links visibility
 */
export function sharedInfraLinkControllerupdateLinksVisibility({ workspaceId, sharedInfraId, body }: {
    workspaceId: string;
    sharedInfraId: string;
    body: UpdateSharedInfraLinkVisibilityRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/links/visibility-management`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Update activity status by workspace.
 */
export function managerRunControllerupdateRun({ accountId, workspaceId, runId, updateRunRequest }: {
    accountId: string;
    workspaceId: string;
    runId: string;
    updateRunRequest: UpdateRunRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/run/${encodeURIComponent(runId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateRunRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Update plugin status by workspace.
 */
export function managerRunControllerrunTask({ accountId, workspaceId, runId, runTaskRequest }: {
    accountId: string;
    workspaceId: string;
    runId: string;
    runTaskRequest: RunTaskRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/run/${encodeURIComponent(runId)}/plugin`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: runTaskRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Deployment tag register.
 */
export function managerRunControllerdeploymentTag({ accountId, workspaceId, runId, deploymentTagRequest }: {
    accountId: string;
    workspaceId: string;
    runId: string;
    deploymentTagRequest: DeploymentTagRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/run/${encodeURIComponent(runId)}/deploy-tag`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: deploymentTagRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Request to upsert connection interfaces attributes.
 */
export function connectionInterfaceControllerupdateConnectionInterfaceAttributes({ workspaceId, connectionInterfaceId, updateConnectionInterfaceAttributesRequest }: {
    workspaceId: string;
    connectionInterfaceId: string;
    updateConnectionInterfaceAttributesRequest: UpdateConnectionInterfaceAttributesRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/${encodeURIComponent(connectionInterfaceId)}/attributes`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateConnectionInterfaceAttributesRequest
    })));
}
/**
 * Request to upsert connection interfaces.
 */
export function connectionInterfaceControllerupsertBatch({ workspaceId, body }: {
    workspaceId: string;
    body: UpsertConnectionInterfaceRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/batch`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Update application status.
 */
export function workspaceApplicationControllerupdateStatus({ workspaceId, applicationId, accountId, updateApplicationStatusRequest }: {
    workspaceId: string;
    applicationId: string;
    accountId?: string;
    updateApplicationStatusRequest: UpdateApplicationStatusRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/status`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateApplicationStatusRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Get application links visibility
 */
export function applicationLinkControllergetVisibleLinks({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationVisibleLinksResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/links/visibility-management`, {
        ...opts
    }));
}
/**
 * Update application links visibility
 */
export function applicationLinkControllerupdateLinksVisibility({ workspaceId, applicationId, body }: {
    workspaceId: string;
    applicationId: string;
    body: UpdateApplicationLinkVisibilityRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/links/visibility-management`, oazapfts.json({
        ...opts,
        method: "PUT",
        body
    })));
}
/**
 * Get workflow settings with associated stacks.
 */
export function workflowAccountControllerlistSettingsByAccountWorkflow({ workflowId }: {
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowSettingsResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/settings`, {
        ...opts
    }));
}
/**
 * Update settings information of a workflow account.
 */
export function workflowAccountControllerupdateSettingsByAccountWorkflow({ workflowId, settingsWorkflowSaveRequest }: {
    workflowId: string;
    settingsWorkflowSaveRequest: SettingsWorkflowSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/settings`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: settingsWorkflowSaveRequest
    })));
}
/**
 * List all configured context of an workflow account.
 */
export function workflowAccountControllerlistContextByAccountWorkflow({ workflowId }: {
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowContextResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/context`, {
        ...opts
    }));
}
/**
 * Update all context configurations of a workflow account.
 */
export function workflowAccountControllerupdateContextsByAccountWorkflow({ workflowId, contextsWorkflowSaveRequest }: {
    workflowId: string;
    contextsWorkflowSaveRequest: ContextsWorkflowSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: contextsWorkflowSaveRequest
    })));
}
/**
 * List all actions of an workflow account.
 */
export function workflowAccountControllerlistActionsByAccountWorkflow({ workflowId }: {
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowActionsResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/actions`, {
        ...opts
    }));
}
/**
 * Update all actions associations with a workflow account.
 */
export function workflowAccountControllerupdateActionsByAccountWorkflow({ workflowId, actionsWorkflowSaveRequest }: {
    workflowId: string;
    actionsWorkflowSaveRequest: ActionsWorkflowSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}/actions`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: actionsWorkflowSaveRequest
    })));
}
/**
 * Get an account variable by name
 */
export function accountVariableControllerfindByName({ accountId, name }: {
    accountId?: string;
    name: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountVariableResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/variables/${encodeURIComponent(name)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Update account variable
 */
export function accountVariableControllerupdate({ name, updateAccountVariableRequest }: {
    name: string;
    updateAccountVariableRequest: UpdateAccountVariableRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/variables/${encodeURIComponent(name)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: updateAccountVariableRequest
    })));
}
/**
 * Delete account variable
 */
export function accountVariableControllerdelete({ name }: {
    name: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/variables/${encodeURIComponent(name)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get Workflow of a stack in an account.
 */
export function workflowControllergetAccountWorkflow({ stackVersionId, identifier }: {
    stackVersionId: string;
    identifier?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowReadResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/stacks/${encodeURIComponent(stackVersionId)}/workflows${QS.query(QS.explode({
        identifier
    }))}`, {
        ...opts
    }));
}
/**
 * Add step to a workflow belonging to an account.
 */
export function workflowControlleraddWorkflowStepInAccount({ stackVersionId, addAccountWorkflowStepRequest }: {
    stackVersionId: string;
    addAccountWorkflowStepRequest: AddAccountWorkflowStepRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/stacks/${encodeURIComponent(stackVersionId)}/workflows`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: addAccountWorkflowStepRequest
    })));
}
/**
 * Save stack workflow context
 */
export function workflowStackControllersaveStackWorkflowContext({ stackId, workflowId, stackWorkflowCreateRequest }: {
    stackId: string;
    workflowId: string;
    stackWorkflowCreateRequest: StackWorkflowCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/stacks/${encodeURIComponent(stackId)}/workflows/${encodeURIComponent(workflowId)}/context`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: stackWorkflowCreateRequest
    })));
}
/**
 * Get all workspaces without stack version and favorites filter
 */
export function workspaceV2ControllergetWorkspacesWithoutStackFilter({ name, aclOnly, accountId, page, size, sortBy, sortDir }: {
    name?: string;
    aclOnly?: boolean;
    accountId?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "DESCRIPTION";
    sortDir?: "ASC" | "DESC";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspacePageResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces${QS.query(QS.explode({
        name,
        aclOnly,
        page,
        size,
        sortBy,
        sortDir
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get all workspaces
 */
export function workspaceV2ControllergetWorkspaces({ aclOnly, accountId, page, size, sortBy, sortDir, getWorkspacePageRequest }: {
    aclOnly?: boolean;
    accountId?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "DESCRIPTION";
    sortDir?: "ASC" | "DESC";
    getWorkspacePageRequest?: GetWorkspacePageRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspacePageResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces${QS.query(QS.explode({
        aclOnly,
        page,
        size,
        sortBy,
        sortDir
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: getWorkspacePageRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Get shared infra page.
 */
export function workspaceSharedInfrastructureV2ControllergetSharedInfra({ workspaceId, page, size, sortBy, sortDir, getSharedInfraRequestBody }: {
    workspaceId: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "CREATED_AT" | "STACK_VERSION" | "STACK";
    sortDir?: "ASC" | "DESC";
    getSharedInfraRequestBody?: GetSharedInfraRequestBody;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceSharedInfraPageResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/shared-infra${QS.query(QS.explode({
        page,
        size,
        sortBy,
        sortDir
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: getSharedInfraRequestBody
    })));
}
/**
 * Get applications from a workspace with pagination.
 */
export function workspaceApplicationV2ControllergetApplications({ workspaceId, page, size, sortBy, sortDir, getApplicationsRequestBody }: {
    workspaceId: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "CREATED_AT" | "STACK_VERSION" | "STACK";
    sortDir?: "ASC" | "DESC";
    getApplicationsRequestBody?: GetApplicationsRequestBody;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceApplicationPageResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/applications${QS.query(QS.explode({
        page,
        size,
        sortBy,
        sortDir
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: getApplicationsRequestBody
    })));
}
/**
 * List specific shared infra or application usage by plugin list.
 */
export function usageInsightsControllerlistPluginUsageByTargetId({ accountId, targetType, targetId, body }: {
    accountId: string;
    targetType: "applications" | "shared_infra";
    targetId: string;
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: TargetByPluginUsageResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    }>(`/v1/${encodeURIComponent(targetType)}/${encodeURIComponent(targetId)}/plugins/usage`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * List all shared infras or applications usage by plugin list.
 */
export function usageInsightsControllerlistAllPluginUsage({ accountId, targetType, body }: {
    accountId: string;
    targetType: "applications" | "shared_infra";
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PluginUsageByTargetResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    }>(`/v1/${encodeURIComponent(targetType)}/plugins/usage`, oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Get all workspaces
 */
export function workspaceControllergetWorkspaces({ name, aclOnly, accountId }: {
    name?: string;
    aclOnly?: boolean;
    accountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces${QS.query(QS.explode({
        name,
        aclOnly
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Create workspace
 */
export function workspaceControllersave({ newWorkspaceRequest }: {
    newWorkspaceRequest: NewWorkspaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/workspaces", oazapfts.json({
        ...opts,
        method: "POST",
        body: newWorkspaceRequest
    })));
}
/**
 * Get all stacks in a workspace.
 */
export function workspaceStackControllergetStacks({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackInWorkspaceReadResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks`, {
        ...opts
    }));
}
/**
 * Add a stack in a workspace.
 */
export function workspaceStackControlleraddStack({ workspaceId, addStackInWorkspaceRequest }: {
    workspaceId: string;
    addStackInWorkspaceRequest: AddStackInWorkspaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addStackInWorkspaceRequest
    })));
}
/**
 * Get all stacks with context in a workspace.
 */
export function workspaceStackControllergetStacksWithContext({ workspaceId, body }: {
    workspaceId: string;
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackWithWorkspaceContextResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/with-context`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Add a stack version in a workspace.
 */
export function workspaceStackControlleraddUpdateStack({ workspaceId, addStackInWorkspaceRequest }: {
    workspaceId: string;
    addStackInWorkspaceRequest: AddStackInWorkspaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/-/update`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addStackInWorkspaceRequest
    })));
}
/**
 * Get all shared infrastructure of a workspace.
 */
export function workspaceSharedInfrastructureControllergetAllSharedInfrastructure({ workspaceId, name, stackVersionId }: {
    workspaceId: string;
    name?: string;
    stackVersionId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraDetailsResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra${QS.query(QS.explode({
        name,
        stackVersionId
    }))}`, {
        ...opts
    }));
}
/**
 * Create shared infrastructure in a workspace.
 */
export function workspaceSharedInfrastructureControllersave({ workspaceId, fromPortal, createSharedInfraRequest }: {
    workspaceId: string;
    fromPortal?: boolean;
    createSharedInfraRequest: CreateSharedInfraRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra${QS.query(QS.explode({
        fromPortal
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createSharedInfraRequest
    })));
}
/**
 * Get shared infrastructure of a workspace.
 */
export function workspaceSharedInfrastructureControllergetSharedInfrastructure({ workspaceId, sharedInfraId }: {
    workspaceId: string;
    sharedInfraId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraDetailsResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}`, {
        ...opts
    }));
}
/**
 * Recreate a shared infrastructure in a workspace.
 */
export function workspaceSharedInfrastructureControllerrecreate({ workspaceId, sharedInfraId, recreateSharedInfraRequest }: {
    workspaceId: string;
    sharedInfraId: string;
    recreateSharedInfraRequest: RecreateSharedInfraRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: recreateSharedInfraRequest
    })));
}
/**
 * Delete infrastructure in a workspace.
 */
export function workspaceSharedInfrastructureControllerdeleteSharedInfra({ workspaceId, sharedInfraId }: {
    workspaceId: string;
    sharedInfraId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update shared infrastructure in a workspace.
 */
export function workspaceSharedInfrastructureControllerupdate({ workspaceId, sharedInfraId, updateSharedInfraRequest }: {
    workspaceId: string;
    sharedInfraId: string;
    updateSharedInfraRequest: UpdateSharedInfraRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateSharedInfraRequest
    })));
}
/**
 * Get all shared infra links
 */
export function sharedInfraLinkControllergetSharedInfraLinks({ workspaceId, sharedInfraId, name }: {
    workspaceId: string;
    sharedInfraId: string;
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraLinkReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/links${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Create shared infra link
 */
export function sharedInfraLinkControllersave({ workspaceId, sharedInfraId, createSharedInfraLinkRequest }: {
    workspaceId: string;
    sharedInfraId: string;
    createSharedInfraLinkRequest: CreateSharedInfraLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/links`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createSharedInfraLinkRequest
    })));
}
/**
 * Has active deploy with type self hosted?
 */
export function workspaceSharedInfrastructureControllerhasActiveDeployWithTypeSelfHosted({ workspaceId, sharedInfraId }: {
    workspaceId: string;
    sharedInfraId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/has-self-hosted-deploy`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Can the shared infra be destroyed?
 */
export function workspaceSharedInfrastructureControllercanBeDestroyed({ workspaceId, sharedInfraId, environmentId, accountId }: {
    workspaceId: string;
    sharedInfraId: string;
    environmentId: string;
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/environments/${encodeURIComponent(environmentId)}/can-be-destroyed`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Register the snapshot of a shared infrastructure's deploy.
 */
export function workspaceSharedInfrastructureControllerprocessDeploySnapshot({ workspaceId, sharedInfraId, sharedInfraDeploySnapshotRequest }: {
    workspaceId: string;
    sharedInfraId: string;
    sharedInfraDeploySnapshotRequest: SharedInfraDeploySnapshotRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/deploy`, oazapfts.json({
        ...opts,
        method: "POST",
        body: sharedInfraDeploySnapshotRequest
    })));
}
/**
 * Runs an orchestration in a workspace.
 */
export function managerRunControllerstartRun({ accountId, workspaceId, startRunRequest }: {
    accountId: string;
    workspaceId: string;
    startRunRequest: StartRunRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 201;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/run`, oazapfts.json({
        ...opts,
        method: "POST",
        body: startRunRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Get all workspace links
 */
export function workspaceLinkControllergetWorkspaceLinks({ workspaceId, name }: {
    workspaceId: string;
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceLinkReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/links${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Create workspace
 */
export function workspaceLinkControllersave({ workspaceId, createWorkspaceLinkRequest }: {
    workspaceId: string;
    createWorkspaceLinkRequest: CreateWorkspaceLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/links`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createWorkspaceLinkRequest
    })));
}
/**
 * Get all workspace embedded links
 */
export function workspaceEmbeddedLinkControllergetEmbeddedLinks({ workspaceId, name }: {
    workspaceId: string;
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceEmbeddedLinkReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/embedded-links${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Create workspace embedded link
 */
export function workspaceEmbeddedLinkControllersave({ workspaceId, createWorkspaceEmbeddedLinkRequest }: {
    workspaceId: string;
    createWorkspaceEmbeddedLinkRequest: CreateWorkspaceEmbeddedLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/embedded-links`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createWorkspaceEmbeddedLinkRequest
    })));
}
/**
 * List all connection interfaces of a workspace.
 */
export function connectionInterfaceControllergetConnectionInterfaces({ workspaceId, typeId }: {
    workspaceId: string;
    typeId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceSummaryResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces${QS.query(QS.explode({
        typeId
    }))}`, {
        ...opts
    }));
}
/**
 * Create a connection interface in a workspace.
 */
export function connectionInterfaceControlleradd({ workspaceId, createConnectionInterfaceRequest }: {
    workspaceId: string;
    createConnectionInterfaceRequest: CreateConnectionInterfaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createConnectionInterfaceRequest
    })));
}
/**
 * Update the visibility of a connection interface.
 */
export function connectionInterfaceControllerupdateConnectionInterfaceVisibility({ workspaceId, connectionInterfaceId, updateConnectionInterfaceVisibilityRequest }: {
    workspaceId: string;
    connectionInterfaceId: string;
    updateConnectionInterfaceVisibilityRequest: UpdateConnectionInterfaceVisibilityRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/${encodeURIComponent(connectionInterfaceId)}/visibility/batch`, oazapfts.json({
        ...opts,
        method: "POST",
        body: updateConnectionInterfaceVisibilityRequest
    })));
}
/**
 * Get all application of a workspace.
 */
export function workspaceApplicationControllergetApplications({ workspaceId, name, stackVersionId }: {
    workspaceId: string;
    name?: string;
    stackVersionId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationReadResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications${QS.query(QS.explode({
        name,
        stackVersionId
    }))}`, {
        ...opts
    }));
}
/**
 * Create application on workspace.
 */
export function workspaceApplicationControllersave({ workspaceId, fromPortal, createApplicationRequest }: {
    workspaceId: string;
    fromPortal?: boolean;
    createApplicationRequest: CreateApplicationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications${QS.query(QS.explode({
        fromPortal
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createApplicationRequest
    })));
}
/**
 * Get application of a workspace.
 */
export function workspaceApplicationControllergetApplication({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationReadResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}`, {
        ...opts
    }));
}
/**
 * Recreate an application in a workspace.
 */
export function workspaceApplicationControllerrecreate({ workspaceId, applicationId, recreateApplicationRequest }: {
    workspaceId: string;
    applicationId: string;
    recreateApplicationRequest: RecreateApplicationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: recreateApplicationRequest
    })));
}
/**
 * Delete application of a workspace.
 */
export function workspaceApplicationControllerdeleteApplication({ workspaceId, applicationId, accountId }: {
    workspaceId: string;
    applicationId: string;
    accountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Update an application of a workspace.
 */
export function workspaceApplicationControllerupdate({ workspaceId, applicationId, updateApplicationRequest }: {
    workspaceId: string;
    applicationId: string;
    updateApplicationRequest: UpdateApplicationRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateApplicationRequest
    })));
}
/**
 * Get all application links
 */
export function applicationLinkControllergetApplicationLinks({ workspaceId, applicationId, name }: {
    workspaceId: string;
    applicationId: string;
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationLinkReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/links${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Create application link
 */
export function applicationLinkControllersave({ workspaceId, applicationId, createApplicationLinkRequest }: {
    workspaceId: string;
    applicationId: string;
    createApplicationLinkRequest: CreateApplicationLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/links`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createApplicationLinkRequest
    })));
}
/**
 * Has active deploy with type self hosted?
 */
export function workspaceApplicationControllerhasActiveDeployWithTypeSelfHosted({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/has-self-hosted-deploy`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Can the application be destroyed?
 */
export function workspaceApplicationControllercanBeDestroyed({ workspaceId, applicationId, environmentId, accountId }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(environmentId)}/can-be-destroyed`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
export function workspaceApplicationControllerregistryAppDestroy({ workspaceId, applicationId, environmentId, applicationDestroyRequest }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    applicationDestroyRequest: ApplicationDestroyRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(environmentId)}/app-destroy`, oazapfts.json({
        ...opts,
        method: "POST",
        body: applicationDestroyRequest
    })));
}
export function workspaceApplicationControllerregistryAppDeploy({ workspaceId, applicationId, environmentId, applicationDeployRequest }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    applicationDeployRequest: ApplicationDeployRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchText(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(environmentId)}/app-deploy`, oazapfts.json({
        ...opts,
        method: "POST",
        body: applicationDeployRequest
    })));
}
/**
 * Get all application embedded links
 */
export function applicationEmbeddedLinkControllergetEmbeddedLinks({ workspaceId, applicationId, name }: {
    workspaceId: string;
    applicationId: string;
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationEmbeddedLinkReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/embedded-links${QS.query(QS.explode({
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Create application embedded link
 */
export function applicationEmbeddedLinkControllersave({ workspaceId, applicationId, createApplicationEmbeddedLinkRequest }: {
    workspaceId: string;
    applicationId: string;
    createApplicationEmbeddedLinkRequest: CreateApplicationEmbeddedLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/embedded-links`, oazapfts.json({
        ...opts,
        method: "POST",
        body: createApplicationEmbeddedLinkRequest
    })));
}
/**
 * Register the snapshot of an application's deploy.
 */
export function workspaceApplicationControllerprocessDeploySnapshot({ workspaceId, applicationId, applicationDeploySnapshotRequest }: {
    workspaceId: string;
    applicationId: string;
    applicationDeploySnapshotRequest: ApplicationDeploySnapshotRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/deploy`, oazapfts.json({
        ...opts,
        method: "POST",
        body: applicationDeploySnapshotRequest
    })));
}
/**
 * List all workspaces usage by stack version list.
 */
export function usageInsightsControllerlistAllStackUsage({ accountId, body }: {
    accountId: string;
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackUsageByWorkspaceResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    }>("/v1/workspaces/stacks/usage", oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Get plugins applied to a workspace's application.
 */
export function accountSharedInfrastructureControllergetSharedInfraPage({ page, size, sortBy, sortDir, getAllSharedInfraRequestBody }: {
    page?: number;
    size?: number;
    sortBy?: "NAME" | "CREATED_AT" | "STACK_VERSION" | "STACK";
    sortDir?: "ASC" | "DESC";
    getAllSharedInfraRequestBody?: GetAllSharedInfraRequestBody;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraPageResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/shared-infra/-/search${QS.query(QS.explode({
        page,
        size,
        sortBy,
        sortDir
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: getAllSharedInfraRequestBody
    })));
}
/**
 * Retrieve the amount of apps and shared infras that use the given plugin version id list
 */
export function usageInsightsControllerfindPluginUsageAmount({ accountId, body }: {
    accountId: string;
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PluginUsageAmountConsolidatedResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/plugin-usage", oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Get environments
 */
export function environmentControllergetEnvironments({ accountId }: {
    accountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: EnvironmentReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/environments", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Save environment
 */
export function environmentControllersave({ environmentSaveRequest }: {
    environmentSaveRequest: EnvironmentSaveRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/environments", oazapfts.json({
        ...opts,
        method: "POST",
        body: environmentSaveRequest
    })));
}
/**
 * Batch save environment
 */
export function environmentControllersaveBatch({ accountId, body }: {
    accountId?: string;
    body: EnvironmentSaveRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: EnvironmentReadBatchResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/environments/batch", oazapfts.json({
        ...opts,
        method: "POST",
        body,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    })));
}
/**
 * Get plugins applied to a workspace's application.
 */
export function accountApplicationControllergetApplicationPage({ page, size, sortBy, sortDir, getAllApplicationsRequestBody }: {
    page?: number;
    size?: number;
    sortBy?: "NAME" | "CREATED_AT" | "STACK_VERSION" | "STACK";
    sortDir?: "ASC" | "DESC";
    getAllApplicationsRequestBody?: GetAllApplicationsRequestBody;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationPageResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/applications/-/search${QS.query(QS.explode({
        page,
        size,
        sortBy,
        sortDir
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: getAllApplicationsRequestBody
    })));
}
/**
 * List all workflows of an account.
 */
export function workflowAccountControllerlistAccountWorkflows({ $type, name }: {
    $type?: "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
    name?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows${QS.query(QS.explode({
        "type": $type,
        name
    }))}`, {
        ...opts
    }));
}
/**
 * Create an account workflow
 */
export function workflowAccountControllersaveAccountWorkflow({ accountWorkflowCreateRequest }: {
    accountWorkflowCreateRequest: AccountWorkflowCreateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
        data: IdResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/account/workflows", oazapfts.json({
        ...opts,
        method: "POST",
        body: accountWorkflowCreateRequest
    })));
}
/**
 * Find all account variables
 */
export function accountVariableControllerfindAll({ accountId, name, page, size, sortBy, sortDir }: {
    accountId?: string;
    name?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "VALUE" | "CREATED_AT";
    sortDir?: "ASC" | "DESC";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedAccountVariableResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/variables${QS.query(QS.explode({
        name,
        page,
        size,
        sortBy,
        sortDir
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Create account variable
 */
export function accountVariableControllercreate({ createAccountVariableRequest }: {
    createAccountVariableRequest: CreateAccountVariableRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 201;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/account/variables", oazapfts.json({
        ...opts,
        method: "POST",
        body: createAccountVariableRequest
    })));
}
/**
 * List all stacks with context for given account and given stackVersionId list.
 */
export function contextControllerlistStackVersionsWithAccountContext({ body }: {
    body: string[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountStackContextResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/account/stacks-with-context", oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Get workspace by id
 */
export function workspaceControllergetWorkspaceForId({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}`, {
        ...opts
    }));
}
/**
 * Delete workspace.
 */
export function workspaceControllerdelete({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update workspace
 */
export function workspaceControllerupdate({ workspaceId, updateWorkspaceRequest }: {
    workspaceId: string;
    updateWorkspaceRequest: UpdateWorkspaceRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateWorkspaceRequest
    })));
}
/**
 * Get shared infra links by id
 */
export function sharedInfraLinkControllergetSharedInfraLink({ workspaceId, sharedInfraId, linkId }: {
    workspaceId: string;
    sharedInfraId: string;
    linkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: SharedInfraLinkReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/links/${encodeURIComponent(linkId)}`, {
        ...opts
    }));
}
/**
 * Delete shared infra links
 */
export function sharedInfraLinkControllerdelete({ workspaceId, sharedInfraId, linkId }: {
    workspaceId: string;
    sharedInfraId: string;
    linkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/links/${encodeURIComponent(linkId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update shared infra links
 */
export function sharedInfraLinkControllerupdate({ workspaceId, sharedInfraId, linkId, updateSharedInfraLinkRequest }: {
    workspaceId: string;
    sharedInfraId: string;
    linkId: string;
    updateSharedInfraLinkRequest: UpdateSharedInfraLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/links/${encodeURIComponent(linkId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateSharedInfraLinkRequest
    })));
}
/**
 * Get workspace link by id
 */
export function workspaceLinkControllergetWorkspaceLink({ workspaceId, linkId }: {
    workspaceId: string;
    linkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceLinkReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/links/${encodeURIComponent(linkId)}`, {
        ...opts
    }));
}
/**
 * Delete workspace links
 */
export function workspaceLinkControllerdelete({ workspaceId, linkId }: {
    workspaceId: string;
    linkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/links/${encodeURIComponent(linkId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update workspace links
 */
export function workspaceLinkControllerupdate({ workspaceId, linkId, updateWorkspaceLinkRequest }: {
    workspaceId: string;
    linkId: string;
    updateWorkspaceLinkRequest: UpdateWorkspaceLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/links/${encodeURIComponent(linkId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateWorkspaceLinkRequest
    })));
}
/**
 * Get workspace embedded link by id
 */
export function workspaceEmbeddedLinkControllergetEmbeddedLink({ workspaceId, embeddedLinkId }: {
    workspaceId: string;
    embeddedLinkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceEmbeddedLinkReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/embedded-links/${encodeURIComponent(embeddedLinkId)}`, {
        ...opts
    }));
}
/**
 * Delete workspace embedded link
 */
export function workspaceEmbeddedLinkControllerdelete({ workspaceId, embeddedLinkId }: {
    workspaceId: string;
    embeddedLinkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/embedded-links/${encodeURIComponent(embeddedLinkId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update workspace embedded link
 */
export function workspaceEmbeddedLinkControllerupdate({ workspaceId, embeddedLinkId, updateWorkspaceEmbeddedLinkRequest }: {
    workspaceId: string;
    embeddedLinkId: string;
    updateWorkspaceEmbeddedLinkRequest: UpdateWorkspaceEmbeddedLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/embedded-links/${encodeURIComponent(embeddedLinkId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateWorkspaceEmbeddedLinkRequest
    })));
}
/**
 * Upsert workspace embedded link
 */
export function workspaceEmbeddedLinkControllerupsertBatch({ workspaceId, body }: {
    workspaceId: string;
    body: UpsertWorkspaceEmbeddedLinkRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/embedded-links/batch`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Get application link by id
 */
export function applicationLinkControllergetApplicationLink({ workspaceId, applicationId, linkId }: {
    workspaceId: string;
    applicationId: string;
    linkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationLinkReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/links/${encodeURIComponent(linkId)}`, {
        ...opts
    }));
}
/**
 * Delete workspace links
 */
export function applicationLinkControllerdelete({ workspaceId, applicationId, linkId }: {
    workspaceId: string;
    applicationId: string;
    linkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/links/${encodeURIComponent(linkId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update application links
 */
export function applicationLinkControllerupdate({ workspaceId, applicationId, linkId, updateApplicationLinkRequest }: {
    workspaceId: string;
    applicationId: string;
    linkId: string;
    updateApplicationLinkRequest: UpdateApplicationLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/links/${encodeURIComponent(linkId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateApplicationLinkRequest
    })));
}
/**
 * Get application embedded link by id
 */
export function applicationEmbeddedLinkControllergetEmbeddedLink({ workspaceId, applicationId, embeddedLinkId }: {
    workspaceId: string;
    applicationId: string;
    embeddedLinkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationEmbeddedLinkReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/embedded-links/${encodeURIComponent(embeddedLinkId)}`, {
        ...opts
    }));
}
/**
 * Delete application embedded link
 */
export function applicationEmbeddedLinkControllerdelete({ workspaceId, applicationId, embeddedLinkId }: {
    workspaceId: string;
    applicationId: string;
    embeddedLinkId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/embedded-links/${encodeURIComponent(embeddedLinkId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Update application embedded link
 */
export function applicationEmbeddedLinkControllerupdate({ workspaceId, applicationId, embeddedLinkId, updateApplicationEmbeddedLinkRequest }: {
    workspaceId: string;
    applicationId: string;
    embeddedLinkId: string;
    updateApplicationEmbeddedLinkRequest: UpdateApplicationEmbeddedLinkRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/embedded-links/${encodeURIComponent(embeddedLinkId)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: updateApplicationEmbeddedLinkRequest
    })));
}
/**
 * Upsert application embedded link
 */
export function applicationEmbeddedLinkControllerupsertBatch({ workspaceId, applicationId, body }: {
    workspaceId: string;
    applicationId: string;
    body: UpsertApplicationEmbeddedLinkRequest[];
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/embedded-links/batch`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Archive application from a workspace.
 */
export function workspaceApplicationControllerarchiveApplication({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/archive`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Get environment by id
 */
export function environmentControllergetEnvironment({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: EnvironmentReadResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/environments/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Update environment
 */
export function environmentControllerupdate({ id, environmentUpdateRequest }: {
    id: string;
    environmentUpdateRequest: EnvironmentUpdateRequest;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: IdResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/environments/${encodeURIComponent(id)}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body: environmentUpdateRequest
    })));
}
/**
 * Find all workspace variables
 */
export function workspaceVariableV2ControllerfindAll({ workspaceId, accountId, page, size, sortBy, sortDir, mandate, name, showEmptyValues }: {
    workspaceId: string;
    accountId?: string;
    page?: number;
    size?: number;
    sortBy?: "NAME" | "VALUE" | "CREATED_AT";
    sortDir?: "ASC" | "DESC";
    mandate?: boolean;
    name?: string;
    showEmptyValues?: boolean;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedWorkspaceVariableResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/variables${QS.query(QS.explode({
        page,
        size,
        sortBy,
        sortDir,
        mandate,
        name,
        showEmptyValues
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get plugins applied to a workspace's shared infrastructure.
 */
export function workspaceSharedInfrastructureV2ControllergetAppliedPlugins({ workspaceId, sharedInfraId, environmentId }: {
    workspaceId: string;
    sharedInfraId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AppliedPluginV2Response[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/environments/${encodeURIComponent(environmentId)}/plugins`, {
        ...opts
    }));
}
/**
 * List all available connection interfaces for a workspace.
 */
export function availableConnectionInterfaceV2ControllergetAvailableConnectionInterfacesForAWorkspace({ workspaceId, typeId, automaticallyGenerated }: {
    workspaceId: string;
    typeId?: string;
    automaticallyGenerated?: boolean;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceSummaryV2Response[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/available-connection-interfaces${QS.query(QS.explode({
        typeId,
        automaticallyGenerated
    }))}`, {
        ...opts
    }));
}
/**
 * Get available connection interface for a workspace by its slug.
 */
export function availableConnectionInterfaceV2ControllergetAvailableConnectionInterfaceForAWorkspace({ workspaceId, slug, environmentId, accountId }: {
    workspaceId: string;
    slug: string;
    environmentId?: string;
    accountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceDetailsV2Response;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/available-connection-interfaces/${encodeURIComponent(slug)}${QS.query(QS.explode({
        environmentId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get plugins applied to a workspace's application.
 */
export function workspaceApplicationV2ControllergetAppliedPlugins({ workspaceId, applicationId, environmentId }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AppliedPluginV2Response[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(environmentId)}/plugins`, {
        ...opts
    }));
}
/**
 * List all available connection interfaces for an application.
 */
export function availableConnectionInterfaceV2ControllergetAvailableConnectionInterfacesForAnApplication({ applicationId, typeId, automaticallyGenerated }: {
    applicationId: string;
    typeId?: string;
    automaticallyGenerated?: boolean;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceSummaryV2Response[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/applications/${encodeURIComponent(applicationId)}/available-connection-interfaces${QS.query(QS.explode({
        typeId,
        automaticallyGenerated
    }))}`, {
        ...opts
    }));
}
/**
 * Get available connection interface for an application by its slug.
 */
export function availableConnectionInterfaceV2ControllergetAvailableConnectionInterfaceForAnApplication({ applicationId, slug, environmentId }: {
    applicationId: string;
    slug: string;
    environmentId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceDetailsV2Response;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v2/applications/${encodeURIComponent(applicationId)}/available-connection-interfaces/${encodeURIComponent(slug)}${QS.query(QS.explode({
        environmentId
    }))}`, {
        ...opts
    }));
}
/**
 * Get connection interface in use by applicationId or sharedInfraId.
 */
export function connectionInterfaceInUseControllergetConnectionInterfaceInUseTargetId({ workspaceId, targetId, targetType, slug, accountId, environmentId }: {
    workspaceId: string;
    targetId: string;
    targetType: "applications" | "shared_infra";
    slug: string;
    accountId: string;
    environmentId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceInUseByTargetIdDetailsResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/${encodeURIComponent(targetType)}/${encodeURIComponent(targetId)}/connection-interfaces-in-use/${encodeURIComponent(slug)}/plugins${QS.query(QS.explode({
        environmentId
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get all workspace's variables v1
 */
export function workspaceVariableV1ControllerfindAll({ workspaceId, filter }: {
    workspaceId: string;
    filter?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceVariableReadV1Response[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/variables${QS.query(QS.explode({
        filter
    }))}`, {
        ...opts
    }));
}
/**
 * Get stack in a workspace by stack version id.
 */
export function workspaceStackControllergetStackById({ workspaceId, stackVersionId }: {
    workspaceId: string;
    stackVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StackInWorkspaceReadResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}`, {
        ...opts
    }));
}
/**
 * Delete a stack from a workspace.
 */
export function workspaceStackControllerdeleteStack({ workspaceId, stackVersionId }: {
    workspaceId: string;
    stackVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * List the consolidated input context of all stack elements for given type in a workspace.
 */
export function contextControllerlistConsolidatedContext({ workspaceId, stackVersionId, $type, envName, externalId }: {
    workspaceId: string;
    stackVersionId: string;
    $type: "plugin" | "action";
    envName?: string;
    externalId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConsolidatedContextResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackVersionId)}/${encodeURIComponent($type)}-context${QS.query(QS.explode({
        envName,
        externalId
    }))}`, {
        ...opts
    }));
}
/**
 * Get workflow settings with context by stack and workflowType
 */
export function workflowWorkspaceControllerlistWorkflowByStackIdAndWorkflowType({ workspaceId, stackId, workflowType }: {
    workspaceId: string;
    stackId: string;
    workflowType: "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountWorkflowResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/stacks/${encodeURIComponent(stackId)}/workflows/${encodeURIComponent(workflowType)}`, {
        ...opts
    }));
}
/**
 * Get plugins applied to a workspace's shared infrastructure.
 */
export function workspaceSharedInfrastructureControllergetAppliedPlugins({ workspaceId, sharedInfraId, environmentId, $type }: {
    workspaceId: string;
    sharedInfraId: string;
    environmentId: string;
    $type?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AppliedPluginResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/environments/${encodeURIComponent(environmentId)}/plugins${QS.query(QS.explode({
        "type": $type
    }))}`, {
        ...opts
    }));
}
/**
 * List shared infrastructure's activities.
 */
export function workspaceSharedInfrastructureControllerlistActivities({ workspaceId, sharedInfraId, environmentId, $type, page, size }: {
    workspaceId: string;
    sharedInfraId: string;
    environmentId: string;
    $type?: ("DEPLOY" | "ACTION" | "WORKFLOW" | "DRIFT" | "DESTROY" | "ROLLBACK" | "IAC" | "DEPLOY_SELF_HOSTED" | "DESTROY_SELF_HOSTED" | "ROLLBACK_SELF_HOSTED" | "IAC_SELF_HOSTED" | "CUSTOMER_WORKFLOW")[];
    page: number;
    size: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedActivityResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/environments/${encodeURIComponent(environmentId)}/activities${QS.query(QS.explode({
        "type": $type,
        page,
        size
    }))}`, {
        ...opts
    }));
}
/**
 * Get all in use connection interface dependencies from the shared infra.
 */
export function workspaceSharedInfrastructureControllergetDependencyTree({ workspaceId, sharedInfraId }: {
    workspaceId: string;
    sharedInfraId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AppAndInfraDependencyResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/${encodeURIComponent(sharedInfraId)}/dependencies`, {
        ...opts
    }));
}
/**
 * Get stack version in use by shared infra in a workspace
 */
export function workspaceSharedInfrastructureControllergetStackVersionsInUse({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StacksVersionIdsResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/shared-infra/stacks/versions`, {
        ...opts
    }));
}
/**
 * Get a connection interface by connection interface id.
 */
export function connectionInterfaceControllergetConnectionInterface({ workspaceId, connectionInterfaceId, environmentId }: {
    workspaceId: string;
    connectionInterfaceId: string;
    environmentId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceDetailsResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/${encodeURIComponent(connectionInterfaceId)}${QS.query(QS.explode({
        environmentId
    }))}`, {
        ...opts
    }));
}
/**
 * Get all connection interface in use.
 */
export function connectionInterfaceInUseControllergetAllConnectionInterfaceInUse({ workspaceId, accountId }: {
    workspaceId: string;
    accountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceInUseDetailsResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces-in-use`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get connection interface in use by slug.
 */
export function connectionInterfaceInUseControllergetConnectionInterfaceInUseBySlug({ workspaceId, slug, accountId }: {
    workspaceId: string;
    slug: string;
    accountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceInUseSummaryResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces-in-use/${encodeURIComponent(slug)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * List all available connection interfaces for a workspace.
 */
export function availableConnectionInterfaceControllergetAvailableConnectionInterfacesForAWorkspace({ workspaceId, typeId }: {
    workspaceId: string;
    typeId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceSummaryResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/available-connection-interfaces${QS.query(QS.explode({
        typeId
    }))}`, {
        ...opts
    }));
}
/**
 * Get available connection interface for a workspace by its id.
 */
export function availableConnectionInterfaceControllergetAvailableConnectionInterfaceForAWorkspace({ workspaceId, connectionInterfaceId, environmentId }: {
    workspaceId: string;
    connectionInterfaceId: string;
    environmentId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceDetailsResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/available-connection-interfaces/${encodeURIComponent(connectionInterfaceId)}${QS.query(QS.explode({
        environmentId
    }))}`, {
        ...opts
    }));
}
/**
 * Get plugins applied to a workspace's application.
 */
export function workspaceApplicationControllergetAppliedPlugins({ workspaceId, applicationId, environmentId, $type }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    $type?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AppliedPluginResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(environmentId)}/plugins${QS.query(QS.explode({
        "type": $type
    }))}`, {
        ...opts
    }));
}
export function workspaceApplicationControllergetAppDeployInfo({ workspaceId, applicationId, environmentId, accountId, tenant }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    accountId: string;
    tenant: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationDeployInfoResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(environmentId)}/deploy-info`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId,
            tenant
        })
    }));
}
/**
 * List application's activities.
 */
export function workspaceApplicationControllerlistActivities({ workspaceId, applicationId, environmentId, $type, page, size }: {
    workspaceId: string;
    applicationId: string;
    environmentId: string;
    $type?: ("DEPLOY" | "ACTION" | "WORKFLOW" | "DRIFT" | "DESTROY" | "ROLLBACK" | "IAC" | "DEPLOY_SELF_HOSTED" | "DESTROY_SELF_HOSTED" | "ROLLBACK_SELF_HOSTED" | "IAC_SELF_HOSTED" | "CUSTOMER_WORKFLOW")[];
    page: number;
    size: number;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PaginatedActivityResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/environments/${encodeURIComponent(environmentId)}/activities${QS.query(QS.explode({
        "type": $type,
        page,
        size
    }))}`, {
        ...opts
    }));
}
/**
 * Get all in use connection interface dependencies from the application.
 */
export function workspaceApplicationControllergetDependencyTree({ workspaceId, applicationId }: {
    workspaceId: string;
    applicationId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AppAndInfraDependencyResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/dependencies`, {
        ...opts
    }));
}
/**
 * Can the application be destroyed?
 */
export function workspaceApplicationControllercanBeDeleted({ workspaceId, applicationId, accountId, tenant }: {
    workspaceId: string;
    applicationId: string;
    accountId: string;
    tenant: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
        data: ApplicationCanBeDeletedResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/can-be-deleted`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId,
            tenant
        })
    }));
}
/**
 * Get stack version in use by applications in a workspace
 */
export function workspaceApplicationControllergetStackVersionsInUse({ workspaceId }: {
    workspaceId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StacksVersionIdsResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/stacks/versions`, {
        ...opts
    }));
}
/**
 * Get all workspaces within user permission
 */
export function workspaceControllergetWorkspacesFromUserPermission({ resource, action }: {
    resource: string;
    action: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkspaceReadResponse[];
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/from-user-permission/${encodeURIComponent(resource)}/${encodeURIComponent(action)}`, {
        ...opts
    }));
}
/**
 * Get shared infrastructure information by id.
 */
export function accountSharedInfrastructureControllergetSharedInfra({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetSharedInfraResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    }>(`/v1/shared-infra/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * Check if there is/are any deployment active for the provided plugin
 */
export function pluginInUseControllercheckPluginInUse({ pluginVersionId }: {
    pluginVersionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: PluginInUseDetailsResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/plugin-in-use/${encodeURIComponent(pluginVersionId)}`, {
        ...opts
    }));
}
/**
 * Check availability of connection slug in account.
 */
export function checkConnectionSlugAvailabilityControllercheckConnectionSlugAvailability({ slug, applicationId, sharedInfraId }: {
    slug: string;
    applicationId?: string;
    sharedInfraId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: CheckConnectionSlugAvailabilityResponse;
    } | {
        status: 404;
        data: CheckConnectionSlugAvailabilityResponse;
    }>(`/v1/connection-interfaces/${encodeURIComponent(slug)}/availability${QS.query(QS.explode({
        applicationId,
        sharedInfraId
    }))}`, {
        ...opts
    }));
}
/**
 * Get an application list by repo url.
 */
export function accountApplicationControllergetApplicationsByUrl({ accountId, repoUrl }: {
    accountId: string;
    repoUrl: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ApplicationRepoUrlReadResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/applications${QS.query(QS.explode({
        repoUrl
    }))}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get application information by id.
 */
export function accountApplicationControllergetApplication({ id }: {
    id: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: GetApplicationResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    }>(`/v1/applications/${encodeURIComponent(id)}`, {
        ...opts
    }));
}
/**
 * List all available connection interfaces for an application.
 */
export function availableConnectionInterfaceControllergetAvailableConnectionInterfacesForAnApplication({ applicationId, typeId }: {
    applicationId: string;
    typeId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceSummaryResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/applications/${encodeURIComponent(applicationId)}/available-connection-interfaces${QS.query(QS.explode({
        typeId
    }))}`, {
        ...opts
    }));
}
/**
 * Get available connection interface for an application by its id.
 */
export function availableConnectionInterfaceControllergetAvailableConnectionInterfaceForAnApplication({ applicationId, connectionInterfaceId, environmentId }: {
    applicationId: string;
    connectionInterfaceId: string;
    environmentId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: ConnectionInterfaceDetailsResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/applications/${encodeURIComponent(applicationId)}/available-connection-interfaces/${encodeURIComponent(connectionInterfaceId)}${QS.query(QS.explode({
        environmentId
    }))}`, {
        ...opts
    }));
}
/**
 * Get action version ranges used in workflows given an actionId.
 */
export function workflowAccountControllerlistActionVersionRangesUsage({ accountId, actionId }: {
    accountId?: string;
    actionId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowActionUsageResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows/actions/${encodeURIComponent(actionId)}/versions`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Get account variable usage
 */
export function accountVariableControllerusage({ accountId, name }: {
    accountId?: string;
    name: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountVariableUsageResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/variables/${encodeURIComponent(name)}/usage`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * List the input context of all stack elements for given type in account.
 */
export function contextControllerlistAccountContext({ stackVersionId, $type, envName, externalId }: {
    stackVersionId: string;
    $type: "plugin" | "action";
    envName?: string;
    externalId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountContextResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/stacks/${encodeURIComponent(stackVersionId)}/${encodeURIComponent($type)}-context${QS.query(QS.explode({
        envName,
        externalId
    }))}`, {
        ...opts
    }));
}
/**
 * Get workflow of a stackId by workflow type.
 */
export function workflowStackControllerlistWorkflowByStackIdAndType({ stackId, workflowType }: {
    stackId: string;
    workflowType: "CREATE_API" | "CREATE_APP" | "CREATE_INFRA";
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountWorkflowResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/stacks/${encodeURIComponent(stackId)}/workflows/${encodeURIComponent(workflowType)}`, {
        ...opts
    }));
}
/**
 * Get workflows by a stackId.
 */
export function workflowStackControllerlistWorkflowsByStackId({ stackId }: {
    stackId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: AccountWorkflowResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/stacks/${encodeURIComponent(stackId)}/all-workflows`, {
        ...opts
    }));
}
/**
 * List all stacks with workflow by account.
 */
export function accountStackControllerlistStacksWithWorkflow(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: StacksByAccountResponse;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/account/stacks/with-workflow", {
        ...opts
    }));
}
/**
 * List all stacks that are most commonly used by applications and shared infrastructure.
 */
export function accountStackControllerlistMostUsedStacks({ accountId }: {
    accountId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: string[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/account/stacks/most-used", {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * List all default workflows of an account.
 */
export function workflowAccountControllerlistDefaultAccountWorkflows(opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 200;
        data: WorkflowResponse[];
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>("/v1/account/default-workflows", {
        ...opts
    }));
}
/**
 * Delete completely connection interfaces from all environments.
 */
export function connectionInterfaceControllerdeleteConnectionInterfaceFromAllEnvironments({ workspaceId, slug }: {
    workspaceId: string;
    slug: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/${encodeURIComponent(slug)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete connection interfaces.
 */
export function connectionInterfaceControllerdeleteConnectionInterface({ workspaceId, slug, environmentId, accountId }: {
    workspaceId: string;
    slug: string;
    environmentId: string;
    accountId?: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/${encodeURIComponent(slug)}/environment/${encodeURIComponent(environmentId)}`, {
        ...opts,
        method: "DELETE",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            accountId
        })
    }));
}
/**
 * Delete connection interface attributes by environment.
 */
export function connectionInterfaceControllerdeleteConnectionInterfaceAttributesFromEnvironment({ workspaceId, slug, environmentId }: {
    workspaceId: string;
    slug: string;
    environmentId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 404;
        data: ErrorResponse;
    } | {
        status: 422;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/workspaces/${encodeURIComponent(workspaceId)}/connection-interfaces/${encodeURIComponent(slug)}/attributes${QS.query(QS.explode({
        environmentId
    }))}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete an account workflow.
 */
export function workflowAccountControllerdeleteAccountWorkflow({ workflowId }: {
    workflowId: string;
}, opts?: Oazapfts.RequestOpts) {
    return oazapfts.ok(oazapfts.fetchJson<{
        status: 204;
    } | {
        status: 400;
        data: ErrorResponse;
    } | {
        status: 403;
        data: ErrorResponse;
    } | {
        status: 500;
        data: ErrorResponse;
    }>(`/v1/account/workflows/${encodeURIComponent(workflowId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
