// @stack-spot/portal-layout@2.20.1/dist/components/NotificationCenter/dictionary.js
import { useTranslate } from '@stack-spot/portal-translate';
const dictionary = {
    en: {
        notifications: 'Notifications',
        'ALL.ariaLabel': 'Show all notifications',
        'UNREAD.ariaLabel': 'Show unread notifications',
        'HIGH.ariaLabel': 'Show critical notifications',
        'MEDIUM.ariaLabel': 'Show notifications with medium criticality',
        'LOW.ariaLabel': 'Show notifications with low criticality',
        'ALL.label': 'All',
        'UNREAD.label': 'Unread',
        'HIGH.label': 'High',
        'MEDIUM.label': 'Medium',
        'LOW.label': 'Low',
        seeAll: 'See all notifications',
        openNotifications: 'View notifications',
        hasUnread: 'Has Unread notifications',
        close: 'Close',
        loadingNotifications: "Loading notifications. You'll be able to click this button and see them in a few seconds.",
        tour: 'Now we have Notifications, this will help you stay informed of the latest updates about your Workspaces, Studios, and the Stackspot platform.',
    },
    pt: {
        notifications: 'Notificações',
        'ALL.ariaLabel': 'Mostrar todas as notificações',
        'UNREAD.ariaLabel': 'Mostrar notificações não lidas',
        'HIGH.ariaLabel': 'Mostrar notificações críticas',
        'MEDIUM.ariaLabel': 'Mostrar notificações com criticidade média',
        'LOW.ariaLabel': 'Mostrar notificações com criticidade baixa',
        'ALL.label': 'Todas',
        'UNREAD.label': 'Não lidas',
        'HIGH.label': 'Alto',
        'MEDIUM.label': 'Médio',
        'LOW.label': 'Baixo',
        seeAll: 'Ver todas as notificações',
        openNotifications: 'Visualizar notificações',
        hasUnread: 'Existem notificações não lidas',
        close: 'Fechar',
        loadingNotifications: 'Carregando notificações. Você poderá clicar neste botão e vê-las em alguns segundos.',
        tour: 'Agora temos Notificações, isso irá ajudá-lo a se manter informado sobre as últimas atualizações sobre seus Workspaces, estúdios e a plataforma Stackspot.',
    },
};
export const useNotificationsDictionary = () => useTranslate(dictionary);
