// @stack-spot/portal-components@2.11.0/dist/components/FileTreeView/More.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Box, IconBox, Text } from '@citric/core';
import * as icons from '@citric/icons';
import { EllipsisVertical } from '@citric/icons';
import { IconButton } from '@citric/ui';
import { theme } from '@stack-spot/portal-theme';
import { createElement, useState } from 'react';
const styles = {
    MoreWrapper: {
        position: 'relative',
        'button': {
            width: '2rem',
            height: '2rem',
        },
        '.drop': {
            zIndex: 1,
            position: 'absolute',
            listStyle: 'none',
            display: 'flex',
            margin: '0',
            padding: '0.25rem',
            flexDirection: 'column',
            justifyContent: 'center',
            alignItems: 'flex-start',
            gap: '0.5rem',
            borderRadius: '0.25rem',
            border: `0.063rem solid ${theme.color.light[400]}`,
            backgroundColor: `${theme.color.light[500]}`,
            boxShadow: `0 0 0 0.125rem ${theme.color.gray[600]}`,
            right: '2px',
            width: 'max-content',
            'ul': {
                padding: 0,
                margin: 0,
            },
            '.item-list-float': {
                minWidth: '230px',
                listStyleType: 'none',
                padding: '8px',
                paddingRight: '12px',
                display: 'flex',
                alignItems: 'center',
                cursor: 'pointer',
                '&:hover': {
                    backgroundColor: `${theme.color.light[600]}`,
                },
            },
        },
    },
};
export function More({ loading, menu }) {
    const [opened, setOpened] = useState(false);
    return (_jsxs(Box, { sx: styles.MoreWrapper, children: [_jsx(IconButton, { color: "light", size: "sm", appearance: "circle", "aria-label": "More options", disabled: loading, onClick: () => setOpened((old) => !old), onBlur: () => setTimeout(() => setOpened(false), 200), children: _jsx(IconBox, { size: "xs", children: _jsx(EllipsisVertical, {}) }) }), opened && (_jsx("div", { className: "drop", children: _jsx("ul", { className: "drop-wrapper", children: menu.map((menuItem, key) => (_jsxs("li", { role: "button", className: "item-list-float", onClick: menuItem.onClick, children: [_jsx(IconBox, { size: "sm", children: createElement(icons[menuItem.icon]) }), _jsx(Text, { appearance: "body2", children: menuItem.label })] }, key))) }) }))] }));
}
