// @stack-spot/portal-network@0.66.0/dist/network/ManualQuery.js
/* eslint-disable react-hooks/rules-of-hooks */
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ManualQuery_instances, _ManualQuery_useQueryResult;
import { useQuery, useSuspenseQuery } from '@tanstack/react-query';
import { ManualOperation } from './ManualOperation.js';
import { queryClient } from './react-query-client.js';
export class ManualQuery extends ManualOperation {
    constructor(config) {
        super(config);
        _ManualQuery_instances.add(this);
    }
    async makeRequest(variables, signal) {
        try {
            return await this.config.request(...[signal, variables]);
        }
        catch (error) {
            throw this.config.transformError(error);
        }
    }
    query(...[variables]) {
        return queryClient.fetchQuery({
            queryKey: this.getKey(variables),
            queryFn: ({ signal }) => this.makeRequest(variables, signal),
        });
    }
    useQuery(...args) {
        const result = __classPrivateFieldGet(this, _ManualQuery_instances, "m", _ManualQuery_useQueryResult).call(this, false, ...args);
        return result.data;
    }
    useStatefulQuery(...args) {
        const result = __classPrivateFieldGet(this, _ManualQuery_instances, "m", _ManualQuery_useQueryResult).call(this, false, ...args);
        return [result.data, result.isPending, result.error, result];
    }
    invalidate(variables) {
        return queryClient.invalidateQueries({ queryKey: this.getKey(variables) });
    }
    getKey(variables) {
        return [this.config.apiName, this.config.name, variables];
    }
    cancel(variables) {
        queryClient.cancelQueries({ queryKey: this.getKey(variables) });
    }
}
_ManualQuery_instances = new WeakSet(), _ManualQuery_useQueryResult = function _ManualQuery_useQueryResult(suspense, ...args) {
    /* `this.config.request` is a function with arity 1 or 2. If it accepts variables, its arity is 2: the 1st parameter is the
    AbortSignal and the 2nd is the variables. If it doesn't accept variables, its arity is one: it accepts only the AbortSignal.
    We can use this information to determine what the type of `args` actually is at runtime. If variables are accepted, than the 1st
    argument is the variables and the 2nd is the query options, otherwise, it has a single argument, which is the query options. */
    const [variables, options] = this.config.request.length === 2 ? args : [undefined, args[0]];
    const use = suspense ? useSuspenseQuery : useQuery;
    return use({
        ...options,
        queryKey: this.getKey(variables),
        queryFn: ({ signal }) => this.makeRequest(variables, signal),
    }, queryClient);
};
