// @citric/core@6.4.0/src/field-base.tsx
import { SxProp } from './component-style'
import { css } from './styled'
import { OneOfColorSchemes, Theme } from './theme-types'
import { getColor, getFontStyle, getRadius, getSpacing } from './utils/theme'

interface FieldProps {
  sx?: SxProp,
  colorScheme?: OneOfColorSchemes,
}

const getFieldBaseStyle = (theme: Theme, colorScheme?: OneOfColorSchemes) => css`
  width: 100%;
  height: 2.5rem;
  padding: ${getSpacing(theme, 3)};
  outline: none;
  box-sizing: border-box;
  background-color: ${getColor(theme, 'light.300')};
  border: 1px solid ${getColor(theme, colorScheme || 'light.600')};
  border-radius: ${getRadius(theme, 'xs')};
  font: ${getFontStyle(theme, 'body2')};
  color: ${getColor(theme, 'light.contrastText')};
  caret-color: ${getColor(theme, 'light.contrastText')};

  &::placeholder {
    opacity: 60%;
    color: ${getColor(theme, 'light.contrastText')};
  }

  &:focus {
    border-color: ${getColor(theme, colorScheme || 'primary')};
    box-shadow: inset 0 0 0 1px ${getColor(theme, colorScheme || 'primary')};
  }

  &:disabled {
    background-color: ${getColor(theme, 'light.500')};
    border-color: ${getColor(theme, 'light.600')};
    cursor: not-allowed;
  }

  &:-webkit-autofill,
  &:-webkit-autofill:hover, 
  &:-webkit-autofill:focus {
    -webkit-text-fill-color: ${getColor(theme, 'light.contrastText')};
    box-shadow: inset 0 0 0 1000px ${getColor(theme, 'light.300')} ;
    transition: background-color 50000s ease-in-out 0s;
  }
`

export { getFieldBaseStyle }
export type { FieldProps }
