// @citric/core@6.4.0/src/Switch.tsx
import { SxProp, extendStyle, overwriteStyle } from './component-style'
import { styled } from './styled'
import { defaultConfig } from './utils/config'
import { getColor, getRadius } from './utils/theme'

export interface SwitchProps {
  sx?: SxProp,
}

const Switch = styled.input.withConfig(defaultConfig).attrs({ type: 'checkbox' }) <SwitchProps>`
  ${extendStyle('Switch')};

  background-color: ${({ theme }) => getColor(theme, 'light.700')};
  border-radius: ${({ theme }) => getRadius(theme, 'md')};

  cursor: pointer;
  appearance: none;
  position: relative;
  width: 2.5rem;
  height: 1.25rem;
  box-sizing: border-box;

  &::before {
    content: '';
    background-color: ${({ theme }) => getColor(theme, 'success.contrastText')};
    transition: 0.32s;
    position: absolute;
    width: 0.75rem;
    height: 0.75rem;
    left: 0.25rem;
    top: 0.25rem;
    border-radius: ${({ theme }) => getRadius(theme, 'lg')};
  }

  &:checked {
    background-color: ${({ theme }) => getColor(theme, 'success.500')};

    &::before {
      transform: translateX(20px);
    }
  }

  &:disabled {
    opacity: 0.24;
    pointer-events: none;
  }

  ${overwriteStyle('Switch')};
`

export default Switch
