// @stack-spot/portal-components@2.11.0/dist/components/ScrollView.js
import { getColor } from '@citric/core/dist/utils/theme.js';
import { styled } from 'styled-components';
function getDirectionStyle({ direction = 'both', scrollType = 'auto' }) {
    const x = direction === 'both' || direction === 'horizontal' ? `overflow-x: ${scrollType};` : '';
    const y = direction === 'both' || direction === 'vertical' ? `overflow-x: ${scrollType};` : '';
    return `${x}${y}`;
}
/**
 * Renders a component that provides scrolling.
 *
 * @param options the props for rendering the component: {@link Props}.
 */
export const ScrollView = styled.div `
  ${getDirectionStyle}
  &::-webkit-scrollbar {
    width: 4px;
    height: 4px;
  }
  &::-webkit-scrollbar-track {
    background: ${({ theme, trackColor = 'light.400' }) => getColor(theme, trackColor)};
    border-radius: 2px;
    left: 40px;
  }
  &::-webkit-scrollbar-thumb,
  &::-webkit-scrollbar-thumb:hover {
    background: ${({ theme, thumbColor = 'light.700' }) => getColor(theme, thumbColor)};
    border-radius: 2px;
  }
`;
