// @stack-spot/portal-components@2.11.0/dist/components/TimelineSection.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Text } from '@citric/core';
import { useMonthDictionary, useTranslate } from '@stack-spot/portal-translate';
import { styled } from 'styled-components';
const Styled = styled.div `
  display: flex;
  flex-direction: row;
  align-items: start;

  .timeline-section-header {
    width: 80px;
  }

  .timeline-section-content {
    min-width: 0;
    flex: 1;
  }
`;
/**
 * Renders a section in a timeline where the left side is a header with the date and the right side is the content (children).
 * @param props React props: {@link Props}.
 */
export const TimelineSection = ({ children, day, month, className, style }) => {
    const months = useMonthDictionary();
    const t = useTranslate(locale);
    const ariaLabel = t.dayMonth.replace('%d', `${day}`).replace('%m', months[month]);
    return (_jsxs(Styled, { style: style, className: className, children: [_jsxs("div", { className: "timeline-section-header", children: [_jsx(Text, { appearance: "h2", "aria-label": ariaLabel, children: day >= 10 ? day : `0${day}` }), _jsx(Text, { appearance: "subtitle4", "aria-hidden": "true", children: months[month].toUpperCase().substring(0, 3) })] }), _jsx("div", { className: "timeline-section-content", children: children })] }));
};
const locale = {
    pt: {
        dayMonth: '%d de %m',
    },
    en: {
        dayMonth: '%m %d',
    },
};
