// @citric/ui@6.5.5/dist/Card.js
import { extendStyle, overwriteStyle } from '@citric/core/dist/component-style.js';
import { css } from '@citric/core/dist/styled.js';
import { defaultConfig } from '@citric/core/dist/utils/config.js';
import { getColor, getRadius, getSpacing } from '@citric/core/dist/utils/theme.js';
import { styled } from 'styled-components';
const CardHeader = styled.div `
  ${extendStyle('CardHeader')};
  ${overwriteStyle('CardHeader')};
`;
const CardContent = styled.div `
  ${extendStyle('CardContent')};
  ${overwriteStyle('CardContent')};
`;
const CardFooter = styled.div `
  ${extendStyle('CardFooter')};
  ${overwriteStyle('CardFooter')};
`;
const Card = styled.div.withConfig(defaultConfig) `
  display: block;
  box-sizing: border-box;
  text-decoration: none;
  color: inherit;
  

  ${({ clickable }) => clickable &&
    css `
      &:hover {
        cursor: pointer;
        box-shadow: 0px 0px 0px 2px ${props => getColor(props.theme, 'light.600')};
      }
    `}

  ${extendStyle('Card')};

  padding: ${({ theme }) => getSpacing(theme, 5)};
  background-color: ${({ theme }) => getColor(theme, 'light.400')};
  border: 1px solid ${({ theme }) => getColor(theme, 'light')};
  border-radius: ${({ theme }) => getRadius(theme, 'xs')};
  
  ${overwriteStyle('Card')};
`;
export { Card, CardContent, CardFooter, CardHeader };
