// @citric/ui@6.5.5/dist/Toast.js
import { extendStyle, overwriteStyle } from '@citric/core/dist/component-style.js';
import { css, styled } from '@citric/core/dist/styled.js';
import { defaultConfig } from '@citric/core/dist/utils/config.js';
import { getColor, getRadius } from '@citric/core/dist/utils/theme.js';
import IconButton from './IconButton.js';
const getToastPosition = (position) => {
    const toastPosition = {
        'top-right': css `
      top: 0.75rem;
      right: 0.75rem;
    `,
        'top-left': css `
      top: 0.75rem;
      left: 0.75rem;
    `,
        'bottom-right': css `
      bottom: 0.75rem;
      right: 0.75rem;
    `,
        'bottom-left': css `
      bottom: 0.75rem;
      left: 0.75rem;
    `,
    };
    return toastPosition[position];
};
const Toast = styled.div.withConfig(defaultConfig) `
  position: fixed;
  z-index: 2000;
  width: 21rem;

  box-sizing: border-box;

  ${extendStyle('Toast')};
  ${(props) => getToastPosition(props?.position || 'top-right')}
  border-radius: ${({ theme }) => getRadius(theme, 'xs')};
  background-color: ${({ theme, colorScheme }) => getColor(theme, colorScheme || 'primary.500')};
  color: ${({ theme, colorScheme }) => getColor(theme, colorScheme ? `${colorScheme}.contrastText` : 'primary.contrastText')};

  ${overwriteStyle('Toast')};
`;
const ToastButton = styled(IconButton) `
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
  ${extendStyle('ToastButton')};
  ${overwriteStyle('ToastButton')};
`;
export { Toast, ToastButton };
