// @stack-spot/portal-layout@2.20.1/dist/components/tour/StepContainer.js
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { theme } from '@stack-spot/portal-theme';
import { styled } from 'styled-components';
import { StepNavigation } from './StepNavigation.js';
import { StepTitle } from './StepTitle.js';
import { useTour } from './hook.js';
/**
 * Tutorial: the overlay component for showing a step on React Tour.
 * @param props the react props for the component {@link StepContainerProps}.
 */
export const StepContainer = ({ title, stepKey, customNavigation, position, children }) => {
    const { closeTour } = useTour();
    return _jsxs(BoxWithPointingArrow, { "$position": position, children: [_jsx(StepTitle, { title: title, onClose: () => closeTour() }), children, _jsx(StepNavigation, { stepKey: stepKey, ...(customNavigation || {}) })] });
};
const BoxWithPointingArrow = styled.div `
  position: relative;
  width: 100%;
  background-color: ${theme.color.inverse[500]};
  &::after {
    content: '';
    position: absolute;
    border-width: 10px;
    border-style: solid;
    border-color: transparent;
    margin-top: -5px;
    border-right-color: ${theme.color.inverse[500]};
    ${({ $position, $top }) => $position === 'right' ?
    `
      top: ${$top || '16px'};
      left: -18px;
    ` : ''}
    ${({ $position, $right }) => $position === 'top' ?
    `
    bottom: 96%;
    right: ${$right || '16px'};
    transform: rotate(90deg);
    ` : ''}
    ${({ $position, $top }) => $position === 'left' ?
    `
      top: ${$top || '16px'};
      right: -18px;
      transform: rotate(180deg);
    ` : ''}
    ${({ $position, $right }) => $position === 'bottom' ?
    `
    top: -13px;
    right: ${$right || '16px'};
    transform: rotate(90deg);
    ` : ''}
  }
`;
