// @citric/ui@6.5.5/src/ProgressLinear.tsx
import { SxProp, extendStyle, overwriteStyle } from '@citric/core/dist/component-style'
import { AsProp } from '@citric/core/dist/forward-ref-as'
import { styled } from '@citric/core/dist/styled'
import { OneOfColorSchemes } from '@citric/core/dist/theme-types'
import { defaultConfig } from '@citric/core/dist/utils/config'
import { getColor, getRadius } from '@citric/core/dist/utils/theme'

export interface ProgressLinearProps {
  sx?: SxProp,
  as?: AsProp,
  percentage?: number,
  colorScheme?: OneOfColorSchemes,
}

const ProgressLinear = styled.div.withConfig(defaultConfig) <ProgressLinearProps>`
  position: relative;
  border-radius: ${({ theme }) => getRadius(theme, 'lg')};
  width: 100%;
  height: 0.25rem;

  ${extendStyle('ProgressLinear')};

  background: ${({ theme }) => getColor(theme, 'light.600')};

  &::after {
    content: '';
    position: absolute;
    width: ${({ percentage = 50 }) => percentage}%;
    height: 0.125rem;
    top: 1px;
    left: 1px;
    border-radius: ${({ theme }) => getRadius(theme, 'lg')};
    background: ${({ theme, colorScheme = 'primary' }) => getColor(theme, colorScheme)};
  }

  ${overwriteStyle('ProgressLinear')};
`

export default ProgressLinear
