// @stack-spot/portal-components@2.11.0/src/components/notification/NotificationPlaceholder.tsx
import { AsProp } from '@citric/core/dist/forward-ref-as'
import { SxProperties } from '@citric/core/dist/sx'
import { Dictionary, useTranslate } from '@stack-spot/portal-translate'
import { PlaceholderCallToAction } from '../Placeholder'

interface Props {
  isSearch?: boolean,
  style?: React.CSSProperties,
  className?: string,
  sx?: SxProperties,
  sxCard?: SxProperties,
  titleAs?: AsProp,
}

export const NotificationPlaceholder = ({ isSearch, sx, sxCard, className, style, titleAs }: Props) => {
  const t = useTranslate(dictionary)

  return <PlaceholderCallToAction
    title={isSearch ? t.placeholderSearchTitle : t.placeholderTitle}
    description={isSearch ? t.placeholderSearchDescription : t.placeholderDescription}
    fullWidth
    sx={sx}
    sxCard={sxCard}
    className={className}
    style={style}
    titleAs={titleAs}
  />
}

const dictionary = {
  en: {
    placeholderTitle: 'No notifications',
    placeholderDescription: 'Your notifications will appear here.',
    placeholderSearchTitle: 'No notifications found',
    placeholderSearchDescription: 'You can change the current filters in the fields above',
  },
  pt: {
    placeholderTitle: 'Nenhuma notificação',
    placeholderDescription: 'Suas notificações aparecerão aqui.',
    placeholderSearchTitle: 'Nenhuma notificação encontrada',
    placeholderSearchDescription: 'Você pode mudar os filtros aplicados nos campos acima',
  },
} satisfies Dictionary
