// @stack-spot/portal-components@2.11.0/dist/components/notification/NotificationPlaceholder.js
import { jsx as _jsx } from "react/jsx-runtime";
import { useTranslate } from '@stack-spot/portal-translate';
import { PlaceholderCallToAction } from '../Placeholder.js';
export const NotificationPlaceholder = ({ isSearch, sx, sxCard, className, style, titleAs }) => {
    const t = useTranslate(dictionary);
    return _jsx(PlaceholderCallToAction, { title: isSearch ? t.placeholderSearchTitle : t.placeholderTitle, description: isSearch ? t.placeholderSearchDescription : t.placeholderDescription, fullWidth: true, sx: sx, sxCard: sxCard, className: className, style: style, titleAs: titleAs });
};
const dictionary = {
    en: {
        placeholderTitle: 'No notifications',
        placeholderDescription: 'Your notifications will appear here.',
        placeholderSearchTitle: 'No notifications found',
        placeholderSearchDescription: 'You can change the current filters in the fields above',
    },
    pt: {
        placeholderTitle: 'Nenhuma notificação',
        placeholderDescription: 'Suas notificações aparecerão aqui.',
        placeholderSearchTitle: 'Nenhuma notificação encontrada',
        placeholderSearchDescription: 'Você pode mudar os filtros aplicados nos campos acima',
    },
};
