// @stack-spot/portal-extension@0.5.0/src/layout.ts
import {
  AlertOptions,
  ConfirmOptions,
  CustomModalOptions,
  CustomRightPanelOptions,
  ModalOptions,
  RightPanelOptions,
  ToasterOptions,
  createRequestMessage,
} from '@stack-spot/portal-extension-commons'
import { messageHandler } from './MessageHandler'

export async function showModal(options: ModalOptions): Promise<void> {
  await messageHandler.sendRequest(createRequestMessage('showModal', options))
}

export async function showRightPanel(options: RightPanelOptions): Promise<void> {
  await messageHandler.sendRequest(createRequestMessage('showRightPanel', options))
}

export async function showCustomModal(options: CustomModalOptions): Promise<void> {
  await messageHandler.sendRequest(createRequestMessage('showCustomModal', options))
}

export async function showCustomRightPanel(options: CustomRightPanelOptions): Promise<void> {
  await messageHandler.sendRequest(createRequestMessage('showCustomRightPanel', options))
}

export async function closeModal(): Promise<void> {
  await messageHandler.sendRequest(createRequestMessage('closeModal'))
}

export async function closeRightPanel(): Promise<void> {
  await messageHandler.sendRequest(createRequestMessage('closeRightPanel'))
}

export async function alert(options: AlertOptions): Promise<void> {
  await messageHandler.sendRequest(createRequestMessage('alert', options))
}

export async function confirm(options: ConfirmOptions): Promise<boolean> {
  const response = await messageHandler.sendRequest(createRequestMessage('confirm', options))
  return response.result
}

export async function showToaster(options: ToasterOptions): Promise<void> {
  await messageHandler.sendRequest(createRequestMessage('showToaster', options))
}
