// @stack-spot/portal-extension@0.5.0/dist/layout.js
import { createRequestMessage, } from '@stack-spot/portal-extension-commons';
import { messageHandler } from './MessageHandler';
export async function showModal(options) {
    await messageHandler.sendRequest(createRequestMessage('showModal', options));
}
export async function showRightPanel(options) {
    await messageHandler.sendRequest(createRequestMessage('showRightPanel', options));
}
export async function showCustomModal(options) {
    await messageHandler.sendRequest(createRequestMessage('showCustomModal', options));
}
export async function showCustomRightPanel(options) {
    await messageHandler.sendRequest(createRequestMessage('showCustomRightPanel', options));
}
export async function closeModal() {
    await messageHandler.sendRequest(createRequestMessage('closeModal'));
}
export async function closeRightPanel() {
    await messageHandler.sendRequest(createRequestMessage('closeRightPanel'));
}
export async function alert(options) {
    await messageHandler.sendRequest(createRequestMessage('alert', options));
}
export async function confirm(options) {
    const response = await messageHandler.sendRequest(createRequestMessage('confirm', options));
    return response.result;
}
export async function showToaster(options) {
    await messageHandler.sendRequest(createRequestMessage('showToaster', options));
}
