// @citric/ui@6.5.5/src/Popover/utils.ts
import { css } from '@citric/core/dist/styled'

type PositionProps = 'left' | 'right' | 'top' | 'bottom'

export const getPopoverPosition = (position: PositionProps) => {
  const PopoverPosition = {
    top: css`
      bottom: 125%;
      left: 50%;
      transform: translateX(-50%);
    `,
    bottom: css`
      top: 135%;
      left: 50%;
      transform: translateX(-50%);
    `,
    right: css`
      top: 50%;
      transform: translateY(-50%);
      left: 110%;
    `,
    left: css`
      top: 50%;
      transform: translateY(-50%);
      right: 110%;
    `,
  }

  return PopoverPosition[position]
}

export const getPointerPosition = (position: PositionProps, color?: string) => {
  const pointerPosition = {
    top: css`
      top: 100%;
      left: 50%;
      margin-left: -5px;
      border-top-color: ${color};
    `,
    right: css`
      top: 50%;
      right: 100%;
      margin-top: -5px;
      border-right-color: ${color};
    `,
    bottom: css`
      bottom: 100%;
      left: 50%;
      margin-left: -5px;
      border-bottom-color: ${color};
    `,
    left: css`
      top: 50%;
      left: 100%;
      margin-top: -5px;
      border-left-color: ${color};
    `,
  }

  return pointerPosition[position]
}
