// @stack-spot/portal-layout@2.20.1/src/components/NotificationCenter/styled.ts
import { theme } from '@stack-spot/portal-theme'
import { styled } from 'styled-components'
import { Backdrop } from '../Backdrop'

export const NotificationBox = styled.div`
  .notification-button {
    border: none;
    background: transparent;
    margin: 0 40px;
    position: relative;

    &.loading {
      cursor: progress;
      opacity: 0.5;
    }

    &:before {
      content: '';
      position: absolute;
      right: 9px;
      top: 9px;
      width: 12px;
      height: 12px;
      border-radius: 50%;
      background-color: ${theme.color.danger['500']};
      transform: scale(0);
      transition: transform ease-in 0.3s;
    }

    &.unread:before {
      transform: scale(1);
    }
  }
`

export const StyledBackdrop = styled(Backdrop)`
  position: absolute;
  top: calc(var(--header-height) + 10px);
  right: 60px;
  box-shadow: 4px 4px 48px ${theme.color.danger.contrastText};

  .notification-panel {
    width: 368px;
    padding: 16px;
    border-radius: 4px;
    background-color: ${theme.color.light[300]};
    border: 1px solid ${theme.color.light[400]};
  }

  .filter-list {
    list-style: none;
    margin: 16px 0;
    padding: 0;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    .filter-btn {
      &:focus {
        border-color: transparent;
      }
      &[aria-pressed="true"] {
        border-color: ${theme.color.primary[500]};
      }
    }
  }

  .see-all {
    margin-top: 8px;
    width: 100%;
  }

  .error-feedback {
    flex-direction: column;
  }
`
