// @stack-spot/portal-extension@0.5.0/dist/network/clients.js
import { createNetworkProxy } from './proxy';
export const cloudRuntimesClient = createNetworkProxy('cloudRuntimesClient');
export const aiClient = createNetworkProxy('aiClient');
export const agentClient = createNetworkProxy('agentClient');
export const cloudAccountClient = createNetworkProxy('cloudAccountClient');
export const accountClient = createNetworkProxy('accountClient');
export const secretsClient = createNetworkProxy('secretsClient');
export const cloudServicesClient = createNetworkProxy('cloudServicesClient');
export const cloudPlatformClient = createNetworkProxy('cloudPlatformClient');
export const contentClient = createNetworkProxy('contentClient');
export const eventBusClient = createNetworkProxy('eventBusClient');
export const insightsClient = createNetworkProxy('insightsClient');
export const workflowClient = createNetworkProxy('workflowClient');
export const workspaceClient = createNetworkProxy('workspaceClient');
export const notificationClient = createNetworkProxy('notificationClient');
export const runtimeManagerClient = createNetworkProxy('runtimeManagerClient');
export const cloudPlatformHorizonClient = createNetworkProxy('cloudPlatformHorizonClient');
export const workspaceSearchClient = createNetworkProxy('workspaceSearchClient');
export const workspaceManagerClient = createNetworkProxy('workspaceManagerClient');
