// @stack-spot/portal-network@0.66.0/dist/api/eventBus.js
/**
 * EventBusApi
 * 0.2.7
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
export const defaults = {
    headers: {},
    baseUrl: "https://idp-event-bus-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://idp-event-bus-api.dev.stackspot.com"
};
export function sendEvents({ stackspotCustomerIp, sendEventsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/events", oazapfts.json({
        ...opts,
        method: "POST",
        body: sendEventsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    })));
}
export function sendEncryptedEvents({ stackspotCustomerIp, sendEncryptedEventsRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/events/workspace/${encodeURIComponent(workspaceSlug)}/event/${encodeURIComponent(eventName)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: sendEncryptedEventsRequest,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    })));
}
export function getPublicKey(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/events/key", {
        ...opts
    }));
}
