// @stack-spot/portal-network@0.66.0/dist/api/insights.js
/**
 * Insights API
 * v1.0.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "https://insights-insights-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://insights-insights-api.dev.stackspot.com"
};
/**
 * Create Panel
 */
export function save({ newPanelRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/panels", oazapfts.json({
        ...opts,
        method: "POST",
        body: newPanelRequest
    })));
}
/**
 * Add Panel to Dashboard Studio
 */
export function addToDashboardStudio({ accountSlug, studioSlug, addPanelToDashboardRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/panels/account/${encodeURIComponent(accountSlug)}/studio/${encodeURIComponent(studioSlug)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: addPanelToDashboardRequest
    })));
}
/**
 * Remove Panel from Dashboard Studio
 */
export function removeFromDashboardStudio({ accountSlug, studioSlug, panelId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/panels/account/${encodeURIComponent(accountSlug)}/studio/${encodeURIComponent(studioSlug)}${QS.query(QS.explode({
        panelId
    }))}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Create Dataset
 */
export function save1({ newDatasetRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/datasets", oazapfts.json({
        ...opts,
        method: "POST",
        body: newDatasetRequest
    })));
}
/**
 * Download Studio Data
 */
export function downloadStudioData({ accountSlug, studioSlug, startDate, endDate, stackspotCustomerIp }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/assets/account/${encodeURIComponent(accountSlug)}/studio/${encodeURIComponent(studioSlug)}/csv${QS.query(QS.explode({
        startDate,
        endDate
    }))}`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    }));
}
/**
 * Download Event account data file
 */
export function streamingAccountData({ accountSlug, startDate, endDate, dataset }, opts) {
    return oazapfts.ok(oazapfts.fetchBlob(`/v1/assets/account/${encodeURIComponent(accountSlug)}/export${QS.query(QS.explode({
        startDate,
        endDate,
        dataset
    }))}`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Download Account Data
 */
export function downloadAccountData({ accountSlug, startDate, endDate, stackspotCustomerIp }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/assets/account/${encodeURIComponent(accountSlug)}/csv${QS.query(QS.explode({
        startDate,
        endDate
    }))}`, {
        ...opts,
        method: "POST",
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    }));
}
/**
 * Get Dashboard Studio
 */
export function getDashboardStudio({ accountSlug, studioSlug, stackspotCustomerIp }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/dashboards/account/${encodeURIComponent(accountSlug)}/studio/${encodeURIComponent(studioSlug)}`, {
        ...opts,
        headers: oazapfts.mergeHeaders(opts?.headers, {
            "stackspot-customer-ip": stackspotCustomerIp
        })
    }));
}
