// @stack-spot/portal-network@0.66.0/dist/api/serviceCatalog.js
/**
 * OpenAPI definition
 * v0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "https://catalog-tech-products-catalog-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://catalog-tech-products-catalog-api.dev.stackspot.com"
};
/**
 * Get one product by product_id
 */
export function getOne({ workspaceId, productId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}`, {
        ...opts
    }));
}
/**
 * Update product's name and/or description
 */
export function updateProduct({ workspaceId, productId, productRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: productRequest
    })));
}
/**
 * Get one product by product_id
 */
export function deleteProduct({ workspaceId, productId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Change product status
 */
export function changeProductStatus({ workspaceId, productId, action }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}${QS.query(QS.explode({
        action
    }))}`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Get Client Application by ID
 */
export function getOneClientApplication({ clientApplicationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/client-applications/${encodeURIComponent(clientApplicationId)}`, {
        ...opts
    }));
}
/**
 * Update a Client Application
 */
export function updateClientApplication({ clientApplicationId, clientApplicationRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/client-applications/${encodeURIComponent(clientApplicationId)}`, oazapfts.json({
        ...opts,
        method: "PUT",
        body: clientApplicationRequest
    })));
}
/**
 * Delete a Client Application by ID
 */
export function deleteClientApplication({ clientApplicationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/client-applications/${encodeURIComponent(clientApplicationId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get all products by workspace Id
 */
export function getAll({ workspaceId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products`, {
        ...opts
    }));
}
/**
 * save product in a workspace Id
 */
export function saveProduct({ workspaceId, productWithApiAndEnvironmentRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products`, oazapfts.json({
        ...opts,
        method: "POST",
        body: productWithApiAndEnvironmentRequest
    })));
}
export function linkProductApi({ workspaceId, productId, apiId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts,
        method: "POST"
    }));
}
export function unlinkProductApi({ workspaceId, productId, apiId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Get one API by application id
 */
export function getOne1({ workspaceId, applicationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api`, {
        ...opts
    }));
}
/**
 * save API in a workspace Id
 */
export function saveApi({ workspaceId, applicationId, catalogApiRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api`, oazapfts.json({
        ...opts,
        method: "POST",
        body: catalogApiRequest
    })));
}
/**
 * delete one API by application id
 */
export function deleteApi({ workspaceId, applicationId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * get API definition
 */
export function getApiDefinition({ workspaceId, applicationId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api/environments/${encodeURIComponent(environmentId)}/definition`, {
        ...opts
    }));
}
/**
 * save API definition in a workspace Id
 */
export function saveApiDefinition({ workspaceId, applicationId, environmentId, gatewayUrl, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api/environments/${encodeURIComponent(environmentId)}/definition${QS.query(QS.explode({
        gatewayUrl
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * update API definition in a workspace Id
 */
export function updateApiDefinition({ workspaceId, applicationId, environmentId, gatewayUrl, connectionInterfaceSlug, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api/environments/${encodeURIComponent(environmentId)}/definition${QS.query(QS.explode({
        gatewayUrl,
        connectionInterfaceSlug
    }))}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Get all apis by workspace Id
 */
export function getAll1({ workspaceId, offset, limit, status }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis${QS.query(QS.explode({
        offset,
        limit,
        status
    }))}`, {
        ...opts
    }));
}
/**
 * save API in a workspace Id
 */
export function saveApi1({ workspaceId, catalogApiRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis`, oazapfts.json({
        ...opts,
        method: "POST",
        body: catalogApiRequest
    })));
}
/**
 * get API definition
 */
export function getApiDefinition1({ workspaceId, apiId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}/definition`, {
        ...opts
    }));
}
/**
 * save API definition in a workspace Id
 */
export function saveNewApiDefinition({ workspaceId, apiId, environmentId, gatewayUrl, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}/definition${QS.query(QS.explode({
        gatewayUrl
    }))}`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * update API definition in a workspace Id
 */
export function updateApiDefinition1({ workspaceId, apiId, environmentId, gatewayUrl, connectionInterfaceSlug, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}/definition${QS.query(QS.explode({
        gatewayUrl,
        connectionInterfaceSlug
    }))}`, oazapfts.json({
        ...opts,
        method: "PATCH",
        body
    })));
}
/**
 * Grant partner access permission to access the product
 */
export function givePermission({ partnerAccountId, environmentId, productId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Remove partner access permission from the product
 */
export function removePermission({ partnerAccountId, environmentId, productId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Grant partner access permission to access the API
 */
export function givePermission1({ partnerAccountId, environmentId, apiId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts,
        method: "POST"
    }));
}
/**
 * Remove partner access permission from the API
 */
export function removePermission1({ partnerAccountId, environmentId, apiId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Request access to many client application by Product
 */
export function createAccessRequestByProduct({ environmentId, productId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}/access-requests`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Request access to many client application by Api
 */
export function createAccessRequestByApi({ environmentId, apiId, body }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/access-requests`, oazapfts.json({
        ...opts,
        method: "POST",
        body
    })));
}
/**
 * Update access request by status and clientID
 */
export function updateAccessRequest({ environmentId, accessRequestId, accessRequestRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/environments/${encodeURIComponent(environmentId)}/access-requests/${encodeURIComponent(accessRequestId)}`, oazapfts.json({
        ...opts,
        method: "POST",
        body: accessRequestRequest
    })));
}
/**
 * Get all Client Applications from Account
 */
export function getClientApplication(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/partner/client-applications", {
        ...opts
    }));
}
/**
 * Save a new Client Application for Account
 */
export function createClientApplication({ clientApplicationRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/partner/client-applications", oazapfts.json({
        ...opts,
        method: "POST",
        body: clientApplicationRequest
    })));
}
/**
 * Update batch access request by status and clientID
 */
export function updateAccessRequestBatch({ accessRequestBatchRequest }, opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/partner/access-requests", oazapfts.json({
        ...opts,
        method: "POST",
        body: accessRequestBatchRequest
    })));
}
/**
 * Update API status by application id
 */
export function updateApiStatusByWorkspaceAndApplicationId({ workspaceId, applicationId, environmentId, action }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/applications/${encodeURIComponent(applicationId)}/api/environments/${encodeURIComponent(environmentId)}${QS.query(QS.explode({
        action
    }))}`, {
        ...opts,
        method: "PATCH"
    }));
}
/**
 * Update API status
 */
export function updateApiStatusByWorkspaceAndApiId({ workspaceId, apiId, envId, action }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(envId)}${QS.query(QS.explode({
        action
    }))}`, {
        ...opts,
        method: "PATCH"
    }));
}
export function getApisByProductId({ workspaceId, productId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/products/${encodeURIComponent(productId)}/apis`, {
        ...opts
    }));
}
/**
 * Get one API by api_id
 */
export function getOne2({ workspaceId, apiId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts
    }));
}
/**
 * Get all products of an API
 */
export function getProductsByApiId({ workspaceId, apiId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/products`, {
        ...opts
    }));
}
/**
 * get API Details
 */
export function apiDetails({ workspaceId, apiId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}`, {
        ...opts
    }));
}
/**
 * get API Definition
 */
export function getApiWorkspaceMethods({ workspaceId, apiId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/workspaces/${encodeURIComponent(workspaceId)}/apis/${encodeURIComponent(apiId)}/environments/${encodeURIComponent(environmentId)}/methods`, {
        ...opts
    }));
}
/**
 * List all the Account's Products
 */
export function getProducts(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/service-catalog/products", {
        ...opts
    }));
}
export function getApisByProductId1({ productId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}/apis`, {
        ...opts
    }));
}
export function getApisByProductId2({ productId, environmentId, accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}/apis`, {
        ...opts
    }));
}
/**
 * Get one product by product_id
 */
export function getOneProduct({ accountId, productId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}`, {
        ...opts
    }));
}
/**
 * Get one product by product_id
 */
export function getOneProduct1({ productId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/products/${encodeURIComponent(productId)}`, {
        ...opts
    }));
}
/**
 * Get all products
 */
export function getAllProducts({ accountId, environmentId, limit }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/products${QS.query(QS.explode({
        limit
    }))}`, {
        ...opts
    }));
}
/**
 * Get all products
 */
export function getAllProducts1({ environmentId, limit }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/products${QS.query(QS.explode({
        limit
    }))}`, {
        ...opts
    }));
}
/**
 * Get all products of an API
 */
export function getProductsByApiId1({ apiId, environmentId, accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/products`, {
        ...opts
    }));
}
/**
 * Get all products of an API
 */
export function getProductsByApiId2({ apiId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/products`, {
        ...opts
    }));
}
/**
 * get API definition
 */
export function getApiMethods({ accountId, apiId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/methods`, {
        ...opts
    }));
}
/**
 * get API definition
 */
export function getApiMethods1({ apiId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/methods`, {
        ...opts
    }));
}
/**
 * get API definition
 */
export function getApiDefinition2({ apiId, environmentId, accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/details`, {
        ...opts
    }));
}
/**
 * get API definition
 */
export function getApiDefinition3({ apiId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/details`, {
        ...opts
    }));
}
/**
 * Get one API by api_id
 */
export function getOneApi({ apiId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts
    }));
}
/**
 * Get one API by api_id
 */
export function getOneApi1({ apiId, environmentId, accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}`, {
        ...opts
    }));
}
/**
 * Get all apis
 */
export function getAllApis({ environmentId, limit }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/service-catalog/environments/${encodeURIComponent(environmentId)}/apis${QS.query(QS.explode({
        limit
    }))}`, {
        ...opts
    }));
}
/**
 * Get all apis
 */
export function getAllApis1({ accountId, environmentId, limit }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/service-catalog/environments/${encodeURIComponent(environmentId)}/apis${QS.query(QS.explode({
        limit
    }))}`, {
        ...opts
    }));
}
/**
 * List all the Account's Apis
 */
export function getApis(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/service-catalog/apis", {
        ...opts
    }));
}
/**
 * Get all partner APIs whose status is requested or granted
 */
export function getRequestedOrGrantedAccessApis({ partnerAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/requested-or-granted-access-apis`, {
        ...opts
    }));
}
/**
 * Get all Partner's products
 */
export function getProducts1({ partnerAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/products`, {
        ...opts
    }));
}
/**
 * Get all Partner's products by environment
 */
export function getProductsByEnv({ partnerAccountId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/products`, {
        ...opts
    }));
}
/**
 * Get all Partner's apis by environment
 */
export function getApisByEnv({ partnerAccountId, environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/environments/${encodeURIComponent(environmentId)}/apis`, {
        ...opts
    }));
}
/**
 * Get all Partner's apis
 */
export function getApis1({ partnerAccountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/${encodeURIComponent(partnerAccountId)}/apis`, {
        ...opts
    }));
}
/**
 * Get all Client Applications that are eligible to request access to API
 */
export function getEligibleClientApplications({ environmentId, apiId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/client-applications`, {
        ...opts
    }));
}
export function getAllApplications({ environmentId, apiId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/environments/${encodeURIComponent(environmentId)}/apis/${encodeURIComponent(apiId)}/approved-client-applications`, {
        ...opts
    }));
}
/**
 * access request list requested
 */
export function findAccessRequestRequestedStatus({ environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/environments/${encodeURIComponent(environmentId)}/access-requests`, {
        ...opts
    }));
}
/**
 * Access request list history
 */
export function findAccessRequestRequestedHistory({ environmentId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/partner/environments/${encodeURIComponent(environmentId)}/access-requests/history`, {
        ...opts
    }));
}
/**
 * Get allowed environments
 */
export function getEnvironmentsByAccountId({ accountId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/accounts/${encodeURIComponent(accountId)}/environments`, {
        ...opts
    }));
}
/**
 * Get health status
 */
export function healthz(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/healthz", {
        ...opts
    }));
}
