// @stack-spot/portal-network@0.66.0/dist/api/workspaceSearchEngine.js
/**
 * Search Engine API
 * v1.0.0
 * DO NOT MODIFY - This file has been generated using oazapfts.
 * See https://www.npmjs.com/package/oazapfts
 */
import * as Oazapfts from "@oazapfts/runtime";
import * as QS from "@oazapfts/runtime/query";
export const defaults = {
    headers: {},
    baseUrl: "https://workspace-search-engine-api.dev.stackspot.com",
};
const oazapfts = Oazapfts.runtime(defaults);
export const servers = {
    generatedServerUrl: "https://workspace-search-engine-api.dev.stackspot.com"
};
/**
 * Search all Stackspot resources using semantic search.
 */
export function search({ q, category, page }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/search${QS.query(QS.explode({
        q,
        category,
        page
    }))}`, {
        ...opts
    }));
}
/**
 * Get the logged-in user search history list.
 */
export function getUserHistory(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/search/history", {
        ...opts
    }));
}
/**
 * Clear the logged-in user search history.
 */
export function clearUserHistory(opts) {
    return oazapfts.ok(oazapfts.fetchJson("/v1/search/history", {
        ...opts,
        method: "DELETE"
    }));
}
/**
 * Delete the logged-in user specific search history item.
 */
export function deleteUserHistoryItem({ itemId }, opts) {
    return oazapfts.ok(oazapfts.fetchJson(`/v1/search/history/${encodeURIComponent(itemId)}`, {
        ...opts,
        method: "DELETE"
    }));
}
