// @stack-spot/portal-layout@2.20.1/src/components/Contact/show-contact-modal.tsx
import { Box, Button, Flex, IconBox, Text } from '@citric/core'
import { ExternalLink } from '@citric/icons'
import { Card } from '@citric/ui'
import { useAnchorTag } from '@stack-spot/portal-components/anchor'
import { Dictionary, useTranslate } from '@stack-spot/portal-translate'
import { ReactElement } from 'react'
import { overlay } from '../../LayoutOverlayManager'

interface CardItemProps {
  title: string,
  description: string,
  buttonText: string,
  onClick?: () => void,
  url?: string,
  icon: ReactElement,
  hasMargin?: boolean,
}

const CardItem = ({ title, description, hasMargin = false, buttonText, url, icon, onClick }: CardItemProps) => {
  const Link = useAnchorTag()
  return <Card sx={{ p: '7', bg: 'light.500', w: '100%' }}>
    <Flex flexDirection="column">
      <Text appearance="h5" mb="3">{title}</Text>
      <Text appearance="microtext1" colorScheme="light.700" mb="8">{description}</Text>
      <Box mt={hasMargin ? 5: 0}>
        <Link onClick={onClick} href={url} target="_blank">
          <Button size="md" colorScheme="inverse">
            {buttonText}
            <IconBox sx={{ ml: 2 }} colorIcon="inverse.contrastText">
              {icon}
            </IconBox>
          </Button>
        </Link>
      </Box>
    </Flex>
  </Card>
}

const ContactUsComponent = ({ serviceNowUrl }: {serviceNowUrl:string}) => {
  const t = useTranslate(dictionary)

  return <Flex sx={{ gap: '8px' }} w={12} mb={10} flexWrap="nowrap">
    <CardItem title={t.support} description={t.supportDescription} buttonText={t.supportButton} url={serviceNowUrl}
      icon={<ExternalLink />} />
  </Flex>
}


export const showContactUsModal = ({ title, serviceNowUrl }:{ serviceNowUrl:string, title: string }) => {
  overlay.showModal({
    title,
    children: <ContactUsComponent serviceNowUrl={serviceNowUrl} />,
    size: 'small',
  })
}

const dictionary = {
  en: {
    contactUs: 'Contact us',
    support: 'Support and feedback',
    supportDescription: 'Need help or have a feedback? Connect with our Support team',
    supportButton: 'Get in touch',
  },
  pt: {
    contactUs: 'Fale conosco',
    support: 'Suporte e feedback',
    supportDescription: 'Precisa de ajuda ou tem algum feedback? Fale com o nosso time de Suporte',
    supportButton: 'Entre em contato',
  },
} satisfies Dictionary
